{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00f332d4-1c1c-42c7-8382-7cb5fcab05d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyranges import PyRanges\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b272d5b-706a-4c50-9f84-cc3e31f479e3",
   "metadata": {},
   "source": [
    "<h1> T2T CHM13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b0442f5-74b7-4d26-9bc8-dab9a2b8ff71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_55371/1838600794.py:2: DtypeWarning: Columns (23,36) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  t2t_genes = pd.read_csv('../data/annotation_files/T2T-chm13-catLiftOffGenes.csv')\n"
     ]
    }
   ],
   "source": [
    "## Anotations\n",
    "t2t_genes = pd.read_csv('../data/annotation_files/T2T-chm13-catLiftOffGenes.csv')\n",
    "t2t_telomeres = pd.read_csv('../data/annotation_files/chm13v2.0_telomere.bed', sep='\\t', names=['Chromosome','Start', 'End'])\n",
    "t2t_sds = pd.read_csv('../data/annotation_files/chm13v2.0_SD.full.bed', sep='\\t')\n",
    "t2t_sds = t2t_sds[['#chr1', 'start1', 'end1']]\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "t2t_sds.columns =['Chromosome','Start','End']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d730b68f-174e-41e2-ac7a-b23e1aa2e6eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_55371/3804510778.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  chm13 = pd.read_csv('../data/call_cnvr_chm13/t2t-chm13_cnvs.txt', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "# is not working \n",
    "chm13 = pd.read_csv('../data/call_cnvr_chm13/t2t-chm13_cnvs.txt', sep='\\t')\n",
    "#chm13 = pd.read_csv('/Users/luciabazan/Downloads/AMR_Mayan_LP6005441-DNA_G07', sep='\\t', header= None)\n",
    "#chm13.columns = ['Chromosome', 'Start', 'End','CNV_Value','ext1','ext2','ext3','ext4','ext5']\n",
    "#chm13[[chm13['CNV_Value'] != '2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "55cc3484-958c-4a32-94f5-7b0e88698454",
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13 = chm13.dropna()\n",
    "chm13['Length_cnvr'] = chm13['End'] - chm13['Start']\n",
    "\n",
    "\n",
    "chm13 = chm13[chm13['Sample_ID'] == 'LP6005441-DNA_G07']\n",
    "#chm13['Sample_ID'] = 'LP6005441-DNA_G07'\n",
    "#chm13 = chm13[~chm13['CNV_Value'].isin([2])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "55b427d6-ec1f-4c01-8621-426f5f660809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNVR_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>CNV_Start</th>\n",
       "      <th>CNV_End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Length</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8285141</td>\n",
       "      <td>8289096</td>\n",
       "      <td>1</td>\n",
       "      <td>3956</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8292271</td>\n",
       "      <td>8294316</td>\n",
       "      <td>1</td>\n",
       "      <td>2046</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8294316</td>\n",
       "      <td>8297751</td>\n",
       "      <td>1</td>\n",
       "      <td>3436</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8297751</td>\n",
       "      <td>8303368</td>\n",
       "      <td>1</td>\n",
       "      <td>5618</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8304488</td>\n",
       "      <td>8306331</td>\n",
       "      <td>1</td>\n",
       "      <td>1844</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92805</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248364795</td>\n",
       "      <td>248366351</td>\n",
       "      <td>26</td>\n",
       "      <td>1557</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92806</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248366351</td>\n",
       "      <td>248367622</td>\n",
       "      <td>27</td>\n",
       "      <td>1272</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92807</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248367622</td>\n",
       "      <td>248368745</td>\n",
       "      <td>27</td>\n",
       "      <td>1124</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92808</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248368745</td>\n",
       "      <td>248370339</td>\n",
       "      <td>42</td>\n",
       "      <td>1595</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92809</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248370339</td>\n",
       "      <td>248387328</td>\n",
       "      <td>37</td>\n",
       "      <td>16990</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92810 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr        Start          End      CNVR_ID          Sample_ID  \\\n",
       "0      22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "1      22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "2      22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "3      22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "4      22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "...    ..          ...          ...          ...                ...   \n",
       "92805   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "92806   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "92807   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "92808   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "92809   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "\n",
       "       CNV_Start    CNV_End  CNV_Value  Length  Length_cnvr  \\\n",
       "0        8285141    8289096          1    3956   12764469.0   \n",
       "1        8292271    8294316          1    2046   12764469.0   \n",
       "2        8294316    8297751          1    3436   12764469.0   \n",
       "3        8297751    8303368          1    5618   12764469.0   \n",
       "4        8304488    8306331          1    1844   12764469.0   \n",
       "...          ...        ...        ...     ...          ...   \n",
       "92805  248364795  248366351         26    1557      82796.0   \n",
       "92806  248366351  248367622         27    1272      82796.0   \n",
       "92807  248367622  248368745         27    1124      82796.0   \n",
       "92808  248368745  248370339         42    1595      82796.0   \n",
       "92809  248370339  248387328         37   16990      82796.0   \n",
       "\n",
       "                  Sample   Region  \n",
       "0      LP6005441-DNA_G07  America  \n",
       "1      LP6005441-DNA_G07  America  \n",
       "2      LP6005441-DNA_G07  America  \n",
       "3      LP6005441-DNA_G07  America  \n",
       "4      LP6005441-DNA_G07  America  \n",
       "...                  ...      ...  \n",
       "92805  LP6005441-DNA_G07  America  \n",
       "92806  LP6005441-DNA_G07  America  \n",
       "92807  LP6005441-DNA_G07  America  \n",
       "92808  LP6005441-DNA_G07  America  \n",
       "92809  LP6005441-DNA_G07  America  \n",
       "\n",
       "[92810 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Keeping only samples that are on old cnvs \n",
    "chm13 = chm13.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])\n",
    "#chm13.drop(columns=['ext1','ext2','ext3','ext4','ext5'], inplace=True)\n",
    "chm13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d29ef0bb-a274-4317-9ad6-b1fff5735bfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    9.281000e+04\n",
       "mean     5.466143e+06\n",
       "std      1.126691e+07\n",
       "min      1.000000e+03\n",
       "25%      9.444700e+04\n",
       "50%      5.516410e+05\n",
       "75%      3.502745e+06\n",
       "max      4.509068e+07\n",
       "Name: Length_cnvr, dtype: float64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13['Length_cnvr'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ac0ab3f0-db9c-4b57-a3aa-06fdecb87f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13#.value_counts(by=[\"Sample_ID\"])\n",
    "chm13.columns = ['Chromosome', 'Start', 'End', 'CNVR_ID\t','Sample_ID', 'CNV_start', 'CNV_end','CNV_Value', 'Length', 'Length_cnvr','Sample','Region']\n",
    "\n",
    "chm13= chm13[chm13['Chromosome'] != 'M']\n",
    "chm13 = chm13[chm13['Chromosome'] != 'X']\n",
    "chm13 = chm13[chm13['Chromosome'] != 'Y']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dca9507e-4cd0-46d6-811b-db80775baa27",
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13['Chromosome'] = chm13['Chromosome'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a4ff43d7-b75b-406c-8601-e49e532c98b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>17.124031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>436980.0</td>\n",
       "      <td>452948.0</td>\n",
       "      <td>15968.0</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>453948.0</td>\n",
       "      <td>466894.0</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>518975.0</td>\n",
       "      <td>605035.0</td>\n",
       "      <td>86060.0</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>650556.0</td>\n",
       "      <td>662579.0</td>\n",
       "      <td>12023.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7675</th>\n",
       "      <td>9</td>\n",
       "      <td>148771046.0</td>\n",
       "      <td>149276995.0</td>\n",
       "      <td>505949.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7676</th>\n",
       "      <td>9</td>\n",
       "      <td>149278663.0</td>\n",
       "      <td>149484014.0</td>\n",
       "      <td>205351.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7677</th>\n",
       "      <td>9</td>\n",
       "      <td>149620306.0</td>\n",
       "      <td>149794372.0</td>\n",
       "      <td>174066.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7678</th>\n",
       "      <td>9</td>\n",
       "      <td>149803678.0</td>\n",
       "      <td>149871435.0</td>\n",
       "      <td>67757.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7679</th>\n",
       "      <td>9</td>\n",
       "      <td>150171200.0</td>\n",
       "      <td>150617247.0</td>\n",
       "      <td>446047.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7680 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome        Start          End  Length_cnvr  LP6005441-DNA_G07\n",
       "0               1          0.0     403567.0     403567.0          17.124031\n",
       "1               1     436980.0     452948.0      15968.0           6.000000\n",
       "2               1     453948.0     466894.0      12946.0           1.000000\n",
       "3               1     518975.0     605035.0      86060.0           3.000000\n",
       "4               1     650556.0     662579.0      12023.0           1.000000\n",
       "...           ...          ...          ...          ...                ...\n",
       "7675            9  148771046.0  149276995.0     505949.0           1.000000\n",
       "7676            9  149278663.0  149484014.0     205351.0           1.000000\n",
       "7677            9  149620306.0  149794372.0     174066.0           1.000000\n",
       "7678            9  149803678.0  149871435.0      67757.0           1.000000\n",
       "7679            9  150171200.0  150617247.0     446047.0           1.000000\n",
       "\n",
       "[7680 rows x 5 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv = chm13.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='CNV_Value',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "final_cnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1811726a-f1b0-48c7-9268-d5cb962533de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>CNV_start</th>\n",
       "      <th>CNV_end</th>\n",
       "      <th>Length</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9083</td>\n",
       "      <td>9084</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9083</td>\n",
       "      <td>10803</td>\n",
       "      <td>1721</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10803</td>\n",
       "      <td>14487</td>\n",
       "      <td>3685</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14487</td>\n",
       "      <td>18731</td>\n",
       "      <td>4245</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>18731</td>\n",
       "      <td>20013</td>\n",
       "      <td>1283</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27269</th>\n",
       "      <td>9</td>\n",
       "      <td>149267269</td>\n",
       "      <td>149269905</td>\n",
       "      <td>2637</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27270</th>\n",
       "      <td>9</td>\n",
       "      <td>149470985</td>\n",
       "      <td>149472450</td>\n",
       "      <td>1466</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27271</th>\n",
       "      <td>9</td>\n",
       "      <td>149662164</td>\n",
       "      <td>149666176</td>\n",
       "      <td>4013</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27272</th>\n",
       "      <td>9</td>\n",
       "      <td>149856286</td>\n",
       "      <td>149857564</td>\n",
       "      <td>1279</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27273</th>\n",
       "      <td>9</td>\n",
       "      <td>150225245</td>\n",
       "      <td>150233612</td>\n",
       "      <td>8368</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27274 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome  CNV_start    CNV_end  Length  LP6005441-DNA_G07\n",
       "0               1          0       9083    9084               44.0\n",
       "1               1       9083      10803    1721               34.0\n",
       "2               1      10803      14487    3685               20.0\n",
       "3               1      14487      18731    4245               14.0\n",
       "4               1      18731      20013    1283               14.0\n",
       "...           ...        ...        ...     ...                ...\n",
       "27269           9  149267269  149269905    2637                1.0\n",
       "27270           9  149470985  149472450    1466                1.0\n",
       "27271           9  149662164  149666176    4013                1.0\n",
       "27272           9  149856286  149857564    1279                1.0\n",
       "27273           9  150225245  150233612    8368                1.0\n",
       "\n",
       "[27274 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_cnv = chm13.pivot_table(index=[\"Chromosome\", 'CNV_start', 'CNV_end', 'Length'], \n",
    "                    columns='Sample', \n",
    "                    values='CNV_Value',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "final_cnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8ee60b12-5353-4636-9d8f-bab3747f6a83",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2t_deletions_gb = final_cnv[final_cnv['LP6005441-DNA_G07'] < 3]\n",
    "t2t_deletions_gb_mean = t2t_deletions_gb['LP6005441-DNA_G07'].mean()\n",
    "t2t_duplications_gb = final_cnv[final_cnv['LP6005441-DNA_G07'] > 1]\n",
    "t2t_duplications_gb_mean = t2t_duplications_gb['LP6005441-DNA_G07'].mean()\n",
    "t2t_duplications_gb = PyRanges(t2t_duplications_gb)\n",
    "t2t_deletions_gb = PyRanges(t2t_deletions_gb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c3dd5fce-15e1-4360-ab58-17d628a16655",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>17.124031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>436980</td>\n",
       "      <td>452948</td>\n",
       "      <td>15968.0</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>453948</td>\n",
       "      <td>466894</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>518975</td>\n",
       "      <td>605035</td>\n",
       "      <td>86060.0</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>650556</td>\n",
       "      <td>662579</td>\n",
       "      <td>12023.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7675</th>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>26278097</td>\n",
       "      <td>26278097.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7676</th>\n",
       "      <td>20</td>\n",
       "      <td>29521296</td>\n",
       "      <td>66210255</td>\n",
       "      <td>36688959.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7677</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7678</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>137.948830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7679</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>9.741935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7680 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+---------------+---------------------+\n",
       "| Chromosome   | Start     | End       | Length_cnvr   | LP6005441-DNA_G07   |\n",
       "| (category)   | (int64)   | (int64)   | (float64)     | (float64)           |\n",
       "|--------------+-----------+-----------+---------------+---------------------|\n",
       "| 1            | 0         | 403567    | 403567.0      | 17.124031007751938  |\n",
       "| 1            | 436980    | 452948    | 15968.0       | 6.0                 |\n",
       "| 1            | 453948    | 466894    | 12946.0       | 1.0                 |\n",
       "| 1            | 518975    | 605035    | 86060.0       | 3.0                 |\n",
       "| ...          | ...       | ...       | ...           | ...                 |\n",
       "| 20           | 29521296  | 66210255  | 36688959.0    | 1.0                 |\n",
       "| 20           | 0         | 26278097  | 26278097.0    | 1.0                 |\n",
       "| 21           | 0         | 45090682  | 45090682.0    | 1.0                 |\n",
       "| 22           | 15787747  | 51324926  | 35537179.0    | 9.741935483870968   |\n",
       "| 22           | 0         | 12764469  | 12764469.0    | 137.94882961350027  |\n",
       "+--------------+-----------+-----------+---------------+---------------------+\n",
       "Unstranded PyRanges object has 7,680 rows and 5 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_cnv_ranges = PyRanges(final_cnv)\n",
    "final_cnv_ranges "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1ebf6094-c3e2-453d-b1f9-3b733adfb293",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2t_genes.columns = ['Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'thickStart', 'thickEnd', 'reserved', 'blockCount', 'blockSizes',\n",
    "       'chromStarts', 'name2', 'cdsStartStat', 'cdsEndStat', 'exonFrames',\n",
    "       'txId', 'type', 'geneName', 'geneType', 'sourceGene',\n",
    "       'sourceTranscript', 'alignmentId', 'alternativeSourceTranscripts',\n",
    "       'Paralogy', 'UnfilteredParalogy', 'collapsedGeneIds',\n",
    "       'collapsedGeneNames', 'frameshift', 'exonAnnotationSupport',\n",
    "       'intronAnnotationSupport', 'transcriptClass', 'transcriptModes',\n",
    "       'validStart', 'validStop', 'properOrf', 'extra_paralog']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7ab381cf-130e-4900-b541-f7756d20c889",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2t_genes['Chromosome'] = t2t_genes['Chromosome'].str.replace('chr', '')\n",
    "t2t_genes = PyRanges(t2t_genes)\n",
    "\n",
    "t2t_telomeres['Chromosome'] = t2t_telomeres['Chromosome'].str.replace('chr', '')\n",
    "t2t_telomeres = PyRanges(t2t_telomeres)\n",
    "\n",
    "t2t_sds['Chromosome'] = t2t_sds['Chromosome'].str.replace('chr', '')\n",
    "t2t_sds = PyRanges(t2t_sds)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "270bf858-3acc-485c-90d3-88cb319bd888",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6679ce82-0b3c-45fd-97d4-3fced891d9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### t2t genes \n",
    "coordinates_genes = final_cnv_ranges.nearest(t2t_genes)\n",
    "coordinates_genes = coordinates_genes.as_df()\n",
    "coordinates_genes = coordinates_genes[coordinates_genes['Distance'] <1 ]\n",
    "\n",
    "## T2T telomeres\n",
    "coordinates_telomeres = final_cnv_ranges.nearest(t2t_telomeres)\n",
    "coordinates_telomeres = coordinates_telomeres.as_df()\n",
    "coordinates_telomeres = coordinates_telomeres[coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs \n",
    "coordinates_sds = final_cnv_ranges.nearest(t2t_sds)\n",
    "coordinates_sds = coordinates_sds.as_df()\n",
    "coordinates_sds = coordinates_sds[coordinates_sds['Distance'] <1 ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "975af7c8-8fdf-4a44-934a-68192a6ad882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>frameshift</th>\n",
       "      <th>exonAnnotationSupport</th>\n",
       "      <th>intronAnnotationSupport</th>\n",
       "      <th>transcriptClass</th>\n",
       "      <th>transcriptModes</th>\n",
       "      <th>validStart</th>\n",
       "      <th>validStop</th>\n",
       "      <th>properOrf</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>17.124031</td>\n",
       "      <td>11135</td>\n",
       "      <td>12457</td>\n",
       "      <td>AL627309.3-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>436980</td>\n",
       "      <td>452948</td>\n",
       "      <td>15968.0</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>431292</td>\n",
       "      <td>444671</td>\n",
       "      <td>ISG15-202</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>453948</td>\n",
       "      <td>466894</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>450246</td>\n",
       "      <td>486181</td>\n",
       "      <td>AGRN-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>518975</td>\n",
       "      <td>605035</td>\n",
       "      <td>86060.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>510430</td>\n",
       "      <td>544696</td>\n",
       "      <td>C1orf159-204</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1,1,1,1,1,1,1,1,1</td>\n",
       "      <td>2,2,2,2,2,2,1,1,1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>650556</td>\n",
       "      <td>662579</td>\n",
       "      <td>12023.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>645364</td>\n",
       "      <td>660458</td>\n",
       "      <td>SDF4-202</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1,1,1,1</td>\n",
       "      <td>2,2,1,1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7675</th>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>26278097</td>\n",
       "      <td>26278097.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5700</td>\n",
       "      <td>7747</td>\n",
       "      <td>AL078621.3-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7676</th>\n",
       "      <td>20</td>\n",
       "      <td>29521296</td>\n",
       "      <td>66210255</td>\n",
       "      <td>36688959.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>29561837</td>\n",
       "      <td>29599815</td>\n",
       "      <td>FRG1CP-204</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1,1,1</td>\n",
       "      <td>1,1,1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7677</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>45408</td>\n",
       "      <td>46269</td>\n",
       "      <td>BNIP3P41-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7678</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>137.948830</td>\n",
       "      <td>47083</td>\n",
       "      <td>47944</td>\n",
       "      <td>BNIP3P41-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7679</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>9.741935</td>\n",
       "      <td>15801274</td>\n",
       "      <td>15818302</td>\n",
       "      <td>AP001464.1-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6267 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome     Start       End  Length_cnvr  LP6005441-DNA_G07   Start_b  \\\n",
       "0             1         0    403567     403567.0          17.124031     11135   \n",
       "1             1    436980    452948      15968.0           6.000000    431292   \n",
       "2             1    453948    466894      12946.0           1.000000    450246   \n",
       "3             1    518975    605035      86060.0           3.000000    510430   \n",
       "4             1    650556    662579      12023.0           1.000000    645364   \n",
       "...         ...       ...       ...          ...                ...       ...   \n",
       "7675         20         0  26278097   26278097.0           1.000000      5700   \n",
       "7676         20  29521296  66210255   36688959.0           1.000000  29561837   \n",
       "7677         21         0  45090682   45090682.0           1.000000     45408   \n",
       "7678         22         0  12764469   12764469.0         137.948830     47083   \n",
       "7679         22  15787747  51324926   35537179.0           9.741935  15801274   \n",
       "\n",
       "         End_b          name  score strand  ...  frameshift  \\\n",
       "0        12457  AL627309.3-1      0      -  ...         NaN   \n",
       "1       444671     ISG15-202      0      +  ...         NaN   \n",
       "2       486181      AGRN-201      0      +  ...       False   \n",
       "3       544696  C1orf159-204      0      -  ...         NaN   \n",
       "4       660458      SDF4-202      0      -  ...         NaN   \n",
       "...        ...           ...    ...    ...  ...         ...   \n",
       "7675      7747  AL078621.3-1      0      -  ...         NaN   \n",
       "7676  29599815    FRG1CP-204      0      +  ...         NaN   \n",
       "7677     46269    BNIP3P41-1      0      -  ...         NaN   \n",
       "7678     47944    BNIP3P41-1      0      -  ...         NaN   \n",
       "7679  15818302  AP001464.1-1      0      -  ...         NaN   \n",
       "\n",
       "                                  exonAnnotationSupport  \\\n",
       "0                                                   NaN   \n",
       "1                                                 1,1,1   \n",
       "2     1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "3                               1,1,1,1,1,1,1,1,1,1,1,1   \n",
       "4                                         1,1,1,1,1,1,1   \n",
       "...                                                 ...   \n",
       "7675                                                NaN   \n",
       "7676                                        1,1,1,1,1,1   \n",
       "7677                                                NaN   \n",
       "7678                                                NaN   \n",
       "7679                                                NaN   \n",
       "\n",
       "                                intronAnnotationSupport  transcriptClass  \\\n",
       "0                                                   NaN          paralog   \n",
       "1                                                   1,1         ortholog   \n",
       "2     2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...         ortholog   \n",
       "3                                 2,2,2,2,2,2,1,1,1,1,1         ortholog   \n",
       "4                                           2,2,1,1,1,1         ortholog   \n",
       "...                                                 ...              ...   \n",
       "7675                                                NaN          paralog   \n",
       "7676                                          1,1,1,1,1         ortholog   \n",
       "7677                                                NaN          paralog   \n",
       "7678                                                NaN          paralog   \n",
       "7679                                                NaN          paralog   \n",
       "\n",
       "     transcriptModes validStart validStop properOrf extra_paralog Distance  \n",
       "0            Liftoff        NaN       NaN       NaN          True        0  \n",
       "1           transMap       True      True      True         False        0  \n",
       "2           transMap       True      True      True         False        0  \n",
       "3           transMap       True      True      True         False        0  \n",
       "4           transMap       True      True      True         False        0  \n",
       "...              ...        ...       ...       ...           ...      ...  \n",
       "7675         Liftoff        NaN       NaN       NaN          True        0  \n",
       "7676        transMap       True      True      True         False        0  \n",
       "7677         Liftoff        NaN       NaN       NaN          True        0  \n",
       "7678         Liftoff        NaN       NaN       NaN          True        0  \n",
       "7679         Liftoff        NaN       NaN       NaN          True        0  \n",
       "\n",
       "[6267 rows x 42 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4428325c-0e80-48a8-a7e4-85bf0538832f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>frameshift</th>\n",
       "      <th>exonAnnotationSupport</th>\n",
       "      <th>intronAnnotationSupport</th>\n",
       "      <th>transcriptClass</th>\n",
       "      <th>transcriptModes</th>\n",
       "      <th>validStart</th>\n",
       "      <th>validStop</th>\n",
       "      <th>properOrf</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>17.124031</td>\n",
       "      <td>11135</td>\n",
       "      <td>12457</td>\n",
       "      <td>AL627309.3-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>436980</td>\n",
       "      <td>452948</td>\n",
       "      <td>15968.0</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>431292</td>\n",
       "      <td>444671</td>\n",
       "      <td>ISG15-202</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>453948</td>\n",
       "      <td>466894</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>450246</td>\n",
       "      <td>486181</td>\n",
       "      <td>AGRN-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>518975</td>\n",
       "      <td>605035</td>\n",
       "      <td>86060.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>510430</td>\n",
       "      <td>544696</td>\n",
       "      <td>C1orf159-204</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1,1,1,1,1,1,1,1,1</td>\n",
       "      <td>2,2,2,2,2,2,1,1,1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>650556</td>\n",
       "      <td>662579</td>\n",
       "      <td>12023.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>645364</td>\n",
       "      <td>660458</td>\n",
       "      <td>SDF4-202</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1,1,1,1</td>\n",
       "      <td>2,2,1,1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7675</th>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>26278097</td>\n",
       "      <td>26278097.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5700</td>\n",
       "      <td>7747</td>\n",
       "      <td>AL078621.3-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7676</th>\n",
       "      <td>20</td>\n",
       "      <td>29521296</td>\n",
       "      <td>66210255</td>\n",
       "      <td>36688959.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>29561837</td>\n",
       "      <td>29599815</td>\n",
       "      <td>FRG1CP-204</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1,1,1</td>\n",
       "      <td>1,1,1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7677</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>45408</td>\n",
       "      <td>46269</td>\n",
       "      <td>BNIP3P41-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7678</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>137.948830</td>\n",
       "      <td>47083</td>\n",
       "      <td>47944</td>\n",
       "      <td>BNIP3P41-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7679</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>9.741935</td>\n",
       "      <td>15801274</td>\n",
       "      <td>15818302</td>\n",
       "      <td>AP001464.1-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6267 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome     Start       End  Length_cnvr  LP6005441-DNA_G07   Start_b  \\\n",
       "0             1         0    403567     403567.0          17.124031     11135   \n",
       "1             1    436980    452948      15968.0           6.000000    431292   \n",
       "2             1    453948    466894      12946.0           1.000000    450246   \n",
       "3             1    518975    605035      86060.0           3.000000    510430   \n",
       "4             1    650556    662579      12023.0           1.000000    645364   \n",
       "...         ...       ...       ...          ...                ...       ...   \n",
       "7675         20         0  26278097   26278097.0           1.000000      5700   \n",
       "7676         20  29521296  66210255   36688959.0           1.000000  29561837   \n",
       "7677         21         0  45090682   45090682.0           1.000000     45408   \n",
       "7678         22         0  12764469   12764469.0         137.948830     47083   \n",
       "7679         22  15787747  51324926   35537179.0           9.741935  15801274   \n",
       "\n",
       "         End_b          name  score strand  ...  frameshift  \\\n",
       "0        12457  AL627309.3-1      0      -  ...         NaN   \n",
       "1       444671     ISG15-202      0      +  ...         NaN   \n",
       "2       486181      AGRN-201      0      +  ...       False   \n",
       "3       544696  C1orf159-204      0      -  ...         NaN   \n",
       "4       660458      SDF4-202      0      -  ...         NaN   \n",
       "...        ...           ...    ...    ...  ...         ...   \n",
       "7675      7747  AL078621.3-1      0      -  ...         NaN   \n",
       "7676  29599815    FRG1CP-204      0      +  ...         NaN   \n",
       "7677     46269    BNIP3P41-1      0      -  ...         NaN   \n",
       "7678     47944    BNIP3P41-1      0      -  ...         NaN   \n",
       "7679  15818302  AP001464.1-1      0      -  ...         NaN   \n",
       "\n",
       "                                  exonAnnotationSupport  \\\n",
       "0                                                   NaN   \n",
       "1                                                 1,1,1   \n",
       "2     1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "3                               1,1,1,1,1,1,1,1,1,1,1,1   \n",
       "4                                         1,1,1,1,1,1,1   \n",
       "...                                                 ...   \n",
       "7675                                                NaN   \n",
       "7676                                        1,1,1,1,1,1   \n",
       "7677                                                NaN   \n",
       "7678                                                NaN   \n",
       "7679                                                NaN   \n",
       "\n",
       "                                intronAnnotationSupport  transcriptClass  \\\n",
       "0                                                   NaN          paralog   \n",
       "1                                                   1,1         ortholog   \n",
       "2     2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...         ortholog   \n",
       "3                                 2,2,2,2,2,2,1,1,1,1,1         ortholog   \n",
       "4                                           2,2,1,1,1,1         ortholog   \n",
       "...                                                 ...              ...   \n",
       "7675                                                NaN          paralog   \n",
       "7676                                          1,1,1,1,1         ortholog   \n",
       "7677                                                NaN          paralog   \n",
       "7678                                                NaN          paralog   \n",
       "7679                                                NaN          paralog   \n",
       "\n",
       "     transcriptModes validStart validStop properOrf extra_paralog Distance  \n",
       "0            Liftoff        NaN       NaN       NaN          True        0  \n",
       "1           transMap       True      True      True         False        0  \n",
       "2           transMap       True      True      True         False        0  \n",
       "3           transMap       True      True      True         False        0  \n",
       "4           transMap       True      True      True         False        0  \n",
       "...              ...        ...       ...       ...           ...      ...  \n",
       "7675         Liftoff        NaN       NaN       NaN          True        0  \n",
       "7676        transMap       True      True      True         False        0  \n",
       "7677         Liftoff        NaN       NaN       NaN          True        0  \n",
       "7678         Liftoff        NaN       NaN       NaN          True        0  \n",
       "7679         Liftoff        NaN       NaN       NaN          True        0  \n",
       "\n",
       "[6267 rows x 42 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "19a26eee-0ee7-43ca-a74c-ca187b130d69",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Deletions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2edbccd4-87d0-49d4-9161-0ee32c091008",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Genes \n",
    "deletions_t2t_coordinates_genes = t2t_deletions_gb.nearest(t2t_genes)\n",
    "deletions_t2t_coordinates_genes = deletions_t2t_coordinates_genes.as_df()\n",
    "deletions_t2t_coordinates_genes = deletions_t2t_coordinates_genes[deletions_t2t_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_mean_individual_genes= deletions_t2t_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_t2t_coordinates_telomeres = t2t_deletions_gb.nearest(t2t_telomeres)\n",
    "deletions_t2t_coordinates_telomeres = deletions_t2t_coordinates_telomeres.as_df()\n",
    "deletions_t2t_coordinates_telomeres = deletions_t2t_coordinates_telomeres[deletions_t2t_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_t2t_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_t2t_mean_individual_telomeres= deletions_t2t_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_t2t_coordinates_sds = t2t_deletions_gb.nearest(t2t_sds)\n",
    "deletions_t2t_coordinates_sds = deletions_t2t_coordinates_sds.as_df()\n",
    "deletions_t2t_coordinates_sds = deletions_t2t_coordinates_sds[deletions_t2t_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_mean_individual_sds= deletions_t2t_coordinates_sds['LP6005441-DNA_G07'].mean()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "49a1d1e1-3ea6-4c9a-a9fb-39eb950a80ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bb365324-be37-4e25-b96e-cc6640763704",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Genes \n",
    "duplications_t2t_coordinates_genes = t2t_duplications_gb.nearest(t2t_genes)\n",
    "duplications_t2t_coordinates_genes = duplications_t2t_coordinates_genes.as_df()\n",
    "duplications_t2t_coordinates_genes = duplications_t2t_coordinates_genes[duplications_t2t_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_mean_individual_genes= duplications_t2t_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_t2t_coordinates_telomeres = t2t_duplications_gb.nearest(t2t_telomeres)\n",
    "duplications_t2t_coordinates_telomeres = duplications_t2t_coordinates_telomeres.as_df()\n",
    "duplications_t2t_coordinates_telomeres = duplications_t2t_coordinates_telomeres[duplications_t2t_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_t2t_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_t2t_mean_individual_telomeres= duplications_t2t_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_t2t_coordinates_sds = t2t_duplications_gb.nearest(t2t_sds)\n",
    "duplications_t2t_coordinates_sds = duplications_t2t_coordinates_sds.as_df()\n",
    "duplications_t2t_coordinates_sds = duplications_t2t_coordinates_sds[duplications_t2t_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_mean_individual_sds= duplications_t2t_coordinates_sds['LP6005441-DNA_G07'].mean()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccab1fb8-1a97-43df-9041-235962b1cff4",
   "metadata": {},
   "source": [
    "<h2> T2T CNVnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d68610aa-58f1-4657-9dde-03ef84a89829",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>25001</td>\n",
       "      <td>45000</td>\n",
       "      <td>3.3208</td>\n",
       "      <td>19999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>48001</td>\n",
       "      <td>78000</td>\n",
       "      <td>3.2608</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>88001</td>\n",
       "      <td>96000</td>\n",
       "      <td>0.5038</td>\n",
       "      <td>7999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>2133001</td>\n",
       "      <td>2162000</td>\n",
       "      <td>7.1520</td>\n",
       "      <td>28999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>2193001</td>\n",
       "      <td>2223000</td>\n",
       "      <td>5.9894</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408</th>\n",
       "      <td>22</td>\n",
       "      <td>24379001</td>\n",
       "      <td>24409000</td>\n",
       "      <td>3.0316</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3409</th>\n",
       "      <td>22</td>\n",
       "      <td>24409001</td>\n",
       "      <td>24427000</td>\n",
       "      <td>1.0380</td>\n",
       "      <td>17999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3410</th>\n",
       "      <td>22</td>\n",
       "      <td>24430001</td>\n",
       "      <td>24463000</td>\n",
       "      <td>1.1298</td>\n",
       "      <td>32999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3412</th>\n",
       "      <td>22</td>\n",
       "      <td>39109001</td>\n",
       "      <td>39114000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413</th>\n",
       "      <td>22</td>\n",
       "      <td>39369001</td>\n",
       "      <td>39373000</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1855 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome     Start       End  CNV_value  Length_cnvr\n",
       "0             1     25001     45000     3.3208        19999\n",
       "1             1     48001     78000     3.2608        29999\n",
       "2             1     88001     96000     0.5038         7999\n",
       "6             1   2133001   2162000     7.1520        28999\n",
       "8             1   2193001   2223000     5.9894        29999\n",
       "...         ...       ...       ...        ...          ...\n",
       "3408         22  24379001  24409000     3.0316        29999\n",
       "3409         22  24409001  24427000     1.0380        17999\n",
       "3410         22  24430001  24463000     1.1298        32999\n",
       "3412         22  39109001  39114000     0.0000         4999\n",
       "3413         22  39369001  39373000     0.0014         3999\n",
       "\n",
       "[1855 rows x 5 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header_list = ['Type', 'Position', 'Lenght', 'RD', 'T-TEST-EVAL', 'GAUSSIAN-EVAL', 'T-TEST-EVAL(MIDDLE)','GAUSSIAN-EVAL(MIDDLE)', 'FRAQ-READ-QUALITY', 't', 't2']\n",
    "\n",
    "t2t_data = pd.read_csv('/Users/luciabazan/t2t_LP6005441-DNA_G07_cnvs.tsv', sep='\\t' , names=header_list)\n",
    "t2t_data\n",
    "\n",
    "\n",
    "t2t_data = t2t_data[t2t_data['T-TEST-EVAL'] <= 0.05]\n",
    "t2t_data\n",
    "\n",
    "cnvnator_data = t2t_data\n",
    "\n",
    "#SEPARATE COORDINATES\n",
    "coordinates = cnvnator_data['Position'].str.split(':', expand=True)\n",
    "CHROMOSOME = coordinates[0]\n",
    "\n",
    "split_position = coordinates[1].str.split('-', expand=True)\n",
    "START = split_position[0]\n",
    "STOP = split_position[1]\n",
    "\n",
    "cnvnator_style= pd.DataFrame()\n",
    "cnvnator_style['Chromosome'] = CHROMOSOME\n",
    "cnvnator_style['Start'] = START\n",
    "cnvnator_style['End'] = STOP\n",
    "cnvnator_style['CNV_value'] = cnvnator_data['RD']\n",
    "\n",
    "\n",
    "cnvnator_style['Start'] = cnvnator_style['Start'].astype(int)\n",
    "cnvnator_style['End'] = cnvnator_style['End'].astype(int)\n",
    "cnvnator_style['CNV_value'] = cnvnator_style['CNV_value'].astype(float)\n",
    "\n",
    "#cnvnator_style['Type'] = cnvnator_data['Type']\n",
    "\n",
    "cnvnator_style['CNV_value'] = cnvnator_style['CNV_value'] * 2\n",
    "\n",
    "cnvnator_style['Chromosome'] = cnvnator_style['Chromosome'].str.replace('chr','')\n",
    "\n",
    "t2t_cnvnator = cnvnator_style\n",
    "t2t_cnvnator['Length_cnvr'] = t2t_cnvnator['End'] - t2t_cnvnator['Start']\n",
    "t2t_cnvnator = t2t_cnvnator[t2t_cnvnator['Chromosome'] != 'M']\n",
    "t2t_cnvnator = t2t_cnvnator[t2t_cnvnator['Chromosome'] != 'X']\n",
    "t2t_cnvnator = t2t_cnvnator[t2t_cnvnator['Chromosome'] != 'Y']\n",
    "t2t_cnvnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5c1906b5-86f4-481e-89e2-42c528a62792",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>25001</td>\n",
       "      <td>45000</td>\n",
       "      <td>3.3208</td>\n",
       "      <td>19999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>48001</td>\n",
       "      <td>78000</td>\n",
       "      <td>3.2608</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>88001</td>\n",
       "      <td>96000</td>\n",
       "      <td>0.5038</td>\n",
       "      <td>7999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2133001</td>\n",
       "      <td>2162000</td>\n",
       "      <td>7.1520</td>\n",
       "      <td>28999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2193001</td>\n",
       "      <td>2223000</td>\n",
       "      <td>5.9894</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850</th>\n",
       "      <td>22</td>\n",
       "      <td>24379001</td>\n",
       "      <td>24409000</td>\n",
       "      <td>3.0316</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1851</th>\n",
       "      <td>22</td>\n",
       "      <td>24409001</td>\n",
       "      <td>24427000</td>\n",
       "      <td>1.0380</td>\n",
       "      <td>17999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1852</th>\n",
       "      <td>22</td>\n",
       "      <td>24430001</td>\n",
       "      <td>24463000</td>\n",
       "      <td>1.1298</td>\n",
       "      <td>32999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>22</td>\n",
       "      <td>39109001</td>\n",
       "      <td>39114000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>22</td>\n",
       "      <td>39369001</td>\n",
       "      <td>39373000</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1855 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "| Chromosome   | Start     | End       | CNV_value   | Length_cnvr   |\n",
       "| (category)   | (int64)   | (int64)   | (float64)   | (int64)       |\n",
       "|--------------+-----------+-----------+-------------+---------------|\n",
       "| 1            | 25001     | 45000     | 3.3208      | 19999         |\n",
       "| 1            | 48001     | 78000     | 3.2608      | 29999         |\n",
       "| 1            | 88001     | 96000     | 0.5038      | 7999          |\n",
       "| 1            | 2133001   | 2162000   | 7.152       | 28999         |\n",
       "| ...          | ...       | ...       | ...         | ...           |\n",
       "| 22           | 24409001  | 24427000  | 1.038       | 17999         |\n",
       "| 22           | 24430001  | 24463000  | 1.1298      | 32999         |\n",
       "| 22           | 39109001  | 39114000  | 0.0         | 4999          |\n",
       "| 22           | 39369001  | 39373000  | 0.0014      | 3999          |\n",
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "Unstranded PyRanges object has 1,855 rows and 5 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2t_deletions_cnvnator = t2t_cnvnator[t2t_cnvnator['CNV_value'] < 2]\n",
    "t2t_deletions_cnvnator_mean = t2t_deletions_cnvnator['CNV_value'].mean()\n",
    "t2t_duplications_cnvnator = t2t_cnvnator[t2t_cnvnator['CNV_value'] > 2]\n",
    "t2t_duplications_cnvnator_mean = t2t_duplications_cnvnator['CNV_value'].mean()\n",
    "t2t_duplications_cnvnator = PyRanges(t2t_duplications_cnvnator)\n",
    "t2t_deletions_cnvnator = PyRanges(t2t_deletions_cnvnator)\n",
    "\n",
    "t2t_cnvnator= PyRanges(t2t_cnvnator)\n",
    "t2t_cnvnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "083a418f-d419-43af-9452-5a3a7d30d028",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8627f457-8192-48d5-93cf-0caf1a2186ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### t2t genes \n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator.nearest(t2t_genes)\n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator_coordinates_genes.as_df()\n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator_coordinates_genes[t2t_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "\n",
    "## T2T telomeres\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator.nearest(t2t_telomeres)\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator_coordinates_telomeres.as_df()\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator_coordinates_telomeres[t2t_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs \n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator.nearest(t2t_sds)\n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator_coordinates_sds.as_df()\n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator_coordinates_sds[t2t_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b938fddc-1393-401e-9c13-b20443820381",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_55371/2452310715.py:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  deletions_t2t_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "#Deletions\n",
    "#### Genes \n",
    "deletions_t2t_cnvnator_coordinates_genes = t2t_deletions_cnvnator.nearest(t2t_genes)\n",
    "deletions_t2t_cnvnator_coordinates_genes = deletions_t2t_cnvnator_coordinates_genes.as_df()\n",
    "deletions_t2t_cnvnator_coordinates_genes = deletions_t2t_cnvnator_coordinates_genes[deletions_t2t_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_cnvnator_mean_individual_genes= deletions_t2t_cnvnator_coordinates_genes['CNV_value'].mean()\n",
    "deletions_t2t_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_t2t_cnvnator_coordinates_telomeres = t2t_deletions_cnvnator.nearest(t2t_telomeres)\n",
    "deletions_t2t_cnvnator_coordinates_telomeres = deletions_t2t_cnvnator_coordinates_telomeres.as_df()\n",
    "deletions_t2t_cnvnator_coordinates_telomeres = deletions_t2t_cnvnator_coordinates_telomeres[deletions_t2t_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_t2t_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_t2t_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_cnvnator_mean_individual_telomeres= deletions_t2t_cnvnator_coordinates_telomeres['CNV_value'].mean()\n",
    "deletions_t2t_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_t2t_cnvnator_coordinates_sds = t2t_deletions_cnvnator.nearest(t2t_sds)\n",
    "deletions_t2t_cnvnator_coordinates_sds = deletions_t2t_cnvnator_coordinates_sds.as_df()\n",
    "deletions_t2t_cnvnator_coordinates_sds = deletions_t2t_cnvnator_coordinates_sds[deletions_t2t_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_cnvnator_mean_individual_sds= deletions_t2t_cnvnator_coordinates_sds['CNV_value'].mean()\n",
    "deletions_t2t_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8e147424-6af8-4dec-a7d2-afd746f88002",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.981063976945245"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deletions_t2t_cnvnator_mean_individual_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "59b41445-b950-4946-90e5-ae4dd6f8e960",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_55371/4006543835.py:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  duplications_t2t_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3.5933225225225227"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Duplications\n",
    "#### Genes \n",
    "duplications_t2t_cnvnator_coordinates_genes = t2t_duplications_cnvnator.nearest(t2t_genes)\n",
    "duplications_t2t_cnvnator_coordinates_genes = duplications_t2t_cnvnator_coordinates_genes.as_df()\n",
    "duplications_t2t_cnvnator_coordinates_genes = duplications_t2t_cnvnator_coordinates_genes[duplications_t2t_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_genes= duplications_t2t_cnvnator_coordinates_genes['CNV_value'].mean()\n",
    "duplications_t2t_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_t2t_cnvnator_coordinates_telomeres = t2t_duplications_cnvnator.nearest(t2t_telomeres)\n",
    "duplications_t2t_cnvnator_coordinates_telomeres = duplications_t2t_cnvnator_coordinates_telomeres.as_df()\n",
    "duplications_t2t_cnvnator_coordinates_telomeres = duplications_t2t_cnvnator_coordinates_telomeres[duplications_t2t_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_t2t_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_t2t_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_telomeres= duplications_t2t_cnvnator_coordinates_telomeres['CNV_value'].mean()\n",
    "duplications_t2t_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_t2t_cnvnator_coordinates_sds = t2t_duplications_cnvnator.nearest(t2t_sds)\n",
    "duplications_t2t_cnvnator_coordinates_sds = duplications_t2t_cnvnator_coordinates_sds.as_df()\n",
    "duplications_t2t_cnvnator_coordinates_sds = duplications_t2t_cnvnator_coordinates_sds[duplications_t2t_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "\n",
    "duplications_t2t_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_sds= duplications_t2t_cnvnator_coordinates_sds['CNV_value'].mean()\n",
    "duplications_t2t_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_sds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7ffae1-94ba-45c0-95d7-e4426e4b28ae",
   "metadata": {},
   "source": [
    "<h1> GRCh37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "21bf4b13-6a3b-4874-bebc-d5d9f6eca33d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Anotations\n",
    "grch37_genes = pd.read_csv('../data/annotation_files/grch37-genes.csv')\n",
    "grch37_telomeres = pd.read_csv('../data/annotation_files/grch37-telomers-centromers.csv')\n",
    "grch37_sds = pd.read_csv('../data/annotation_files/grch37-sds.csv')\n",
    "#grch37_sds = grch37_sds[['#chr1', 'start1', 'end1']]\n",
    "\n",
    "\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "\n",
    "\n",
    "#t2t_sds.columns =['Chromosome','Start','End']\n",
    "#t2t_sds\n",
    "grch37_genes.columns = ['#\"bin\"', 'name', 'Chromosome', 'strand', 'Start', 'End', 'cdsStart',\n",
    "       'cdsEnd', 'exonCount', 'exonStarts', 'exonEnds', 'score', 'name2',\n",
    "       'cdsStartStat', 'cdsEndStat', 'exonFrames'] \n",
    "grch37_telomeres = grch37_telomeres[grch37_telomeres['type'] == 'telomere']\n",
    "grch37_sds.columns = ['#\"bin\"', 'Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'otherChrom', 'otherStart', 'otherEnd', 'otherSize', 'uid',\n",
    "       'posBasesHit', 'testResult', 'verdict', 'chits', 'ccov', 'alignfile',\n",
    "       'alignL', 'indelN', 'indelS', 'alignB', 'matchB', 'mismatchB',\n",
    "       'transitionsB', 'transversionsB', 'fracMatch', 'fracMatchIndel', 'jcK',\n",
    "       'k2K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "92a4acd0-37cc-4be4-a1ce-1ac6fba433d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_55371/3671410005.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_old['RD'] = sample_old['RD'] * 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Type</th>\n",
       "      <th>RD</th>\n",
       "      <th>Lenght</th>\n",
       "      <th>T-TEST-EVAL</th>\n",
       "      <th>GAUSSIAN-EVAL</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51663</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>234001</td>\n",
       "      <td>238300</td>\n",
       "      <td>3.0</td>\n",
       "      <td>duplication</td>\n",
       "      <td>4.170320</td>\n",
       "      <td>4300.0</td>\n",
       "      <td>2.965070e-10</td>\n",
       "      <td>1.407390e+02</td>\n",
       "      <td>4299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51664</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.010921</td>\n",
       "      <td>50700.0</td>\n",
       "      <td>3.143440e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51665</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>50100.0</td>\n",
       "      <td>3.181090e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51666</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.444038</td>\n",
       "      <td>3900.0</td>\n",
       "      <td>4.086470e-11</td>\n",
       "      <td>2.141070e-31</td>\n",
       "      <td>3899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51667</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.013284</td>\n",
       "      <td>3400.0</td>\n",
       "      <td>4.687430e-11</td>\n",
       "      <td>1.314760e-38</td>\n",
       "      <td>3399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52386</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138194201</td>\n",
       "      <td>138197800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.323550</td>\n",
       "      <td>3600.0</td>\n",
       "      <td>1.411230e-02</td>\n",
       "      <td>2.138410e+09</td>\n",
       "      <td>3599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52387</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138214101</td>\n",
       "      <td>138217600</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.019674</td>\n",
       "      <td>3500.0</td>\n",
       "      <td>4.553500e-11</td>\n",
       "      <td>1.185690e-11</td>\n",
       "      <td>3499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52388</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138479101</td>\n",
       "      <td>138480200</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>1.448840e-10</td>\n",
       "      <td>3.214710e-77</td>\n",
       "      <td>1099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52389</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>139663601</td>\n",
       "      <td>139665300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.604050</td>\n",
       "      <td>1700.0</td>\n",
       "      <td>1.392830e-03</td>\n",
       "      <td>2.771230e-02</td>\n",
       "      <td>1699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52390</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>140253401</td>\n",
       "      <td>140256000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.986446</td>\n",
       "      <td>2600.0</td>\n",
       "      <td>6.129710e-11</td>\n",
       "      <td>5.009630e-27</td>\n",
       "      <td>2599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Sample_ID Chromosome      Start        End  CNV_Value  \\\n",
       "51663  LP6005441-DNA_G07          1     234001     238300        3.0   \n",
       "51664  LP6005441-DNA_G07          1     267101     317800        1.0   \n",
       "51665  LP6005441-DNA_G07          1     471301     521400        1.0   \n",
       "51666  LP6005441-DNA_G07          1    1010801    1014700        1.0   \n",
       "51667  LP6005441-DNA_G07          1    4124401    4127800        1.0   \n",
       "...                  ...        ...        ...        ...        ...   \n",
       "52386  LP6005441-DNA_G07          9  138194201  138197800        1.0   \n",
       "52387  LP6005441-DNA_G07          9  138214101  138217600        1.0   \n",
       "52388  LP6005441-DNA_G07          9  138479101  138480200        1.0   \n",
       "52389  LP6005441-DNA_G07          9  139663601  139665300        1.0   \n",
       "52390  LP6005441-DNA_G07          9  140253401  140256000        1.0   \n",
       "\n",
       "              Type        RD   Lenght   T-TEST-EVAL  GAUSSIAN-EVAL  \\\n",
       "51663  duplication  4.170320   4300.0  2.965070e-10   1.407390e+02   \n",
       "51664     deletion  0.010921  50700.0  3.143440e-12   0.000000e+00   \n",
       "51665     deletion  0.000109  50100.0  3.181090e-12   0.000000e+00   \n",
       "51666     deletion  0.444038   3900.0  4.086470e-11   2.141070e-31   \n",
       "51667     deletion  1.013284   3400.0  4.687430e-11   1.314760e-38   \n",
       "...            ...       ...      ...           ...            ...   \n",
       "52386     deletion  1.323550   3600.0  1.411230e-02   2.138410e+09   \n",
       "52387     deletion  1.019674   3500.0  4.553500e-11   1.185690e-11   \n",
       "52388     deletion  0.000000   1100.0  1.448840e-10   3.214710e-77   \n",
       "52389     deletion  0.604050   1700.0  1.392830e-03   2.771230e-02   \n",
       "52390     deletion  0.986446   2600.0  6.129710e-11   5.009630e-27   \n",
       "\n",
       "       Length_cnvr  \n",
       "51663         4299  \n",
       "51664        50699  \n",
       "51665        50099  \n",
       "51666         3899  \n",
       "51667         3399  \n",
       "...            ...  \n",
       "52386         3599  \n",
       "52387         3499  \n",
       "52388         1099  \n",
       "52389         1699  \n",
       "52390         2599  \n",
       "\n",
       "[728 rows x 11 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"../data/call_cnvr_grch37/grch37_cnvs.txt\", sep='\\t')\n",
    "data = pd.read_csv(\"../data_backup/cnvnator_cnvs.csv\")\n",
    "data.columns = ['Sample_ID','Chromosome', 'Start', 'End','CNV_Value','Type','RD','Lenght', 'T-TEST-EVAL','GAUSSIAN-EVAL',]\n",
    "data = data.dropna()#.describe()\n",
    "data['Length_cnvr'] = data['End'] - data['Start'] \n",
    "sample_old = data[data['Sample_ID'] == 'LP6005441-DNA_G07']\n",
    "#sample_old = old_assembly[old_assembly['Sample_ID'] == 'LP6005441-DNA_B10']\n",
    "#sample_old = sample_old[['Chromosome', 'Start', 'End', 'RD']]\n",
    "sample_old['RD'] = sample_old['RD'] * 2\n",
    "#sample_old['Chromosome'] = str('chr')+sample_old['Chromosome']\n",
    "    #sample_old.to_csv('LP6005441-DNA_B10_grch37.bed', index=False, header=False, sep='\\t')\n",
    "    #list(sample_old.columns)\n",
    "\n",
    "#sample_old\n",
    "\n",
    "sample_old = sample_old[~sample_old['RD'].isin([2])]\n",
    "\n",
    "sample_old = sample_old[sample_old['Chromosome'] != 'M']\n",
    "sample_old = sample_old[sample_old['Chromosome'] != 'X']\n",
    "sample_old = sample_old[sample_old['Chromosome'] != 'Y']\n",
    "\n",
    "sample_old "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f87a0006-3796-4c37-969d-d3dedb628afc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Type</th>\n",
       "      <th>RD</th>\n",
       "      <th>Lenght</th>\n",
       "      <th>T-TEST-EVAL</th>\n",
       "      <th>GAUSSIAN-EVAL</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>234001</td>\n",
       "      <td>238300</td>\n",
       "      <td>3.0</td>\n",
       "      <td>duplication</td>\n",
       "      <td>4.170320</td>\n",
       "      <td>4300.0</td>\n",
       "      <td>2.965070e-10</td>\n",
       "      <td>1.407390e+02</td>\n",
       "      <td>4299</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.010921</td>\n",
       "      <td>50700.0</td>\n",
       "      <td>3.143440e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50699</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>50100.0</td>\n",
       "      <td>3.181090e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50099</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.444038</td>\n",
       "      <td>3900.0</td>\n",
       "      <td>4.086470e-11</td>\n",
       "      <td>2.141070e-31</td>\n",
       "      <td>3899</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.013284</td>\n",
       "      <td>3400.0</td>\n",
       "      <td>4.687430e-11</td>\n",
       "      <td>1.314760e-38</td>\n",
       "      <td>3399</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138194201</td>\n",
       "      <td>138197800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.323550</td>\n",
       "      <td>3600.0</td>\n",
       "      <td>1.411230e-02</td>\n",
       "      <td>2.138410e+09</td>\n",
       "      <td>3599</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138214101</td>\n",
       "      <td>138217600</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.019674</td>\n",
       "      <td>3500.0</td>\n",
       "      <td>4.553500e-11</td>\n",
       "      <td>1.185690e-11</td>\n",
       "      <td>3499</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138479101</td>\n",
       "      <td>138480200</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>1.448840e-10</td>\n",
       "      <td>3.214710e-77</td>\n",
       "      <td>1099</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>139663601</td>\n",
       "      <td>139665300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.604050</td>\n",
       "      <td>1700.0</td>\n",
       "      <td>1.392830e-03</td>\n",
       "      <td>2.771230e-02</td>\n",
       "      <td>1699</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>140253401</td>\n",
       "      <td>140256000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.986446</td>\n",
       "      <td>2600.0</td>\n",
       "      <td>6.129710e-11</td>\n",
       "      <td>5.009630e-27</td>\n",
       "      <td>2599</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample_ID Chromosome      Start        End  CNV_Value  \\\n",
       "0    LP6005441-DNA_G07          1     234001     238300        3.0   \n",
       "1    LP6005441-DNA_G07          1     267101     317800        1.0   \n",
       "2    LP6005441-DNA_G07          1     471301     521400        1.0   \n",
       "3    LP6005441-DNA_G07          1    1010801    1014700        1.0   \n",
       "4    LP6005441-DNA_G07          1    4124401    4127800        1.0   \n",
       "..                 ...        ...        ...        ...        ...   \n",
       "723  LP6005441-DNA_G07          9  138194201  138197800        1.0   \n",
       "724  LP6005441-DNA_G07          9  138214101  138217600        1.0   \n",
       "725  LP6005441-DNA_G07          9  138479101  138480200        1.0   \n",
       "726  LP6005441-DNA_G07          9  139663601  139665300        1.0   \n",
       "727  LP6005441-DNA_G07          9  140253401  140256000        1.0   \n",
       "\n",
       "            Type        RD   Lenght   T-TEST-EVAL  GAUSSIAN-EVAL  Length_cnvr  \\\n",
       "0    duplication  4.170320   4300.0  2.965070e-10   1.407390e+02         4299   \n",
       "1       deletion  0.010921  50700.0  3.143440e-12   0.000000e+00        50699   \n",
       "2       deletion  0.000109  50100.0  3.181090e-12   0.000000e+00        50099   \n",
       "3       deletion  0.444038   3900.0  4.086470e-11   2.141070e-31         3899   \n",
       "4       deletion  1.013284   3400.0  4.687430e-11   1.314760e-38         3399   \n",
       "..           ...       ...      ...           ...            ...          ...   \n",
       "723     deletion  1.323550   3600.0  1.411230e-02   2.138410e+09         3599   \n",
       "724     deletion  1.019674   3500.0  4.553500e-11   1.185690e-11         3499   \n",
       "725     deletion  0.000000   1100.0  1.448840e-10   3.214710e-77         1099   \n",
       "726     deletion  0.604050   1700.0  1.392830e-03   2.771230e-02         1699   \n",
       "727     deletion  0.986446   2600.0  6.129710e-11   5.009630e-27         2599   \n",
       "\n",
       "                Sample   Region  \n",
       "0    LP6005441-DNA_G07  America  \n",
       "1    LP6005441-DNA_G07  America  \n",
       "2    LP6005441-DNA_G07  America  \n",
       "3    LP6005441-DNA_G07  America  \n",
       "4    LP6005441-DNA_G07  America  \n",
       "..                 ...      ...  \n",
       "723  LP6005441-DNA_G07  America  \n",
       "724  LP6005441-DNA_G07  America  \n",
       "725  LP6005441-DNA_G07  America  \n",
       "726  LP6005441-DNA_G07  America  \n",
       "727  LP6005441-DNA_G07  America  \n",
       "\n",
       "[728 rows x 13 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###### Keeping only samples that are on old cnvs \n",
    "sample_old = sample_old.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])\n",
    "\n",
    "#REMOVE SAMPLES WITH ABERRATIONS\n",
    "#sample_old.columns = ['Chromosome', 'Start', 'End', 'CNVR_ID', 'Sample_iD', 'CNV_start', 'CNV_end', 'CNV_Value','Length', 'Length_cnvr','Sample','Region']\n",
    "sample_old "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f5ab003b-6d04-431b-ac90-0be08e8f2049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>234001</td>\n",
       "      <td>238300</td>\n",
       "      <td>4299</td>\n",
       "      <td>4.170320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>50699</td>\n",
       "      <td>0.010921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>50099</td>\n",
       "      <td>0.000109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>3899</td>\n",
       "      <td>0.444038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>3399</td>\n",
       "      <td>1.013284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>9</td>\n",
       "      <td>138194201</td>\n",
       "      <td>138197800</td>\n",
       "      <td>3599</td>\n",
       "      <td>1.323550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>9</td>\n",
       "      <td>138214101</td>\n",
       "      <td>138217600</td>\n",
       "      <td>3499</td>\n",
       "      <td>1.019674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>9</td>\n",
       "      <td>138479101</td>\n",
       "      <td>138480200</td>\n",
       "      <td>1099</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>9</td>\n",
       "      <td>139663601</td>\n",
       "      <td>139665300</td>\n",
       "      <td>1699</td>\n",
       "      <td>0.604050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>9</td>\n",
       "      <td>140253401</td>\n",
       "      <td>140256000</td>\n",
       "      <td>2599</td>\n",
       "      <td>0.986446</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome      Start        End  Length_cnvr  LP6005441-DNA_G07\n",
       "0               1     234001     238300         4299           4.170320\n",
       "1               1     267101     317800        50699           0.010921\n",
       "2               1     471301     521400        50099           0.000109\n",
       "3               1    1010801    1014700         3899           0.444038\n",
       "4               1    4124401    4127800         3399           1.013284\n",
       "..            ...        ...        ...          ...                ...\n",
       "723             9  138194201  138197800         3599           1.323550\n",
       "724             9  138214101  138217600         3499           1.019674\n",
       "725             9  138479101  138480200         1099           0.000000\n",
       "726             9  139663601  139665300         1699           0.604050\n",
       "727             9  140253401  140256000         2599           0.986446\n",
       "\n",
       "[728 rows x 5 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv_grch37 = sample_old.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='RD',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "\n",
    "final_cnv_grch37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3200a16b-2464-4e47-8063-9e46662ba8f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>50699</td>\n",
       "      <td>0.010921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>50099</td>\n",
       "      <td>0.000109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>3899</td>\n",
       "      <td>0.444038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>3399</td>\n",
       "      <td>1.013284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>8360201</td>\n",
       "      <td>8364000</td>\n",
       "      <td>3799</td>\n",
       "      <td>0.017621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>22</td>\n",
       "      <td>21880701</td>\n",
       "      <td>21887200</td>\n",
       "      <td>6499</td>\n",
       "      <td>0.925102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>22</td>\n",
       "      <td>21899601</td>\n",
       "      <td>21905500</td>\n",
       "      <td>5899</td>\n",
       "      <td>1.209440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>22</td>\n",
       "      <td>23478501</td>\n",
       "      <td>23479700</td>\n",
       "      <td>1199</td>\n",
       "      <td>0.186410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>22</td>\n",
       "      <td>37143101</td>\n",
       "      <td>37147900</td>\n",
       "      <td>4799</td>\n",
       "      <td>0.057464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>22</td>\n",
       "      <td>39294101</td>\n",
       "      <td>39298700</td>\n",
       "      <td>4599</td>\n",
       "      <td>0.003570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+---------------+---------------------+\n",
       "| Chromosome   | Start     | End       | Length_cnvr   | LP6005441-DNA_G07   |\n",
       "| (category)   | (int64)   | (int64)   | (int64)       | (float64)           |\n",
       "|--------------+-----------+-----------+---------------+---------------------|\n",
       "| 1            | 267101    | 317800    | 50699         | 0.01092072          |\n",
       "| 1            | 471301    | 521400    | 50099         | 0.0001087124        |\n",
       "| 1            | 1010801   | 1014700   | 3899          | 0.444038            |\n",
       "| 1            | 4124401   | 4127800   | 3399          | 1.013284            |\n",
       "| ...          | ...       | ...       | ...           | ...                 |\n",
       "| 22           | 21899601  | 21905500  | 5899          | 1.20944             |\n",
       "| 22           | 23478501  | 23479700  | 1199          | 0.1864096           |\n",
       "| 22           | 37143101  | 37147900  | 4799          | 0.057464            |\n",
       "| 22           | 39294101  | 39298700  | 4599          | 0.00356992          |\n",
       "+--------------+-----------+-----------+---------------+---------------------+\n",
       "Unstranded PyRanges object has 508 rows and 5 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_deletions_cnvnator = final_cnv_grch37[final_cnv_grch37['LP6005441-DNA_G07'] < 2]\n",
    "grch37_deletions_cnvnator_mean = grch37_deletions_cnvnator['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "grch37_duplications_cnvnator = final_cnv_grch37[final_cnv_grch37['LP6005441-DNA_G07'] > 2]\n",
    "grch37_duplications_cnvnator_mean = grch37_duplications_cnvnator['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "grch37_duplications_cnvnator = PyRanges(grch37_duplications_cnvnator)\n",
    "grch37_deletions_cnvnator = PyRanges(grch37_deletions_cnvnator)\n",
    "\n",
    "\n",
    "final_cnv_ranges_grch37 = PyRanges(final_cnv_grch37)\n",
    "grch37_deletions_cnvnator"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f006e6db-c6d3-4dea-9126-9f0ce4e30ed2",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8e734c44-9ebc-4144-95a1-58803104e4fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "grch37_genes['Chromosome'] = grch37_genes['Chromosome'].str.replace('chr', '')\n",
    "grch37_genes_coordinates = PyRanges(grch37_genes)\n",
    "\n",
    "grch37_telomeres['Chromosome'] = grch37_telomeres['Chromosome'].str.replace('chr', '')\n",
    "grch37_telomeres = PyRanges(grch37_telomeres)\n",
    "\n",
    "grch37_sds['Chromosome'] = grch37_sds['Chromosome'].str.replace('chr', '')\n",
    "grch37_sds = PyRanges(grch37_sds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b3c811e9-1372-4e7f-9742-02d296c392b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genes\n",
    "grch37_coordinates_genes = final_cnv_ranges_grch37.nearest(grch37_genes_coordinates)\n",
    "grch37_coordinates_genes = grch37_coordinates_genes.as_df()\n",
    "grch37_coordinates_genes = grch37_coordinates_genes[grch37_coordinates_genes['Distance'] <1 ]\n",
    "#grch37_coordinates_genes\n",
    "\n",
    "#Telomeres\n",
    "grch37_coordinates_telomeres = final_cnv_ranges_grch37.nearest(grch37_telomeres)\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres.as_df()\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres[grch37_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs\n",
    "grch37_coordinates_sds = final_cnv_ranges_grch37.nearest(grch37_sds)\n",
    "grch37_coordinates_sds = grch37_coordinates_sds.as_df()\n",
    "grch37_coordinates_sds = grch37_coordinates_sds[grch37_coordinates_sds['Distance'] <1 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d95bff56-ba99-456e-8d98-67bbd853687f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#grch37_genes = len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "#grch37_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eb30162-a0ee-47c5-9dce-657e51b770f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3324939e-9862-4985-b021-884aaa8c9347",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Deletions\n",
    "#### Genes \n",
    "deletions_grch37_cnvnator_coordinates_genes = grch37_deletions_cnvnator.nearest(grch37_genes_coordinates)\n",
    "deletions_grch37_cnvnator_coordinates_genes = deletions_grch37_cnvnator_coordinates_genes.as_df()\n",
    "deletions_grch37_cnvnator_coordinates_genes = deletions_grch37_cnvnator_coordinates_genes[deletions_grch37_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_cnvnator_mean_individual_genes= deletions_grch37_cnvnator_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "deletions_grch37_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_grch37_cnvnator_coordinates_telomeres = grch37_deletions_cnvnator.nearest(grch37_telomeres)\n",
    "deletions_grch37_cnvnator_coordinates_telomeres = deletions_grch37_cnvnator_coordinates_telomeres.as_df()\n",
    "deletions_grch37_cnvnator_coordinates_telomeres = deletions_grch37_cnvnator_coordinates_telomeres[deletions_grch37_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_grch37_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_grch37_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_cnvnator_mean_individual_telomeres= deletions_grch37_cnvnator_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "deletions_grch37_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_grch37_cnvnator_coordinates_sds = grch37_deletions_cnvnator.nearest(grch37_sds)\n",
    "deletions_grch37_cnvnator_coordinates_sds = deletions_grch37_cnvnator_coordinates_sds.as_df()\n",
    "deletions_grch37_cnvnator_coordinates_sds = deletions_grch37_cnvnator_coordinates_sds[deletions_grch37_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_cnvnator_mean_individual_sds= deletions_grch37_cnvnator_coordinates_sds['LP6005441-DNA_G07'].mean()\n",
    "deletions_grch37_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b1318180-3c75-4c62-a2d3-2cfc72cc820f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Duplications\n",
    "#### Genes \n",
    "duplications_grch37_cnvnator_coordinates_genes = grch37_duplications_cnvnator.nearest(grch37_genes_coordinates)\n",
    "duplications_grch37_cnvnator_coordinates_genes = duplications_grch37_cnvnator_coordinates_genes.as_df()\n",
    "duplications_grch37_cnvnator_coordinates_genes = duplications_grch37_cnvnator_coordinates_genes[duplications_grch37_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_cnvnator_mean_individual_genes= duplications_grch37_cnvnator_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "duplications_grch37_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_grch37_cnvnator_coordinates_telomeres = grch37_duplications_cnvnator.nearest(grch37_telomeres)\n",
    "duplications_grch37_cnvnator_coordinates_telomeres = duplications_grch37_cnvnator_coordinates_telomeres.as_df()\n",
    "duplications_grch37_cnvnator_coordinates_telomeres = duplications_grch37_cnvnator_coordinates_telomeres[duplications_grch37_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_grch37_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_grch37_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_cnvnator_mean_individual_telomeres= duplications_grch37_cnvnator_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "duplications_grch37_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_grch37_cnvnator_coordinates_sds = grch37_duplications_cnvnator.nearest(grch37_sds)\n",
    "duplications_grch37_cnvnator_coordinates_sds = duplications_grch37_cnvnator_coordinates_sds.as_df()\n",
    "duplications_grch37_cnvnator_coordinates_sds = duplications_grch37_cnvnator_coordinates_sds[duplications_grch37_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_cnvnator_mean_individual_sds= duplications_grch37_cnvnator_coordinates_sds['LP6005441-DNA_G07'].mean()\n",
    "duplications_grch37_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02a2ca0-2689-4d75-a250-04f12f78179b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "258edfda-e062-4a97-b450-3c3ff2798083",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "55c3b568-3b07-4850-8ec0-a81222f3d5d3",
   "metadata": {},
   "source": [
    "<h2> Peter Sudmant data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cf3b31d8-07cb-4c2b-bc22-66e26786d3f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>177033</td>\n",
       "      <td>22</td>\n",
       "      <td>164985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>89237</td>\n",
       "      <td>10</td>\n",
       "      <td>77189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>228416</td>\n",
       "      <td>267160</td>\n",
       "      <td>24</td>\n",
       "      <td>38744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>471226</td>\n",
       "      <td>471326</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>318047</td>\n",
       "      <td>471226</td>\n",
       "      <td>21</td>\n",
       "      <td>153179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4311</th>\n",
       "      <td>22</td>\n",
       "      <td>42973857</td>\n",
       "      <td>42980083</td>\n",
       "      <td>4</td>\n",
       "      <td>6226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4312</th>\n",
       "      <td>22</td>\n",
       "      <td>42973707</td>\n",
       "      <td>42973857</td>\n",
       "      <td>4</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4313</th>\n",
       "      <td>22</td>\n",
       "      <td>49013936</td>\n",
       "      <td>49016261</td>\n",
       "      <td>5</td>\n",
       "      <td>2325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4314</th>\n",
       "      <td>22</td>\n",
       "      <td>51223959</td>\n",
       "      <td>51224959</td>\n",
       "      <td>4</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4315</th>\n",
       "      <td>22</td>\n",
       "      <td>51201920</td>\n",
       "      <td>51223959</td>\n",
       "      <td>4</td>\n",
       "      <td>22039</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4316 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "| Chromosome   | Start     | End       | CNV_value   | Length_cnvr   |\n",
       "| (category)   | (int64)   | (int64)   | (int64)     | (int64)       |\n",
       "|--------------+-----------+-----------+-------------+---------------|\n",
       "| 1            | 12048     | 177033    | 22          | 164985        |\n",
       "| 1            | 12048     | 89237     | 10          | 77189         |\n",
       "| 1            | 228416    | 267160    | 24          | 38744         |\n",
       "| 1            | 471226    | 471326    | 10          | 100           |\n",
       "| ...          | ...       | ...       | ...         | ...           |\n",
       "| 22           | 42973707  | 42973857  | 4           | 150           |\n",
       "| 22           | 49013936  | 49016261  | 5           | 2325          |\n",
       "| 22           | 51223959  | 51224959  | 4           | 1000          |\n",
       "| 22           | 51201920  | 51223959  | 4           | 22039         |\n",
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "Unstranded PyRanges object has 4,316 rows and 5 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sudmant = pd.read_csv('/Users/luciabazan/Downloads/sudmant_data.csv')\n",
    "sudmant = sudmant[['contig','start','end','AMR_Maya_HGDP00855_F']]\n",
    "sudmant.columns =['Chromosome', 'Start', 'End', 'CNV_value']\n",
    "\n",
    "\n",
    "#sudmant['Chromosome'] = sudmant['Chromosome'].str.replace('chr','hs')\n",
    "n_uncalled = np.sum(sudmant==-1,1)\n",
    "sudmant = sudmant[n_uncalled==0]\n",
    "\n",
    "sudmant['Length_cnvr'] = sudmant['End'] - sudmant['Start'] \n",
    "sudmant = sudmant[~sudmant['CNV_value'].isin([2])]\n",
    "\n",
    "\n",
    "sudmant['Chromosome'] = sudmant['Chromosome'].str.replace('chr', '')\n",
    "sudmant = sudmant[sudmant['Chromosome'] != 'M']\n",
    "sudmant = sudmant[sudmant['Chromosome'] != 'X']\n",
    "sudmant = sudmant[sudmant['Chromosome'] != 'Y']\n",
    "\n",
    "\n",
    "grch37_deletions_sudmant = sudmant[sudmant['CNV_value'] < 3]\n",
    "grch37_deletions_sudmant_mean = grch37_deletions_sudmant['CNV_value'].mean()\n",
    "\n",
    "grch37_duplications_sudmant = sudmant[sudmant['CNV_value'] > 1]\n",
    "grch37_duplications_sudmant_mean = grch37_duplications_sudmant['CNV_value'].mean()\n",
    "\n",
    "grch37_duplications_sudmant = PyRanges(grch37_duplications_sudmant)\n",
    "grch37_deletions_sudmant = PyRanges(grch37_deletions_sudmant)\n",
    "\n",
    "\n",
    "sudmant = PyRanges(sudmant)\n",
    "sudmant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a64eefc5-41c8-45f7-b64b-d46d3aa43daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genes\n",
    "peter_coordinates_genes = sudmant.nearest(grch37_genes_coordinates)\n",
    "peter_coordinates_genes = peter_coordinates_genes.as_df()\n",
    "peter_coordinates_genes = peter_coordinates_genes[peter_coordinates_genes['Distance'] <1 ]\n",
    "#grch37_coordinates_genes\n",
    "\n",
    "#Telomeres\n",
    "peter_coordinates_telomeres = sudmant.nearest(grch37_telomeres)\n",
    "peter_coordinates_telomeres = peter_coordinates_telomeres.as_df()\n",
    "peter_coordinates_telomeres = peter_coordinates_telomeres[peter_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs\n",
    "peter_coordinates_sds = sudmant.nearest(grch37_sds)\n",
    "peter_coordinates_sds = peter_coordinates_sds.as_df()\n",
    "peter_coordinates_sds = peter_coordinates_sds[peter_coordinates_sds['Distance'] <1 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "382fef50-c3b3-4737-8d72-1dc5078ba5c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>#\"bin\"</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>...</th>\n",
       "      <th>alignB</th>\n",
       "      <th>matchB</th>\n",
       "      <th>mismatchB</th>\n",
       "      <th>transitionsB</th>\n",
       "      <th>transversionsB</th>\n",
       "      <th>fracMatch</th>\n",
       "      <th>fracMatchIndel</th>\n",
       "      <th>jcK</th>\n",
       "      <th>k2K</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>177033</td>\n",
       "      <td>22</td>\n",
       "      <td>164985</td>\n",
       "      <td>585</td>\n",
       "      <td>10000</td>\n",
       "      <td>87112</td>\n",
       "      <td>chr15:102446355</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>74269</td>\n",
       "      <td>73742</td>\n",
       "      <td>527</td>\n",
       "      <td>332</td>\n",
       "      <td>195</td>\n",
       "      <td>0.992904</td>\n",
       "      <td>0.991956</td>\n",
       "      <td>0.007130</td>\n",
       "      <td>0.007133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>89237</td>\n",
       "      <td>10</td>\n",
       "      <td>77189</td>\n",
       "      <td>585</td>\n",
       "      <td>10000</td>\n",
       "      <td>87112</td>\n",
       "      <td>chr15:102446355</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>74269</td>\n",
       "      <td>73742</td>\n",
       "      <td>527</td>\n",
       "      <td>332</td>\n",
       "      <td>195</td>\n",
       "      <td>0.992904</td>\n",
       "      <td>0.991956</td>\n",
       "      <td>0.007130</td>\n",
       "      <td>0.007133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>228416</td>\n",
       "      <td>267160</td>\n",
       "      <td>24</td>\n",
       "      <td>38744</td>\n",
       "      <td>586</td>\n",
       "      <td>227417</td>\n",
       "      <td>238719</td>\n",
       "      <td>chr17:81176011</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>11249</td>\n",
       "      <td>11065</td>\n",
       "      <td>184</td>\n",
       "      <td>123</td>\n",
       "      <td>61</td>\n",
       "      <td>0.983643</td>\n",
       "      <td>0.981984</td>\n",
       "      <td>0.016538</td>\n",
       "      <td>0.016561</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>471226</td>\n",
       "      <td>471326</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "      <td>73</td>\n",
       "      <td>317719</td>\n",
       "      <td>471368</td>\n",
       "      <td>chr5:180744334</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>153555</td>\n",
       "      <td>153444</td>\n",
       "      <td>111</td>\n",
       "      <td>75</td>\n",
       "      <td>36</td>\n",
       "      <td>0.999277</td>\n",
       "      <td>0.999121</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>318047</td>\n",
       "      <td>471226</td>\n",
       "      <td>21</td>\n",
       "      <td>153179</td>\n",
       "      <td>73</td>\n",
       "      <td>317719</td>\n",
       "      <td>471368</td>\n",
       "      <td>chr5:180744334</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>153555</td>\n",
       "      <td>153444</td>\n",
       "      <td>111</td>\n",
       "      <td>75</td>\n",
       "      <td>36</td>\n",
       "      <td>0.999277</td>\n",
       "      <td>0.999121</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4290</th>\n",
       "      <td>22</td>\n",
       "      <td>42973857</td>\n",
       "      <td>42980083</td>\n",
       "      <td>4</td>\n",
       "      <td>6226</td>\n",
       "      <td>912</td>\n",
       "      <td>42962426</td>\n",
       "      <td>42978456</td>\n",
       "      <td>chr22:42902112</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>13422</td>\n",
       "      <td>13031</td>\n",
       "      <td>391</td>\n",
       "      <td>245</td>\n",
       "      <td>146</td>\n",
       "      <td>0.970869</td>\n",
       "      <td>0.967122</td>\n",
       "      <td>0.029712</td>\n",
       "      <td>0.029771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4291</th>\n",
       "      <td>22</td>\n",
       "      <td>42973707</td>\n",
       "      <td>42973857</td>\n",
       "      <td>4</td>\n",
       "      <td>150</td>\n",
       "      <td>912</td>\n",
       "      <td>42962426</td>\n",
       "      <td>42978456</td>\n",
       "      <td>chr22:42902112</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>13422</td>\n",
       "      <td>13031</td>\n",
       "      <td>391</td>\n",
       "      <td>245</td>\n",
       "      <td>146</td>\n",
       "      <td>0.970869</td>\n",
       "      <td>0.967122</td>\n",
       "      <td>0.029712</td>\n",
       "      <td>0.029771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4292</th>\n",
       "      <td>22</td>\n",
       "      <td>49013936</td>\n",
       "      <td>49016261</td>\n",
       "      <td>5</td>\n",
       "      <td>2325</td>\n",
       "      <td>958</td>\n",
       "      <td>49014584</td>\n",
       "      <td>49015948</td>\n",
       "      <td>chr15:22514585</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1331</td>\n",
       "      <td>1270</td>\n",
       "      <td>61</td>\n",
       "      <td>44</td>\n",
       "      <td>17</td>\n",
       "      <td>0.954170</td>\n",
       "      <td>0.949888</td>\n",
       "      <td>0.047290</td>\n",
       "      <td>0.047556</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4293</th>\n",
       "      <td>22</td>\n",
       "      <td>51223959</td>\n",
       "      <td>51224959</td>\n",
       "      <td>4</td>\n",
       "      <td>1000</td>\n",
       "      <td>975</td>\n",
       "      <td>51198214</td>\n",
       "      <td>51236901</td>\n",
       "      <td>chr2:114370005</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>38593</td>\n",
       "      <td>38225</td>\n",
       "      <td>368</td>\n",
       "      <td>242</td>\n",
       "      <td>126</td>\n",
       "      <td>0.990465</td>\n",
       "      <td>0.989260</td>\n",
       "      <td>0.009597</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4294</th>\n",
       "      <td>22</td>\n",
       "      <td>51201920</td>\n",
       "      <td>51223959</td>\n",
       "      <td>4</td>\n",
       "      <td>22039</td>\n",
       "      <td>975</td>\n",
       "      <td>51198214</td>\n",
       "      <td>51236901</td>\n",
       "      <td>chr2:114370005</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>38593</td>\n",
       "      <td>38225</td>\n",
       "      <td>368</td>\n",
       "      <td>242</td>\n",
       "      <td>126</td>\n",
       "      <td>0.990465</td>\n",
       "      <td>0.989260</td>\n",
       "      <td>0.009597</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3704 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome     Start       End  CNV_value  Length_cnvr  #\"bin\"   Start_b  \\\n",
       "0             1     12048    177033         22       164985     585     10000   \n",
       "1             1     12048     89237         10        77189     585     10000   \n",
       "2             1    228416    267160         24        38744     586    227417   \n",
       "3             1    471226    471326         10          100      73    317719   \n",
       "4             1    318047    471226         21       153179      73    317719   \n",
       "...         ...       ...       ...        ...          ...     ...       ...   \n",
       "4290         22  42973857  42980083          4         6226     912  42962426   \n",
       "4291         22  42973707  42973857          4          150     912  42962426   \n",
       "4292         22  49013936  49016261          5         2325     958  49014584   \n",
       "4293         22  51223959  51224959          4         1000     975  51198214   \n",
       "4294         22  51201920  51223959          4        22039     975  51198214   \n",
       "\n",
       "         End_b             name  score  ...  alignB  matchB  mismatchB  \\\n",
       "0        87112  chr15:102446355      0  ...   74269   73742        527   \n",
       "1        87112  chr15:102446355      0  ...   74269   73742        527   \n",
       "2       238719   chr17:81176011      0  ...   11249   11065        184   \n",
       "3       471368   chr5:180744334      0  ...  153555  153444        111   \n",
       "4       471368   chr5:180744334      0  ...  153555  153444        111   \n",
       "...        ...              ...    ...  ...     ...     ...        ...   \n",
       "4290  42978456   chr22:42902112      0  ...   13422   13031        391   \n",
       "4291  42978456   chr22:42902112      0  ...   13422   13031        391   \n",
       "4292  49015948   chr15:22514585      0  ...    1331    1270         61   \n",
       "4293  51236901   chr2:114370005      0  ...   38593   38225        368   \n",
       "4294  51236901   chr2:114370005      0  ...   38593   38225        368   \n",
       "\n",
       "      transitionsB  transversionsB  fracMatch  fracMatchIndel       jcK  \\\n",
       "0              332             195   0.992904        0.991956  0.007130   \n",
       "1              332             195   0.992904        0.991956  0.007130   \n",
       "2              123              61   0.983643        0.981984  0.016538   \n",
       "3               75              36   0.999277        0.999121  0.000723   \n",
       "4               75              36   0.999277        0.999121  0.000723   \n",
       "...            ...             ...        ...             ...       ...   \n",
       "4290           245             146   0.970869        0.967122  0.029712   \n",
       "4291           245             146   0.970869        0.967122  0.029712   \n",
       "4292            44              17   0.954170        0.949888  0.047290   \n",
       "4293           242             126   0.990465        0.989260  0.009597   \n",
       "4294           242             126   0.990465        0.989260  0.009597   \n",
       "\n",
       "           k2K  Distance  \n",
       "0     0.007133         0  \n",
       "1     0.007133         0  \n",
       "2     0.016561         0  \n",
       "3     0.000723         0  \n",
       "4     0.000723         0  \n",
       "...        ...       ...  \n",
       "4290  0.029771         0  \n",
       "4291  0.029771         0  \n",
       "4292  0.047556         0  \n",
       "4293  0.009604         0  \n",
       "4294  0.009604         0  \n",
       "\n",
       "[3704 rows x 35 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peter_coordinates_sds#[['Chromosome','Start', 'End']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d720a419-476a-4e26-8d5c-9225f2ce38d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3e2e6080-b29f-4c35-b894-00ac5173dcef",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Deletions\n",
    "#### Genes \n",
    "deletions_grch37_sudmant_coordinates_genes = grch37_deletions_sudmant.nearest(grch37_genes_coordinates)\n",
    "deletions_grch37_sudmant_coordinates_genes = deletions_grch37_sudmant_coordinates_genes.as_df()\n",
    "deletions_grch37_sudmant_coordinates_genes = deletions_grch37_sudmant_coordinates_genes[deletions_grch37_sudmant_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_sudmant_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_sudmant_mean_individual_genes= deletions_grch37_sudmant_coordinates_genes['CNV_value'].mean()\n",
    "deletions_grch37_sudmant_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_grch37_sudmant_coordinates_telomeres = grch37_deletions_sudmant.nearest(grch37_telomeres)\n",
    "deletions_grch37_sudmant_coordinates_telomeres = deletions_grch37_sudmant_coordinates_telomeres.as_df()\n",
    "deletions_grch37_sudmant_coordinates_telomeres = deletions_grch37_sudmant_coordinates_telomeres[deletions_grch37_sudmant_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_grch37_sudmant_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_grch37_sudmant_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_sudmant_mean_individual_telomeres= deletions_grch37_sudmant_coordinates_telomeres['CNV_value'].mean()\n",
    "deletions_grch37_sudmant_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_grch37_sudmant_coordinates_sds = grch37_deletions_sudmant.nearest(grch37_sds)\n",
    "deletions_grch37_sudmant_coordinates_sds = deletions_grch37_sudmant_coordinates_sds.as_df()\n",
    "deletions_grch37_sudmant_coordinates_sds = deletions_grch37_sudmant_coordinates_sds[deletions_grch37_sudmant_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_sudmant_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_sudmant_mean_individual_sds= deletions_grch37_sudmant_coordinates_sds['CNV_value'].mean()\n",
    "deletions_grch37_sudmant_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1f5d27a5-ae43-417b-960e-42e27ce8f2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Duplications\n",
    "#### Genes \n",
    "duplications_grch37_sudmant_coordinates_genes = grch37_duplications_sudmant.nearest(grch37_genes_coordinates)\n",
    "duplications_grch37_sudmant_coordinates_genes = duplications_grch37_sudmant_coordinates_genes.as_df()\n",
    "duplications_grch37_sudmant_coordinates_genes = duplications_grch37_sudmant_coordinates_genes[duplications_grch37_sudmant_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_sudmant_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_sudmant_mean_individual_genes= duplications_grch37_sudmant_coordinates_genes['CNV_value'].mean()\n",
    "duplications_grch37_sudmant_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_grch37_sudmant_coordinates_telomeres = grch37_duplications_sudmant.nearest(grch37_telomeres)\n",
    "duplications_grch37_sudmant_coordinates_telomeres = duplications_grch37_sudmant_coordinates_telomeres.as_df()\n",
    "duplications_grch37_sudmant_coordinates_telomeres = duplications_grch37_sudmant_coordinates_telomeres[duplications_grch37_sudmant_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_grch37_sudmant_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_grch37_sudmant_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_sudmant_mean_individual_telomeres= duplications_grch37_sudmant_coordinates_telomeres['CNV_value'].mean()\n",
    "duplications_grch37_sudmant_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_grch37_sudmant_coordinates_sds = grch37_duplications_sudmant.nearest(grch37_sds)\n",
    "duplications_grch37_sudmant_coordinates_sds = duplications_grch37_sudmant_coordinates_sds.as_df()\n",
    "duplications_grch37_sudmant_coordinates_sds = duplications_grch37_sudmant_coordinates_sds[duplications_grch37_sudmant_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_sudmant_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_sudmant_mean_individual_sds= duplications_grch37_sudmant_coordinates_sds['CNV_value'].mean()\n",
    "duplications_grch37_sudmant_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d598726-8adc-4696-b5e0-0c4267040c6e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "90789782-536a-4f10-b4b5-60e5d62a4cee",
   "metadata": {},
   "source": [
    "GRCh37 Duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b2db73c9-02d8-47e0-a506-82ac31c4ba7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Literature</th>\n",
       "      <th>Mean size(Bp) Literature</th>\n",
       "      <th>Standard deviation Literature</th>\n",
       "      <th>Literature  Deletions</th>\n",
       "      <th>Literature Mean CN Deletions</th>\n",
       "      <th>Literature  Duplications</th>\n",
       "      <th>Literature Mean CN Duplications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>4316</td>\n",
       "      <td>32098.0</td>\n",
       "      <td>66435.0</td>\n",
       "      <td>333</td>\n",
       "      <td>0.726727</td>\n",
       "      <td>3983</td>\n",
       "      <td>10.283204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>3087</td>\n",
       "      <td>39603.0</td>\n",
       "      <td>76185.0</td>\n",
       "      <td>180</td>\n",
       "      <td>0.738889</td>\n",
       "      <td>2907</td>\n",
       "      <td>10.553492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>3704</td>\n",
       "      <td>36046.0</td>\n",
       "      <td>70746.0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.850000</td>\n",
       "      <td>3684</td>\n",
       "      <td>10.779045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>7680</td>\n",
       "      <td>91996.0</td>\n",
       "      <td>975897.0</td>\n",
       "      <td>5103</td>\n",
       "      <td>1.021968</td>\n",
       "      <td>2693</td>\n",
       "      <td>5.469177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>6267</td>\n",
       "      <td>104423.0</td>\n",
       "      <td>1079792.0</td>\n",
       "      <td>3883</td>\n",
       "      <td>1.028735</td>\n",
       "      <td>2489</td>\n",
       "      <td>5.618915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>19</td>\n",
       "      <td>11658011.0</td>\n",
       "      <td>15837724.0</td>\n",
       "      <td>12</td>\n",
       "      <td>0.969551</td>\n",
       "      <td>7</td>\n",
       "      <td>32.009488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>1245</td>\n",
       "      <td>342104.0</td>\n",
       "      <td>2404566.0</td>\n",
       "      <td>443</td>\n",
       "      <td>1.072734</td>\n",
       "      <td>831</td>\n",
       "      <td>7.153938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data  Literature  Mean size(Bp) Literature  \\\n",
       "0       GRCh37        4316                   32098.0   \n",
       "1        Genes        3087                   39603.0   \n",
       "2    Telomeres           0                       NaN   \n",
       "3          SDs        3704                   36046.0   \n",
       "4    T2T-CHM13        7680                   91996.0   \n",
       "5        Genes        6267                  104423.0   \n",
       "6    Telomeres          19                11658011.0   \n",
       "7          SDs        1245                  342104.0   \n",
       "\n",
       "   Standard deviation Literature  Literature  Deletions  \\\n",
       "0                        66435.0                    333   \n",
       "1                        76185.0                    180   \n",
       "2                            NaN                      0   \n",
       "3                        70746.0                     20   \n",
       "4                       975897.0                   5103   \n",
       "5                      1079792.0                   3883   \n",
       "6                     15837724.0                     12   \n",
       "7                      2404566.0                    443   \n",
       "\n",
       "   Literature Mean CN Deletions  Literature  Duplications  \\\n",
       "0                      0.726727                      3983   \n",
       "1                      0.738889                      2907   \n",
       "2                           NaN                         0   \n",
       "3                      0.850000                      3684   \n",
       "4                      1.021968                      2693   \n",
       "5                      1.028735                      2489   \n",
       "6                      0.969551                         7   \n",
       "7                      1.072734                       831   \n",
       "\n",
       "   Literature Mean CN Duplications  \n",
       "0                        10.283204  \n",
       "1                        10.553492  \n",
       "2                              NaN  \n",
       "3                        10.779045  \n",
       "4                         5.469177  \n",
       "5                         5.618915  \n",
       "6                        32.009488  \n",
       "7                         7.153938  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "literature = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'\n",
    "             ], \n",
    "     'Literature':[len(sudmant), \n",
    "                  len(peter_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(peter_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(peter_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(final_cnv_ranges), \n",
    "                  len(coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), ], \n",
    "     'Mean size(Bp) Literature':[round(sudmant.as_df()['Length_cnvr'].mean()), \n",
    "                      round(peter_coordinates_genes['Length_cnvr'].mean()), \n",
    "                      peter_coordinates_telomeres['Length_cnvr'].mean(), \n",
    "                      round(peter_coordinates_sds['Length_cnvr'].mean()), \n",
    "                      round(final_cnv['Length_cnvr'].mean()),\n",
    "                      round(coordinates_genes['Length_cnvr'].mean()),\n",
    "                      round(coordinates_telomeres['Length_cnvr'].mean()), \n",
    "                      round(coordinates_sds['Length_cnvr'].mean())],\n",
    "    'Standard deviation Literature':[round(sudmant.as_df()['Length_cnvr'].std()), \n",
    "                      round(peter_coordinates_genes['Length_cnvr'].std()), \n",
    "                      peter_coordinates_telomeres['Length_cnvr'].std(), \n",
    "                      round(peter_coordinates_sds['Length_cnvr'].std()), \n",
    "                      round(final_cnv['Length_cnvr'].std()),\n",
    "                      round(coordinates_genes['Length_cnvr'].std()),\n",
    "                      round(coordinates_telomeres['Length_cnvr'].std()), \n",
    "                      round(coordinates_sds['Length_cnvr'].std())],\n",
    "    'Literature  Deletions':[len(grch37_deletions_sudmant), \n",
    "                              len(deletions_grch37_sudmant_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(deletions_grch37_sudmant_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_grch37_sudmant_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_deletions_gb),\n",
    "                                     len(deletions_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "     \n",
    "     'Literature Mean CN Deletions':[grch37_deletions_sudmant_mean, \n",
    "                                     deletions_grch37_sudmant_mean_individual_genes, \n",
    "                                     deletions_grch37_sudmant_mean_individual_telomeres,\n",
    "                                     deletions_grch37_sudmant_mean_individual_sds,\n",
    "                                     t2t_deletions_gb_mean,\n",
    "                                     deletions_t2t_mean_individual_genes,\n",
    "                                     deletions_t2t_mean_individual_telomeres,\n",
    "                                     deletions_t2t_mean_individual_sds\n",
    "                                    ],\n",
    "\n",
    "     \n",
    "     'Literature  Duplications':[len(grch37_duplications_sudmant), \n",
    "                              len(duplications_grch37_sudmant_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(duplications_grch37_sudmant_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_grch37_sudmant_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_duplications_gb),\n",
    "                                     len(duplications_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "     'Literature Mean CN Duplications':[grch37_duplications_sudmant_mean,\n",
    "                                       duplications_grch37_sudmant_mean_individual_genes,\n",
    "                                       duplications_grch37_sudmant_mean_individual_telomeres,\n",
    "                                       duplications_grch37_sudmant_mean_individual_sds,\n",
    "                                       t2t_duplications_gb_mean,\n",
    "                                        duplications_t2t_mean_individual_genes,\n",
    "                                        duplications_t2t_mean_individual_telomeres,\n",
    "                                        duplications_t2t_mean_individual_sds],\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=literature)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "ed1f8e95-2418-464b-9402-cfa0f13215f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>CNVNator</th>\n",
       "      <th>Mean size(Bp)</th>\n",
       "      <th>Size (std)</th>\n",
       "      <th>Deletions</th>\n",
       "      <th>CNVnator Mean CN Deletions</th>\n",
       "      <th>Duplications</th>\n",
       "      <th>CNVnator Mean CN Duplications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>728</td>\n",
       "      <td>18512.0</td>\n",
       "      <td>40602</td>\n",
       "      <td>508</td>\n",
       "      <td>0.773704</td>\n",
       "      <td>220</td>\n",
       "      <td>4.375335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>383</td>\n",
       "      <td>23790.0</td>\n",
       "      <td>48781</td>\n",
       "      <td>272</td>\n",
       "      <td>0.786549</td>\n",
       "      <td>111</td>\n",
       "      <td>4.271579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>3</td>\n",
       "      <td>47899.0</td>\n",
       "      <td>32020</td>\n",
       "      <td>3</td>\n",
       "      <td>0.099335</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>430</td>\n",
       "      <td>27020.0</td>\n",
       "      <td>49140</td>\n",
       "      <td>234</td>\n",
       "      <td>0.852147</td>\n",
       "      <td>196</td>\n",
       "      <td>4.463832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>1855</td>\n",
       "      <td>66703.0</td>\n",
       "      <td>129066</td>\n",
       "      <td>1420</td>\n",
       "      <td>0.894372</td>\n",
       "      <td>435</td>\n",
       "      <td>3.969565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>386</td>\n",
       "      <td>64483.0</td>\n",
       "      <td>114921</td>\n",
       "      <td>224</td>\n",
       "      <td>0.997807</td>\n",
       "      <td>162</td>\n",
       "      <td>3.344081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>5</td>\n",
       "      <td>101599.0</td>\n",
       "      <td>78341</td>\n",
       "      <td>3</td>\n",
       "      <td>1.049467</td>\n",
       "      <td>2</td>\n",
       "      <td>4.295300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>569</td>\n",
       "      <td>59707.0</td>\n",
       "      <td>101041</td>\n",
       "      <td>347</td>\n",
       "      <td>0.981064</td>\n",
       "      <td>222</td>\n",
       "      <td>3.593323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data  CNVNator  Mean size(Bp)  Size (std)  Deletions  \\\n",
       "0       GRCh37       728        18512.0       40602        508   \n",
       "1        Genes       383        23790.0       48781        272   \n",
       "2    Telomeres         3        47899.0       32020          3   \n",
       "3          SDs       430        27020.0       49140        234   \n",
       "4    T2T-CHM13      1855        66703.0      129066       1420   \n",
       "5        Genes       386        64483.0      114921        224   \n",
       "6    Telomeres         5       101599.0       78341          3   \n",
       "7          SDs       569        59707.0      101041        347   \n",
       "\n",
       "   CNVnator Mean CN Deletions  Duplications  CNVnator Mean CN Duplications  \n",
       "0                    0.773704           220                       4.375335  \n",
       "1                    0.786549           111                       4.271579  \n",
       "2                    0.099335             0                            NaN  \n",
       "3                    0.852147           196                       4.463832  \n",
       "4                    0.894372           435                       3.969565  \n",
       "5                    0.997807           162                       3.344081  \n",
       "6                    1.049467             2                       4.295300  \n",
       "7                    0.981064           222                       3.593323  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'\n",
    "             ], \n",
    "     'CNVNator': [len(final_cnv_grch37), \n",
    "               len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),  \n",
    "               len(t2t_cnvnator), \n",
    "               len(t2t_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(t2t_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(t2t_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "              ], \n",
    "     'Mean size(Bp)':[final_cnv_grch37['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_genes['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_sds['Length_cnvr'].mean().round(), \n",
    "                      t2t_cnvnator.as_df()['Length_cnvr'].mean().round(),\n",
    "                      t2t_cnvnator_coordinates_genes['Length_cnvr'].mean().round(),\n",
    "                      t2t_cnvnator_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      t2t_cnvnator_coordinates_sds['Length_cnvr'].mean().round()],\n",
    "     'Size (std)':[round(final_cnv_grch37['Length_cnvr'].std()), \n",
    "                      round(grch37_coordinates_genes['Length_cnvr'].std()), \n",
    "                      round(grch37_coordinates_telomeres['Length_cnvr'].std()), \n",
    "                      round(grch37_coordinates_sds['Length_cnvr'].std()), \n",
    "                      round(t2t_cnvnator.as_df()['Length_cnvr'].std()),\n",
    "                      round(t2t_cnvnator_coordinates_genes['Length_cnvr'].std()),\n",
    "                      round(t2t_cnvnator_coordinates_telomeres['Length_cnvr'].std()), \n",
    "                      round(t2t_cnvnator_coordinates_sds['Length_cnvr'].std())],\n",
    "     \n",
    "     'Deletions':[len(grch37_deletions_cnvnator), \n",
    "                              len(deletions_grch37_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(deletions_grch37_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_grch37_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_deletions_cnvnator),\n",
    "                                     len(deletions_t2t_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "\n",
    "     'CNVnator Mean CN Deletions':[grch37_deletions_cnvnator_mean,\n",
    "                                   deletions_grch37_cnvnator_mean_individual_genes,\n",
    "                                   deletions_grch37_cnvnator_mean_individual_telomeres,\n",
    "                                   deletions_grch37_cnvnator_mean_individual_sds,\n",
    "                                   t2t_deletions_cnvnator_mean, \n",
    "                                   deletions_t2t_cnvnator_mean_individual_genes,\n",
    "                                   deletions_t2t_cnvnator_mean_individual_telomeres,\n",
    "                                   deletions_t2t_cnvnator_mean_individual_sds],\n",
    "\n",
    "    'Duplications':[len(grch37_duplications_cnvnator), \n",
    "                              len(duplications_grch37_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(duplications_grch37_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_grch37_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_duplications_cnvnator),\n",
    "                                     len(duplications_t2t_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "\n",
    "     'CNVnator Mean CN Duplications':[grch37_duplications_cnvnator_mean,\n",
    "                                      duplications_grch37_cnvnator_mean_individual_genes,\n",
    "                                      duplications_grch37_cnvnator_mean_individual_telomeres,\n",
    "                                      duplications_grch37_cnvnator_mean_individual_sds,\n",
    "                                      t2t_duplications_cnvnator_mean,\n",
    "                                      duplications_t2t_cnvnator_mean_individual_genes,\n",
    "                                      duplications_t2t_cnvnator_mean_individual_telomeres,\n",
    "                                      duplications_t2t_cnvnator_mean_individual_sds\n",
    "                                      \n",
    "                                     ],\n",
    "\n",
    "\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=d)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "95d6753a-e4b7-4fb0-b072-1f7647bae1a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3704"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(peter_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "800cc5c7-e145-4663-b41e-bf515f7f3615",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40602.20857036721"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_cnv_grch37['Length_cnvr'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fa2d77e-0449-4d85-9d29-393adbf42e4f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7edc021c-5cdf-4619-a626-eeacef1a0bb8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9ab70cd-3755-4d70-b838-4772d6a42771",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
