{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "00f332d4-1c1c-42c7-8382-7cb5fcab05d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "from pyranges import PyRanges\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b272d5b-706a-4c50-9f84-cc3e31f479e3",
   "metadata": {},
   "source": [
    "<h1> T2T CHM13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6b0442f5-74b7-4d26-9bc8-dab9a2b8ff71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/3419112927.py:2: DtypeWarning: Columns (23,36) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  t2t_genes = pd.read_csv('../annotation_files/T2T-chm13-catLiftOffGenes.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248384000</td>\n",
       "      <td>248387328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>0</td>\n",
       "      <td>3600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>242693800</td>\n",
       "      <td>242696752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>0</td>\n",
       "      <td>2800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr3</td>\n",
       "      <td>200904400</td>\n",
       "      <td>200905400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr3</td>\n",
       "      <td>201101200</td>\n",
       "      <td>201105948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr4</td>\n",
       "      <td>0</td>\n",
       "      <td>3400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr4</td>\n",
       "      <td>193572600</td>\n",
       "      <td>193574945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr5</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr5</td>\n",
       "      <td>182043800</td>\n",
       "      <td>182045439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr6</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr6</td>\n",
       "      <td>172123600</td>\n",
       "      <td>172126628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr7</td>\n",
       "      <td>0</td>\n",
       "      <td>3600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr7</td>\n",
       "      <td>160565200</td>\n",
       "      <td>160567428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr8</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr8</td>\n",
       "      <td>146257200</td>\n",
       "      <td>146259331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr9</td>\n",
       "      <td>0</td>\n",
       "      <td>3800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr9</td>\n",
       "      <td>150614400</td>\n",
       "      <td>150617247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr10</td>\n",
       "      <td>0</td>\n",
       "      <td>2800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr10</td>\n",
       "      <td>134754800</td>\n",
       "      <td>134758134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr11</td>\n",
       "      <td>0</td>\n",
       "      <td>2200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr11</td>\n",
       "      <td>201200</td>\n",
       "      <td>202800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr11</td>\n",
       "      <td>135125000</td>\n",
       "      <td>135127769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr12</td>\n",
       "      <td>0</td>\n",
       "      <td>3200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr12</td>\n",
       "      <td>133322000</td>\n",
       "      <td>133324548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr13</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr13</td>\n",
       "      <td>113563000</td>\n",
       "      <td>113566686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr14</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr14</td>\n",
       "      <td>101159600</td>\n",
       "      <td>101161492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr15</td>\n",
       "      <td>0</td>\n",
       "      <td>3600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr15</td>\n",
       "      <td>99750200</td>\n",
       "      <td>99753195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr16</td>\n",
       "      <td>0</td>\n",
       "      <td>2600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr16</td>\n",
       "      <td>96327400</td>\n",
       "      <td>96330374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr17</td>\n",
       "      <td>0</td>\n",
       "      <td>2200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr17</td>\n",
       "      <td>84273800</td>\n",
       "      <td>84276897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr18</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr18</td>\n",
       "      <td>80539000</td>\n",
       "      <td>80542538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr19</td>\n",
       "      <td>61704200</td>\n",
       "      <td>61707364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr20</td>\n",
       "      <td>0</td>\n",
       "      <td>2800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr20</td>\n",
       "      <td>66207000</td>\n",
       "      <td>66210255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr21</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr21</td>\n",
       "      <td>45086000</td>\n",
       "      <td>45090682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr22</td>\n",
       "      <td>0</td>\n",
       "      <td>4600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51321800</td>\n",
       "      <td>51324926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chrX</td>\n",
       "      <td>0</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154256800</td>\n",
       "      <td>154259566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chrY</td>\n",
       "      <td>0</td>\n",
       "      <td>5800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62453800</td>\n",
       "      <td>62460029</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome      Start        End\n",
       "0        chr1          0       3000\n",
       "1        chr1  248384000  248387328\n",
       "2        chr2          0       3600\n",
       "3        chr2  242693800  242696752\n",
       "4        chr3          0       2800\n",
       "5        chr3  200904400  200905400\n",
       "6        chr3  201101200  201105948\n",
       "7        chr4          0       3400\n",
       "8        chr4  193572600  193574945\n",
       "9        chr5          0       2400\n",
       "10       chr5  182043800  182045439\n",
       "11       chr6          0       3000\n",
       "12       chr6  172123600  172126628\n",
       "13       chr7          0       3600\n",
       "14       chr7  160565200  160567428\n",
       "15       chr8          0       2400\n",
       "16       chr8  146257200  146259331\n",
       "17       chr9          0       3800\n",
       "18       chr9  150614400  150617247\n",
       "19      chr10          0       2800\n",
       "20      chr10  134754800  134758134\n",
       "21      chr11          0       2200\n",
       "22      chr11     201200     202800\n",
       "23      chr11  135125000  135127769\n",
       "24      chr12          0       3200\n",
       "25      chr12  133322000  133324548\n",
       "26      chr13          0       3000\n",
       "27      chr13  113563000  113566686\n",
       "28      chr14          0       2400\n",
       "29      chr14  101159600  101161492\n",
       "30      chr15          0       3600\n",
       "31      chr15   99750200   99753195\n",
       "32      chr16          0       2600\n",
       "33      chr16   96327400   96330374\n",
       "34      chr17          0       2200\n",
       "35      chr17   84273800   84276897\n",
       "36      chr18          0       2400\n",
       "37      chr18   80539000   80542538\n",
       "38      chr19          0       2400\n",
       "39      chr19   61704200   61707364\n",
       "40      chr20          0       2800\n",
       "41      chr20   66207000   66210255\n",
       "42      chr21          0       3000\n",
       "43      chr21   45086000   45090682\n",
       "44      chr22          0       4600\n",
       "45      chr22   51321800   51324926\n",
       "46       chrX          0       2000\n",
       "47       chrX  154256800  154259566\n",
       "48       chrY          0       5800\n",
       "49       chrY   62453800   62460029"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Anotations\n",
    "t2t_genes = pd.read_csv('../annotation_files/T2T-chm13-catLiftOffGenes.csv')\n",
    "t2t_telomeres = pd.read_csv('../annotation_files/chm13v2.0_telomere.bed', sep='\\t', names=['Chromosome','Start', 'End'])\n",
    "t2t_sds = pd.read_csv('../annotation_files/chm13v2.0_SD.full.bed', sep='\\t')\n",
    "t2t_sds = t2t_sds[['#chr1', 'start1', 'end1']]\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "t2t_sds.columns =['Chromosome','Start','End']\n",
    "t2t_telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d730b68f-174e-41e2-ac7a-b23e1aa2e6eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/2170477776.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  chm13 = pd.read_csv('../call_cnvr_chm13/individual_cnv_cnvr.txt', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "# is not working \n",
    "chm13 = pd.read_csv('../call_cnvr_chm13/individual_cnv_cnvr.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "55cc3484-958c-4a32-94f5-7b0e88698454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNVR_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>CNV_Start</th>\n",
       "      <th>CNV_End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Length</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43107</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20039024</td>\n",
       "      <td>20042704</td>\n",
       "      <td>1</td>\n",
       "      <td>3681</td>\n",
       "      <td>35537179.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43108</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20072827</td>\n",
       "      <td>20080819</td>\n",
       "      <td>1</td>\n",
       "      <td>7993</td>\n",
       "      <td>35537179.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43109</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20093828</td>\n",
       "      <td>20095291</td>\n",
       "      <td>1</td>\n",
       "      <td>1464</td>\n",
       "      <td>35537179.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43110</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20127368</td>\n",
       "      <td>20128368</td>\n",
       "      <td>1</td>\n",
       "      <td>1001</td>\n",
       "      <td>35537179.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43111</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20142860</td>\n",
       "      <td>20144633</td>\n",
       "      <td>1</td>\n",
       "      <td>1774</td>\n",
       "      <td>35537179.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35252344</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248364795</td>\n",
       "      <td>248366351</td>\n",
       "      <td>29</td>\n",
       "      <td>1557</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35252345</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248366351</td>\n",
       "      <td>248367622</td>\n",
       "      <td>30</td>\n",
       "      <td>1272</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35252346</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248367622</td>\n",
       "      <td>248368745</td>\n",
       "      <td>27</td>\n",
       "      <td>1124</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35252347</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248368745</td>\n",
       "      <td>248370339</td>\n",
       "      <td>43</td>\n",
       "      <td>1595</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35252348</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248370339</td>\n",
       "      <td>248387328</td>\n",
       "      <td>39</td>\n",
       "      <td>16990</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25767500 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chr        Start          End      CNVR_ID          Sample_ID  \\\n",
       "43107     22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "43108     22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "43109     22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "43110     22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "43111     22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "...       ..          ...          ...          ...                ...   \n",
       "35252344   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "35252345   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "35252346   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "35252347   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "35252348   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "\n",
       "          CNV_Start    CNV_End  CNV_Value  Length  Length_cnvr  \n",
       "43107      20039024   20042704          1    3681   35537179.0  \n",
       "43108      20072827   20080819          1    7993   35537179.0  \n",
       "43109      20093828   20095291          1    1464   35537179.0  \n",
       "43110      20127368   20128368          1    1001   35537179.0  \n",
       "43111      20142860   20144633          1    1774   35537179.0  \n",
       "...             ...        ...        ...     ...          ...  \n",
       "35252344  248364795  248366351         29    1557      82796.0  \n",
       "35252345  248366351  248367622         30    1272      82796.0  \n",
       "35252346  248367622  248368745         27    1124      82796.0  \n",
       "35252347  248368745  248370339         43    1595      82796.0  \n",
       "35252348  248370339  248387328         39   16990      82796.0  \n",
       "\n",
       "[25767500 rows x 10 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13 = chm13.dropna()\n",
    "chm13['Length_cnvr'] = chm13['End'] - chm13['Start']\n",
    "#chm13 = chm13[chm13['Length_cnvr'] < 7000000]\n",
    "chm13 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "55b427d6-ec1f-4c01-8621-426f5f660809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNVR_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>CNV_Start</th>\n",
       "      <th>CNV_End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Length</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20039024</td>\n",
       "      <td>20042704</td>\n",
       "      <td>1</td>\n",
       "      <td>3681</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20072827</td>\n",
       "      <td>20080819</td>\n",
       "      <td>1</td>\n",
       "      <td>7993</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20093828</td>\n",
       "      <td>20095291</td>\n",
       "      <td>1</td>\n",
       "      <td>1464</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20127368</td>\n",
       "      <td>20128368</td>\n",
       "      <td>1</td>\n",
       "      <td>1001</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20142860</td>\n",
       "      <td>20144633</td>\n",
       "      <td>1</td>\n",
       "      <td>1774</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767495</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248364795</td>\n",
       "      <td>248366351</td>\n",
       "      <td>29</td>\n",
       "      <td>1557</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767496</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248366351</td>\n",
       "      <td>248367622</td>\n",
       "      <td>30</td>\n",
       "      <td>1272</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767497</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248367622</td>\n",
       "      <td>248368745</td>\n",
       "      <td>27</td>\n",
       "      <td>1124</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767498</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248368745</td>\n",
       "      <td>248370339</td>\n",
       "      <td>43</td>\n",
       "      <td>1595</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767499</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248370339</td>\n",
       "      <td>248387328</td>\n",
       "      <td>39</td>\n",
       "      <td>16990</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25767500 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chr        Start          End      CNVR_ID          Sample_ID  \\\n",
       "0         22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "1         22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "2         22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "3         22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "4         22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "...       ..          ...          ...          ...                ...   \n",
       "25767495   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767496   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767497   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767498   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767499   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "\n",
       "          CNV_Start    CNV_End  CNV_Value  Length  Length_cnvr  \\\n",
       "0          20039024   20042704          1    3681   35537179.0   \n",
       "1          20072827   20080819          1    7993   35537179.0   \n",
       "2          20093828   20095291          1    1464   35537179.0   \n",
       "3          20127368   20128368          1    1001   35537179.0   \n",
       "4          20142860   20144633          1    1774   35537179.0   \n",
       "...             ...        ...        ...     ...          ...   \n",
       "25767495  248364795  248366351         29    1557      82796.0   \n",
       "25767496  248366351  248367622         30    1272      82796.0   \n",
       "25767497  248367622  248368745         27    1124      82796.0   \n",
       "25767498  248368745  248370339         43    1595      82796.0   \n",
       "25767499  248370339  248387328         39   16990      82796.0   \n",
       "\n",
       "                     Sample       Region  \n",
       "0         LP6005442-DNA_A11  WestEurasia  \n",
       "1         LP6005442-DNA_A11  WestEurasia  \n",
       "2         LP6005442-DNA_A11  WestEurasia  \n",
       "3         LP6005442-DNA_A11  WestEurasia  \n",
       "4         LP6005442-DNA_A11  WestEurasia  \n",
       "...                     ...          ...  \n",
       "25767495  LP6005441-DNA_B09     EastAsia  \n",
       "25767496  LP6005441-DNA_B09     EastAsia  \n",
       "25767497  LP6005441-DNA_B09     EastAsia  \n",
       "25767498  LP6005441-DNA_B09     EastAsia  \n",
       "25767499  LP6005441-DNA_B09     EastAsia  \n",
       "\n",
       "[25767500 rows x 12 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "## Keeping only samples that are on old cnvs \n",
    "chm13 = chm13.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])\n",
    "\n",
    "\n",
    "chm13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ac0ab3f0-db9c-4b57-a3aa-06fdecb87f48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNVR_ID</th>\n",
       "      <th>Sample_iD</th>\n",
       "      <th>CNV_start</th>\n",
       "      <th>CNV_end</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Length</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20039024</td>\n",
       "      <td>20042704</td>\n",
       "      <td>1</td>\n",
       "      <td>3681</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20072827</td>\n",
       "      <td>20080819</td>\n",
       "      <td>1</td>\n",
       "      <td>7993</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20093828</td>\n",
       "      <td>20095291</td>\n",
       "      <td>1</td>\n",
       "      <td>1464</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20127368</td>\n",
       "      <td>20128368</td>\n",
       "      <td>1</td>\n",
       "      <td>1001</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747.0</td>\n",
       "      <td>51324926.0</td>\n",
       "      <td>CNVR_109275</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>20142860</td>\n",
       "      <td>20144633</td>\n",
       "      <td>1</td>\n",
       "      <td>1774</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>LP6005442-DNA_A11</td>\n",
       "      <td>WestEurasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767495</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248364795</td>\n",
       "      <td>248366351</td>\n",
       "      <td>29</td>\n",
       "      <td>1557</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767496</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248366351</td>\n",
       "      <td>248367622</td>\n",
       "      <td>30</td>\n",
       "      <td>1272</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767497</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248367622</td>\n",
       "      <td>248368745</td>\n",
       "      <td>27</td>\n",
       "      <td>1124</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767498</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248368745</td>\n",
       "      <td>248370339</td>\n",
       "      <td>43</td>\n",
       "      <td>1595</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25767499</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>248370339</td>\n",
       "      <td>248387328</td>\n",
       "      <td>39</td>\n",
       "      <td>16990</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>LP6005441-DNA_B09</td>\n",
       "      <td>EastAsia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25767500 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome        Start          End      CNVR_ID          Sample_iD  \\\n",
       "0                22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "1                22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "2                22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "3                22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "4                22   15787747.0   51324926.0  CNVR_109275  LP6005442-DNA_A11   \n",
       "...             ...          ...          ...          ...                ...   \n",
       "25767495          1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767496          1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767497          1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767498          1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "25767499          1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_B09   \n",
       "\n",
       "          CNV_start    CNV_end  CNV_Value  Length  Length_cnvr  \\\n",
       "0          20039024   20042704          1    3681   35537179.0   \n",
       "1          20072827   20080819          1    7993   35537179.0   \n",
       "2          20093828   20095291          1    1464   35537179.0   \n",
       "3          20127368   20128368          1    1001   35537179.0   \n",
       "4          20142860   20144633          1    1774   35537179.0   \n",
       "...             ...        ...        ...     ...          ...   \n",
       "25767495  248364795  248366351         29    1557      82796.0   \n",
       "25767496  248366351  248367622         30    1272      82796.0   \n",
       "25767497  248367622  248368745         27    1124      82796.0   \n",
       "25767498  248368745  248370339         43    1595      82796.0   \n",
       "25767499  248370339  248387328         39   16990      82796.0   \n",
       "\n",
       "                     Sample       Region  \n",
       "0         LP6005442-DNA_A11  WestEurasia  \n",
       "1         LP6005442-DNA_A11  WestEurasia  \n",
       "2         LP6005442-DNA_A11  WestEurasia  \n",
       "3         LP6005442-DNA_A11  WestEurasia  \n",
       "4         LP6005442-DNA_A11  WestEurasia  \n",
       "...                     ...          ...  \n",
       "25767495  LP6005441-DNA_B09     EastAsia  \n",
       "25767496  LP6005441-DNA_B09     EastAsia  \n",
       "25767497  LP6005441-DNA_B09     EastAsia  \n",
       "25767498  LP6005441-DNA_B09     EastAsia  \n",
       "25767499  LP6005441-DNA_B09     EastAsia  \n",
       "\n",
       "[25767500 rows x 12 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#REMOVE SAMPLES WITH ABERRATIONS\n",
    "samples_with_aberrations = ['LP6005441-DNA_B03', ' LP6005443-DNA_G01', 'LP6005441-DNA_G09', 'LP6005441-DNA_D04'] \n",
    "\n",
    "chm13 = chm13[~chm13.Sample.isin(samples_with_aberrations)]\n",
    "\n",
    "\n",
    "chm13#.value_counts(by=[\"Sample_ID\"])\n",
    "chm13.columns = ['Chromosome', 'Start', 'End', 'CNVR_ID', 'Sample_iD', 'CNV_start', 'CNV_end', 'CNV_Value','Length', 'Length_cnvr','Sample','Region']\n",
    "chm13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a4ff43d7-b75b-406c-8601-e49e532c98b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_A01</th>\n",
       "      <th>LP6005441-DNA_A03</th>\n",
       "      <th>LP6005441-DNA_A04</th>\n",
       "      <th>LP6005441-DNA_A05</th>\n",
       "      <th>LP6005441-DNA_A06</th>\n",
       "      <th>LP6005441-DNA_A08</th>\n",
       "      <th>...</th>\n",
       "      <th>LP6005677-DNA_D03</th>\n",
       "      <th>LP6005677-DNA_E01</th>\n",
       "      <th>LP6005677-DNA_F01</th>\n",
       "      <th>LP6005677-DNA_G01</th>\n",
       "      <th>LP6007068-DNA_A01</th>\n",
       "      <th>LP6007069-DNA_A01</th>\n",
       "      <th>SS6004471</th>\n",
       "      <th>SS6004477</th>\n",
       "      <th>SS6004478</th>\n",
       "      <th>SS6004480</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>18.08</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>18.446154</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>436980.0</td>\n",
       "      <td>452948.0</td>\n",
       "      <td>15968.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.00</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>453948.0</td>\n",
       "      <td>466894.0</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>472779.0</td>\n",
       "      <td>484339.0</td>\n",
       "      <td>11560.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>485339.0</td>\n",
       "      <td>511499.0</td>\n",
       "      <td>26160.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178458</th>\n",
       "      <td>9</td>\n",
       "      <td>149620306.0</td>\n",
       "      <td>149794372.0</td>\n",
       "      <td>174066.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.636364</td>\n",
       "      <td>2.666667</td>\n",
       "      <td>3.166667</td>\n",
       "      <td>2.166667</td>\n",
       "      <td>2.526316</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178459</th>\n",
       "      <td>9</td>\n",
       "      <td>149796085.0</td>\n",
       "      <td>149801998.0</td>\n",
       "      <td>5913.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178460</th>\n",
       "      <td>9</td>\n",
       "      <td>149803678.0</td>\n",
       "      <td>149871435.0</td>\n",
       "      <td>67757.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.555556</td>\n",
       "      <td>1.555556</td>\n",
       "      <td>2.826087</td>\n",
       "      <td>1.428571</td>\n",
       "      <td>1.700000</td>\n",
       "      <td>0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178461</th>\n",
       "      <td>9</td>\n",
       "      <td>149876993.0</td>\n",
       "      <td>150167512.0</td>\n",
       "      <td>290519.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3.3000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.555556</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.800000</td>\n",
       "      <td>2.785714</td>\n",
       "      <td>3.096386</td>\n",
       "      <td>3.083333</td>\n",
       "      <td>2.928571</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178462</th>\n",
       "      <td>9</td>\n",
       "      <td>150171200.0</td>\n",
       "      <td>150617247.0</td>\n",
       "      <td>446047.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>22.2375</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22.848101</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22.225000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>21.734940</td>\n",
       "      <td>23.156627</td>\n",
       "      <td>13.487179</td>\n",
       "      <td>20.476190</td>\n",
       "      <td>21.132530</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178463 rows × 270 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome        Start          End  Length_cnvr  LP6005441-DNA_A01  \\\n",
       "0               1          0.0     403567.0     403567.0                2.0   \n",
       "1               1     436980.0     452948.0      15968.0                2.0   \n",
       "2               1     453948.0     466894.0      12946.0                2.0   \n",
       "3               1     472779.0     484339.0      11560.0                2.0   \n",
       "4               1     485339.0     511499.0      26160.0                2.0   \n",
       "...           ...          ...          ...          ...                ...   \n",
       "178458          9  149620306.0  149794372.0     174066.0                1.0   \n",
       "178459          9  149796085.0  149801998.0       5913.0                2.0   \n",
       "178460          9  149803678.0  149871435.0      67757.0                1.0   \n",
       "178461          9  149876993.0  150167512.0     290519.0                1.0   \n",
       "178462          9  150171200.0  150617247.0     446047.0                2.0   \n",
       "\n",
       "Sample  LP6005441-DNA_A03  LP6005441-DNA_A04  LP6005441-DNA_A05  \\\n",
       "0                   18.08             2.0000                2.0   \n",
       "1                    6.00             2.0000                2.0   \n",
       "2                    2.00             2.0000                2.0   \n",
       "3                    2.00             2.0000                2.0   \n",
       "4                    3.00             2.0000                2.0   \n",
       "...                   ...                ...                ...   \n",
       "178458               1.00             1.0000                1.0   \n",
       "178459               2.00             2.0000                2.0   \n",
       "178460               1.00             1.0000                1.0   \n",
       "178461               1.00             3.3000                2.0   \n",
       "178462               2.00            22.2375                1.0   \n",
       "\n",
       "Sample  LP6005441-DNA_A06  LP6005441-DNA_A08  ...  LP6005677-DNA_D03  \\\n",
       "0                2.000000                2.0  ...                2.0   \n",
       "1                2.000000                2.0  ...                2.0   \n",
       "2                2.000000                2.0  ...                2.0   \n",
       "3                2.000000                2.0  ...                2.0   \n",
       "4                2.000000                2.0  ...                2.0   \n",
       "...                   ...                ...  ...                ...   \n",
       "178458           2.000000                1.0  ...                1.0   \n",
       "178459           2.000000                2.0  ...                2.0   \n",
       "178460           1.000000                1.0  ...                1.0   \n",
       "178461           3.555556                1.0  ...                2.0   \n",
       "178462          22.848101                1.0  ...                1.0   \n",
       "\n",
       "Sample  LP6005677-DNA_E01  LP6005677-DNA_F01  LP6005677-DNA_G01  \\\n",
       "0                2.000000                2.0          18.446154   \n",
       "1                2.000000                2.0           6.000000   \n",
       "2                2.000000                2.0           3.000000   \n",
       "3                2.000000                2.0           2.000000   \n",
       "4                2.000000                2.0           2.000000   \n",
       "...                   ...                ...                ...   \n",
       "178458           1.000000                1.0           2.000000   \n",
       "178459           2.000000                2.0           2.000000   \n",
       "178460           1.000000                1.0           1.000000   \n",
       "178461           3.333333                1.0           1.000000   \n",
       "178462          22.225000                2.0           2.000000   \n",
       "\n",
       "Sample  LP6007068-DNA_A01  LP6007069-DNA_A01  SS6004471  SS6004477  SS6004478  \\\n",
       "0                2.000000           2.000000   2.000000   2.000000   2.000000   \n",
       "1                2.000000           2.000000   2.000000   2.000000   2.000000   \n",
       "2                2.000000           2.000000   2.000000   2.000000   2.000000   \n",
       "3                2.000000           2.000000   2.000000   2.000000   2.000000   \n",
       "4                2.000000           2.000000   2.000000   2.000000   2.000000   \n",
       "...                   ...                ...        ...        ...        ...   \n",
       "178458           2.636364           2.666667   3.166667   2.166667   2.526316   \n",
       "178459           2.000000           2.000000   3.000000   2.000000   2.000000   \n",
       "178460           1.555556           1.555556   2.826087   1.428571   1.700000   \n",
       "178461           2.800000           2.785714   3.096386   3.083333   2.928571   \n",
       "178462          21.734940          23.156627  13.487179  20.476190  21.132530   \n",
       "\n",
       "Sample  SS6004480  \n",
       "0        2.000000  \n",
       "1        2.000000  \n",
       "2        2.000000  \n",
       "3        2.000000  \n",
       "4        2.000000  \n",
       "...           ...  \n",
       "178458   1.000000  \n",
       "178459   2.000000  \n",
       "178460   0.833333  \n",
       "178461   1.000000  \n",
       "178462   1.000000  \n",
       "\n",
       "[178463 rows x 270 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv = chm13.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='CNV_Value',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "\n",
    "final_cnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c3dd5fce-15e1-4360-ab58-17d628a16655",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "final_cnv_ranges = PyRanges(final_cnv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1ebf6094-c3e2-453d-b1f9-3b733adfb293",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2t_genes.columns = ['Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'thickStart', 'thickEnd', 'reserved', 'blockCount', 'blockSizes',\n",
    "       'chromStarts', 'name2', 'cdsStartStat', 'cdsEndStat', 'exonFrames',\n",
    "       'txId', 'type', 'geneName', 'geneType', 'sourceGene',\n",
    "       'sourceTranscript', 'alignmentId', 'alternativeSourceTranscripts',\n",
    "       'Paralogy', 'UnfilteredParalogy', 'collapsedGeneIds',\n",
    "       'collapsedGeneNames', 'frameshift', 'exonAnnotationSupport',\n",
    "       'intronAnnotationSupport', 'transcriptClass', 'transcriptModes',\n",
    "       'validStart', 'validStop', 'properOrf', 'extra_paralog']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "270bf858-3acc-485c-90d3-88cb319bd888",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7ab381cf-130e-4900-b541-f7756d20c889",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "t2t_genes['Chromosome'] = t2t_genes['Chromosome'].str.replace('chr', '')\n",
    "t2t_genes = PyRanges(t2t_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0dfeb712-38ba-40f6-9c41-2674a31bb08b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### t2t genes \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6679ce82-0b3c-45fd-97d4-3fced891d9c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_A01</th>\n",
       "      <th>LP6005441-DNA_A03</th>\n",
       "      <th>LP6005441-DNA_A04</th>\n",
       "      <th>LP6005441-DNA_A05</th>\n",
       "      <th>LP6005441-DNA_A06</th>\n",
       "      <th>LP6005441-DNA_A08</th>\n",
       "      <th>...</th>\n",
       "      <th>frameshift</th>\n",
       "      <th>exonAnnotationSupport</th>\n",
       "      <th>intronAnnotationSupport</th>\n",
       "      <th>transcriptClass</th>\n",
       "      <th>transcriptModes</th>\n",
       "      <th>validStart</th>\n",
       "      <th>validStop</th>\n",
       "      <th>properOrf</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>18.080000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>436980</td>\n",
       "      <td>452948</td>\n",
       "      <td>15968.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>453948</td>\n",
       "      <td>466894</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>472779</td>\n",
       "      <td>484339</td>\n",
       "      <td>11560.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>485339</td>\n",
       "      <td>511499</td>\n",
       "      <td>26160.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178458</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>110.172885</td>\n",
       "      <td>145.875912</td>\n",
       "      <td>158.807910</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>110.482155</td>\n",
       "      <td>160.080597</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178459</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>52.759434</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178460</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>10.165951</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178461</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>111.294022</td>\n",
       "      <td>146.077591</td>\n",
       "      <td>153.602821</td>\n",
       "      <td>206.970825</td>\n",
       "      <td>120.948192</td>\n",
       "      <td>151.150567</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178462</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>10.664289</td>\n",
       "      <td>9.151965</td>\n",
       "      <td>8.807181</td>\n",
       "      <td>0.970000</td>\n",
       "      <td>9.927382</td>\n",
       "      <td>9.674572</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139355 rows × 307 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End  Length_cnvr  LP6005441-DNA_A01  \\\n",
       "0               1         0    403567     403567.0           2.000000   \n",
       "1               1    436980    452948      15968.0           2.000000   \n",
       "2               1    453948    466894      12946.0           2.000000   \n",
       "3               1    472779    484339      11560.0           2.000000   \n",
       "4               1    485339    511499      26160.0           2.000000   \n",
       "...           ...       ...       ...          ...                ...   \n",
       "178458         21         0  45090682   45090682.0         110.172885   \n",
       "178459         22         0  12764469   12764469.0           2.000000   \n",
       "178460         22  15787747  51324926   35537179.0           2.000000   \n",
       "178461         22         0  12764469   12764469.0         111.294022   \n",
       "178462         22  15787747  51324926   35537179.0          10.664289   \n",
       "\n",
       "        LP6005441-DNA_A03  LP6005441-DNA_A04  LP6005441-DNA_A05  \\\n",
       "0               18.080000           2.000000           2.000000   \n",
       "1                6.000000           2.000000           2.000000   \n",
       "2                2.000000           2.000000           2.000000   \n",
       "3                2.000000           2.000000           2.000000   \n",
       "4                3.000000           2.000000           2.000000   \n",
       "...                   ...                ...                ...   \n",
       "178458         145.875912         158.807910           1.000000   \n",
       "178459           2.000000           2.000000          52.759434   \n",
       "178460           2.000000           2.000000          10.165951   \n",
       "178461         146.077591         153.602821         206.970825   \n",
       "178462           9.151965           8.807181           0.970000   \n",
       "\n",
       "        LP6005441-DNA_A06  LP6005441-DNA_A08  ...  frameshift  \\\n",
       "0                2.000000           2.000000  ...         NaN   \n",
       "1                2.000000           2.000000  ...         NaN   \n",
       "2                2.000000           2.000000  ...       False   \n",
       "3                2.000000           2.000000  ...       False   \n",
       "4                2.000000           2.000000  ...       False   \n",
       "...                   ...                ...  ...         ...   \n",
       "178458         110.482155         160.080597  ...         NaN   \n",
       "178459           2.000000           2.000000  ...         NaN   \n",
       "178460           2.000000           2.000000  ...         NaN   \n",
       "178461         120.948192         151.150567  ...         NaN   \n",
       "178462           9.927382           9.674572  ...         NaN   \n",
       "\n",
       "                                    exonAnnotationSupport  \\\n",
       "0                                                     NaN   \n",
       "1                                                   1,1,1   \n",
       "2       1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "3       1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "4       1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "...                                                   ...   \n",
       "178458                                                NaN   \n",
       "178459                                                NaN   \n",
       "178460                                                NaN   \n",
       "178461                                                NaN   \n",
       "178462                                                NaN   \n",
       "\n",
       "                                  intronAnnotationSupport  transcriptClass  \\\n",
       "0                                                     NaN          paralog   \n",
       "1                                                     1,1         ortholog   \n",
       "2       2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...         ortholog   \n",
       "3       2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...         ortholog   \n",
       "4       2,1,2,2,1,0,1,1,1,1,1,1,1,2,2,2,1,1,1,2,2,1,1,...         ortholog   \n",
       "...                                                   ...              ...   \n",
       "178458                                                NaN          paralog   \n",
       "178459                                                NaN          paralog   \n",
       "178460                                                NaN          paralog   \n",
       "178461                                                NaN          paralog   \n",
       "178462                                                NaN          paralog   \n",
       "\n",
       "        transcriptModes  validStart  validStop  properOrf  extra_paralog  \\\n",
       "0               Liftoff         NaN        NaN        NaN           True   \n",
       "1              transMap        True       True       True          False   \n",
       "2              transMap        True       True       True          False   \n",
       "3              transMap        True       True       True          False   \n",
       "4              transMap        True       True       True          False   \n",
       "...                 ...         ...        ...        ...            ...   \n",
       "178458          Liftoff         NaN        NaN        NaN           True   \n",
       "178459          Liftoff         NaN        NaN        NaN           True   \n",
       "178460          Liftoff         NaN        NaN        NaN           True   \n",
       "178461          Liftoff         NaN        NaN        NaN           True   \n",
       "178462          Liftoff         NaN        NaN        NaN           True   \n",
       "\n",
       "        Distance  \n",
       "0              0  \n",
       "1              0  \n",
       "2              0  \n",
       "3              0  \n",
       "4              0  \n",
       "...          ...  \n",
       "178458         0  \n",
       "178459         0  \n",
       "178460         0  \n",
       "178461         0  \n",
       "178462         0  \n",
       "\n",
       "[139355 rows x 307 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_genes = final_cnv_ranges.nearest(t2t_genes)\n",
    "coordinates_genes = coordinates_genes.as_df()\n",
    "coordinates_genes = coordinates_genes[coordinates_genes['Distance'] <1 ]\n",
    "coordinates_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1254733b-7479-424d-80a0-45f18a185c99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "84326"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de30a634-6db2-4111-86a7-9107e7e7c97c",
   "metadata": {},
   "source": [
    "<h2> Telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "05ec3b37-97a2-4222-8816-352478b22cd2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>248384000</td>\n",
       "      <td>248387328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>242693800</td>\n",
       "      <td>242696752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>200904400</td>\n",
       "      <td>200905400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>201101200</td>\n",
       "      <td>201105948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4</td>\n",
       "      <td>193572600</td>\n",
       "      <td>193574945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5</td>\n",
       "      <td>182043800</td>\n",
       "      <td>182045439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>6</td>\n",
       "      <td>172123600</td>\n",
       "      <td>172126628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>3600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>7</td>\n",
       "      <td>160565200</td>\n",
       "      <td>160567428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>8</td>\n",
       "      <td>146257200</td>\n",
       "      <td>146259331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>3800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>9</td>\n",
       "      <td>150614400</td>\n",
       "      <td>150617247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>2800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>10</td>\n",
       "      <td>134754800</td>\n",
       "      <td>134758134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>2200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>11</td>\n",
       "      <td>201200</td>\n",
       "      <td>202800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>11</td>\n",
       "      <td>135125000</td>\n",
       "      <td>135127769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>3200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>12</td>\n",
       "      <td>133322000</td>\n",
       "      <td>133324548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>13</td>\n",
       "      <td>113563000</td>\n",
       "      <td>113566686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>14</td>\n",
       "      <td>101159600</td>\n",
       "      <td>101161492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>3600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>15</td>\n",
       "      <td>99750200</td>\n",
       "      <td>99753195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>16</td>\n",
       "      <td>96327400</td>\n",
       "      <td>96330374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>2200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>17</td>\n",
       "      <td>84273800</td>\n",
       "      <td>84276897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>18</td>\n",
       "      <td>80539000</td>\n",
       "      <td>80542538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>2400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>19</td>\n",
       "      <td>61704200</td>\n",
       "      <td>61707364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>2800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>20</td>\n",
       "      <td>66207000</td>\n",
       "      <td>66210255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>21</td>\n",
       "      <td>45086000</td>\n",
       "      <td>45090682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>4600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>22</td>\n",
       "      <td>51321800</td>\n",
       "      <td>51324926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>X</td>\n",
       "      <td>0</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>X</td>\n",
       "      <td>154256800</td>\n",
       "      <td>154259566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Y</td>\n",
       "      <td>0</td>\n",
       "      <td>5800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Y</td>\n",
       "      <td>62453800</td>\n",
       "      <td>62460029</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+\n",
       "| Chromosome   | Start     | End       |\n",
       "| (category)   | (int64)   | (int64)   |\n",
       "|--------------+-----------+-----------|\n",
       "| 1            | 0         | 3000      |\n",
       "| 1            | 248384000 | 248387328 |\n",
       "| 2            | 0         | 3600      |\n",
       "| 2            | 242693800 | 242696752 |\n",
       "| ...          | ...       | ...       |\n",
       "| X            | 154256800 | 154259566 |\n",
       "| X            | 0         | 2000      |\n",
       "| Y            | 62453800  | 62460029  |\n",
       "| Y            | 0         | 5800      |\n",
       "+--------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 50 rows and 3 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2t_telomeres['Chromosome'] = t2t_telomeres['Chromosome'].str.replace('chr', '')\n",
    "t2t_telomeres = PyRanges(t2t_telomeres)\n",
    "t2t_telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e890b682-23b7-4e09-b497-8c3429f9718a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_A01</th>\n",
       "      <th>LP6005441-DNA_A03</th>\n",
       "      <th>LP6005441-DNA_A04</th>\n",
       "      <th>LP6005441-DNA_A05</th>\n",
       "      <th>LP6005441-DNA_A06</th>\n",
       "      <th>LP6005441-DNA_A08</th>\n",
       "      <th>...</th>\n",
       "      <th>LP6005677-DNA_G01</th>\n",
       "      <th>LP6007068-DNA_A01</th>\n",
       "      <th>LP6007069-DNA_A01</th>\n",
       "      <th>SS6004471</th>\n",
       "      <th>SS6004477</th>\n",
       "      <th>SS6004478</th>\n",
       "      <th>SS6004480</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>18.080000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>18.446154</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>16.488189</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>16.674419</td>\n",
       "      <td>17.312500</td>\n",
       "      <td>17.437500</td>\n",
       "      <td>16.283465</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>15.319728</td>\n",
       "      <td>15.862069</td>\n",
       "      <td>16.807407</td>\n",
       "      <td>15.592593</td>\n",
       "      <td>14.821918</td>\n",
       "      <td>16.277778</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532</td>\n",
       "      <td>248387328</td>\n",
       "      <td>82796.0</td>\n",
       "      <td>20.875000</td>\n",
       "      <td>21.444444</td>\n",
       "      <td>21.111111</td>\n",
       "      <td>25.250000</td>\n",
       "      <td>23.250000</td>\n",
       "      <td>21.750000</td>\n",
       "      <td>...</td>\n",
       "      <td>23.111111</td>\n",
       "      <td>17.909091</td>\n",
       "      <td>18.909091</td>\n",
       "      <td>17.700000</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>18.200000</td>\n",
       "      <td>18.909091</td>\n",
       "      <td>248384000</td>\n",
       "      <td>248387328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13958</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>65036</td>\n",
       "      <td>65036.0</td>\n",
       "      <td>8.400000</td>\n",
       "      <td>9.300000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>10.100000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>3600</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13959</th>\n",
       "      <td>2</td>\n",
       "      <td>242413858</td>\n",
       "      <td>242696752</td>\n",
       "      <td>282894.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>14.642857</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>16.254545</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>242693800</td>\n",
       "      <td>242696752</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178458</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>110.172885</td>\n",
       "      <td>145.875912</td>\n",
       "      <td>158.807910</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>110.482155</td>\n",
       "      <td>160.080597</td>\n",
       "      <td>...</td>\n",
       "      <td>129.256325</td>\n",
       "      <td>98.383085</td>\n",
       "      <td>95.412666</td>\n",
       "      <td>108.014702</td>\n",
       "      <td>153.940968</td>\n",
       "      <td>89.884274</td>\n",
       "      <td>84.726368</td>\n",
       "      <td>0</td>\n",
       "      <td>3000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178459</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>52.759434</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>4600</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178460</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>10.165951</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>51321800</td>\n",
       "      <td>51324926</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178461</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>111.294022</td>\n",
       "      <td>146.077591</td>\n",
       "      <td>153.602821</td>\n",
       "      <td>206.970825</td>\n",
       "      <td>120.948192</td>\n",
       "      <td>151.150567</td>\n",
       "      <td>...</td>\n",
       "      <td>147.741988</td>\n",
       "      <td>118.177010</td>\n",
       "      <td>121.240217</td>\n",
       "      <td>142.797625</td>\n",
       "      <td>172.307776</td>\n",
       "      <td>118.426087</td>\n",
       "      <td>117.115406</td>\n",
       "      <td>0</td>\n",
       "      <td>4600</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178462</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>10.664289</td>\n",
       "      <td>9.151965</td>\n",
       "      <td>8.807181</td>\n",
       "      <td>0.970000</td>\n",
       "      <td>9.927382</td>\n",
       "      <td>9.674572</td>\n",
       "      <td>...</td>\n",
       "      <td>10.482103</td>\n",
       "      <td>7.742838</td>\n",
       "      <td>7.805637</td>\n",
       "      <td>8.050934</td>\n",
       "      <td>8.420014</td>\n",
       "      <td>7.413285</td>\n",
       "      <td>7.399339</td>\n",
       "      <td>51321800</td>\n",
       "      <td>51324926</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83 rows × 273 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start        End  Length_cnvr  LP6005441-DNA_A01  \\\n",
       "0               1          0     403567     403567.0           2.000000   \n",
       "1               1          0     403567     403567.0          16.488189   \n",
       "2               1  248304532  248387328      82796.0          20.875000   \n",
       "13958           2          0      65036      65036.0           8.400000   \n",
       "13959           2  242413858  242696752     282894.0           2.000000   \n",
       "...           ...        ...        ...          ...                ...   \n",
       "178458         21          0   45090682   45090682.0         110.172885   \n",
       "178459         22          0   12764469   12764469.0           2.000000   \n",
       "178460         22   15787747   51324926   35537179.0           2.000000   \n",
       "178461         22          0   12764469   12764469.0         111.294022   \n",
       "178462         22   15787747   51324926   35537179.0          10.664289   \n",
       "\n",
       "        LP6005441-DNA_A03  LP6005441-DNA_A04  LP6005441-DNA_A05  \\\n",
       "0               18.080000           2.000000           2.000000   \n",
       "1                1.000000          16.674419          17.312500   \n",
       "2               21.444444          21.111111          25.250000   \n",
       "13958            9.300000           2.000000           2.000000   \n",
       "13959           14.642857           2.000000           2.000000   \n",
       "...                   ...                ...                ...   \n",
       "178458         145.875912         158.807910           1.000000   \n",
       "178459           2.000000           2.000000          52.759434   \n",
       "178460           2.000000           2.000000          10.165951   \n",
       "178461         146.077591         153.602821         206.970825   \n",
       "178462           9.151965           8.807181           0.970000   \n",
       "\n",
       "        LP6005441-DNA_A06  LP6005441-DNA_A08  ...  LP6005677-DNA_G01  \\\n",
       "0                2.000000           2.000000  ...          18.446154   \n",
       "1               17.437500          16.283465  ...           1.000000   \n",
       "2               23.250000          21.750000  ...          23.111111   \n",
       "13958            2.000000           2.000000  ...          10.100000   \n",
       "13959            2.000000           2.000000  ...          16.254545   \n",
       "...                   ...                ...  ...                ...   \n",
       "178458         110.482155         160.080597  ...         129.256325   \n",
       "178459           2.000000           2.000000  ...           2.000000   \n",
       "178460           2.000000           2.000000  ...           2.000000   \n",
       "178461         120.948192         151.150567  ...         147.741988   \n",
       "178462           9.927382           9.674572  ...          10.482103   \n",
       "\n",
       "        LP6007068-DNA_A01  LP6007069-DNA_A01   SS6004471   SS6004477  \\\n",
       "0                2.000000           2.000000    2.000000    2.000000   \n",
       "1               15.319728          15.862069   16.807407   15.592593   \n",
       "2               17.909091          18.909091   17.700000   16.000000   \n",
       "13958            2.000000           2.000000    2.000000    2.000000   \n",
       "13959            2.000000           2.000000    2.000000    2.000000   \n",
       "...                   ...                ...         ...         ...   \n",
       "178458          98.383085          95.412666  108.014702  153.940968   \n",
       "178459           2.000000           2.000000    2.000000    2.000000   \n",
       "178460           2.000000           2.000000    2.000000    2.000000   \n",
       "178461         118.177010         121.240217  142.797625  172.307776   \n",
       "178462           7.742838           7.805637    8.050934    8.420014   \n",
       "\n",
       "         SS6004478   SS6004480    Start_b      End_b  Distance  \n",
       "0         2.000000    2.000000          0       3000         0  \n",
       "1        14.821918   16.277778          0       3000         0  \n",
       "2        18.200000   18.909091  248384000  248387328         0  \n",
       "13958     2.000000    2.000000          0       3600         0  \n",
       "13959     2.000000    2.000000  242693800  242696752         0  \n",
       "...            ...         ...        ...        ...       ...  \n",
       "178458   89.884274   84.726368          0       3000         0  \n",
       "178459    2.000000    2.000000          0       4600         0  \n",
       "178460    2.000000    2.000000   51321800   51324926         0  \n",
       "178461  118.426087  117.115406          0       4600         0  \n",
       "178462    7.413285    7.399339   51321800   51324926         0  \n",
       "\n",
       "[83 rows x 273 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_telomeres = final_cnv_ranges.nearest(t2t_telomeres)\n",
    "coordinates_telomeres = coordinates_telomeres.as_df()\n",
    "coordinates_telomeres = coordinates_telomeres[coordinates_telomeres['Distance'] <1 ]\n",
    "coordinates_telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "a11425e8-5b0c-4325-a30c-525a50faf896",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f881004-285e-4ec1-bc41-8466659395b7",
   "metadata": {},
   "source": [
    "<h2> SDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d53bcc03-eab4-4e78-a14e-87f647ae18e4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>140066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>198196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83511</th>\n",
       "      <td>Y</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62453929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83512</th>\n",
       "      <td>Y</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62454747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83513</th>\n",
       "      <td>Y</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62456950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83514</th>\n",
       "      <td>Y</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62457871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83515</th>\n",
       "      <td>Y</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62458697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83516 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+\n",
       "| Chromosome   | Start     | End       |\n",
       "| (category)   | (int64)   | (int64)   |\n",
       "|--------------+-----------+-----------|\n",
       "| 1            | 1         | 13731     |\n",
       "| 1            | 3         | 140066    |\n",
       "| 1            | 16        | 13731     |\n",
       "| 1            | 16        | 13731     |\n",
       "| ...          | ...       | ...       |\n",
       "| Y            | 62445192  | 62454747  |\n",
       "| Y            | 62445192  | 62456950  |\n",
       "| Y            | 62445192  | 62457871  |\n",
       "| Y            | 62445192  | 62458697  |\n",
       "+--------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 83,516 rows and 3 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2t_sds['Chromosome'] = t2t_sds['Chromosome'].str.replace('chr', '')\n",
    "t2t_sds = PyRanges(t2t_sds)\n",
    "t2t_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "73c291a7-43fa-4d91-956e-f6f7063b5751",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_A01</th>\n",
       "      <th>LP6005441-DNA_A03</th>\n",
       "      <th>LP6005441-DNA_A04</th>\n",
       "      <th>LP6005441-DNA_A05</th>\n",
       "      <th>LP6005441-DNA_A06</th>\n",
       "      <th>LP6005441-DNA_A08</th>\n",
       "      <th>...</th>\n",
       "      <th>LP6005677-DNA_G01</th>\n",
       "      <th>LP6007068-DNA_A01</th>\n",
       "      <th>LP6007069-DNA_A01</th>\n",
       "      <th>SS6004471</th>\n",
       "      <th>SS6004477</th>\n",
       "      <th>SS6004478</th>\n",
       "      <th>SS6004480</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>18.080000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>18.446154</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>13731</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>453948</td>\n",
       "      <td>466894</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>459773</td>\n",
       "      <td>460982</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>485339</td>\n",
       "      <td>511499</td>\n",
       "      <td>26160.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>505923</td>\n",
       "      <td>507077</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>518975</td>\n",
       "      <td>605035</td>\n",
       "      <td>86060.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>567686</td>\n",
       "      <td>568750</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>663874</td>\n",
       "      <td>679953</td>\n",
       "      <td>16079.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>676952</td>\n",
       "      <td>677960</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178458</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>110.172885</td>\n",
       "      <td>145.875912</td>\n",
       "      <td>158.807910</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>110.482155</td>\n",
       "      <td>160.080597</td>\n",
       "      <td>...</td>\n",
       "      <td>129.256325</td>\n",
       "      <td>98.383085</td>\n",
       "      <td>95.412666</td>\n",
       "      <td>108.014702</td>\n",
       "      <td>153.940968</td>\n",
       "      <td>89.884274</td>\n",
       "      <td>84.726368</td>\n",
       "      <td>4</td>\n",
       "      <td>293068</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178459</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>52.759434</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>66809</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178460</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>10.165951</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>15712005</td>\n",
       "      <td>15938486</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178461</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>111.294022</td>\n",
       "      <td>146.077591</td>\n",
       "      <td>153.602821</td>\n",
       "      <td>206.970825</td>\n",
       "      <td>120.948192</td>\n",
       "      <td>151.150567</td>\n",
       "      <td>...</td>\n",
       "      <td>147.741988</td>\n",
       "      <td>118.177010</td>\n",
       "      <td>121.240217</td>\n",
       "      <td>142.797625</td>\n",
       "      <td>172.307776</td>\n",
       "      <td>118.426087</td>\n",
       "      <td>117.115406</td>\n",
       "      <td>1</td>\n",
       "      <td>66809</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178462</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>10.664289</td>\n",
       "      <td>9.151965</td>\n",
       "      <td>8.807181</td>\n",
       "      <td>0.970000</td>\n",
       "      <td>9.927382</td>\n",
       "      <td>9.674572</td>\n",
       "      <td>...</td>\n",
       "      <td>10.482103</td>\n",
       "      <td>7.742838</td>\n",
       "      <td>7.805637</td>\n",
       "      <td>8.050934</td>\n",
       "      <td>8.420014</td>\n",
       "      <td>7.413285</td>\n",
       "      <td>7.399339</td>\n",
       "      <td>15712005</td>\n",
       "      <td>15938486</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7946 rows × 273 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End  Length_cnvr  LP6005441-DNA_A01  \\\n",
       "0               1         0    403567     403567.0           2.000000   \n",
       "1               1    453948    466894      12946.0           2.000000   \n",
       "2               1    485339    511499      26160.0           2.000000   \n",
       "3               1    518975    605035      86060.0           2.000000   \n",
       "4               1    663874    679953      16079.0           2.000000   \n",
       "...           ...       ...       ...          ...                ...   \n",
       "178458         21         0  45090682   45090682.0         110.172885   \n",
       "178459         22         0  12764469   12764469.0           2.000000   \n",
       "178460         22  15787747  51324926   35537179.0           2.000000   \n",
       "178461         22         0  12764469   12764469.0         111.294022   \n",
       "178462         22  15787747  51324926   35537179.0          10.664289   \n",
       "\n",
       "        LP6005441-DNA_A03  LP6005441-DNA_A04  LP6005441-DNA_A05  \\\n",
       "0               18.080000           2.000000           2.000000   \n",
       "1                2.000000           2.000000           2.000000   \n",
       "2                3.000000           2.000000           2.000000   \n",
       "3                3.000000           2.000000           2.000000   \n",
       "4                2.000000           2.000000           2.000000   \n",
       "...                   ...                ...                ...   \n",
       "178458         145.875912         158.807910           1.000000   \n",
       "178459           2.000000           2.000000          52.759434   \n",
       "178460           2.000000           2.000000          10.165951   \n",
       "178461         146.077591         153.602821         206.970825   \n",
       "178462           9.151965           8.807181           0.970000   \n",
       "\n",
       "        LP6005441-DNA_A06  LP6005441-DNA_A08  ...  LP6005677-DNA_G01  \\\n",
       "0                2.000000           2.000000  ...          18.446154   \n",
       "1                2.000000           2.000000  ...           3.000000   \n",
       "2                2.000000           2.000000  ...           2.000000   \n",
       "3                2.000000           2.000000  ...           2.000000   \n",
       "4                2.000000           2.000000  ...           3.000000   \n",
       "...                   ...                ...  ...                ...   \n",
       "178458         110.482155         160.080597  ...         129.256325   \n",
       "178459           2.000000           2.000000  ...           2.000000   \n",
       "178460           2.000000           2.000000  ...           2.000000   \n",
       "178461         120.948192         151.150567  ...         147.741988   \n",
       "178462           9.927382           9.674572  ...          10.482103   \n",
       "\n",
       "        LP6007068-DNA_A01  LP6007069-DNA_A01   SS6004471   SS6004477  \\\n",
       "0                2.000000           2.000000    2.000000    2.000000   \n",
       "1                2.000000           2.000000    2.000000    2.000000   \n",
       "2                2.000000           2.000000    2.000000    2.000000   \n",
       "3                2.000000           2.000000    2.000000    2.000000   \n",
       "4                2.000000           2.000000    2.000000    2.000000   \n",
       "...                   ...                ...         ...         ...   \n",
       "178458          98.383085          95.412666  108.014702  153.940968   \n",
       "178459           2.000000           2.000000    2.000000    2.000000   \n",
       "178460           2.000000           2.000000    2.000000    2.000000   \n",
       "178461         118.177010         121.240217  142.797625  172.307776   \n",
       "178462           7.742838           7.805637    8.050934    8.420014   \n",
       "\n",
       "         SS6004478   SS6004480   Start_b     End_b  Distance  \n",
       "0         2.000000    2.000000         1     13731         0  \n",
       "1         2.000000    2.000000    459773    460982         0  \n",
       "2         2.000000    2.000000    505923    507077         0  \n",
       "3         2.000000    2.000000    567686    568750         0  \n",
       "4         2.000000    2.000000    676952    677960         0  \n",
       "...            ...         ...       ...       ...       ...  \n",
       "178458   89.884274   84.726368         4    293068         0  \n",
       "178459    2.000000    2.000000         1     66809         0  \n",
       "178460    2.000000    2.000000  15712005  15938486         0  \n",
       "178461  118.426087  117.115406         1     66809         0  \n",
       "178462    7.413285    7.399339  15712005  15938486         0  \n",
       "\n",
       "[7946 rows x 273 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_sds = final_cnv_ranges.nearest(t2t_sds)\n",
    "coordinates_sds = coordinates_sds.as_df()\n",
    "coordinates_sds = coordinates_sds[coordinates_sds['Distance'] <1 ]\n",
    "coordinates_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "59ecbd80-98d7-4d72-ba2d-f4e1ec4ad0a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4038"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e83226f-52a6-4acb-9c59-59e38d7865b9",
   "metadata": {},
   "source": [
    "T2T Deletions and duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "56f232c4-3bd5-4616-ab49-a8ea1e96d079",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/932598781.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  t2t_deletions.replace(2, np.nan, inplace=True)\n",
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/932598781.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  t2t_deletions.replace(np.nan,2, inplace=True)\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "37823"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2t_deletions = final_cnv[(final_cnv[final_cnv.iloc[:,4:280].columns[1:]]<3).all(axis=1)]\n",
    "t2t_deletions.replace(2, np.nan, inplace=True)\n",
    "t2t_mean_individual_deletions = t2t_deletions.iloc[:,4:280].mean()\n",
    "t2t_deletions.replace(np.nan,2, inplace=True)\n",
    "t2t_deletions = PyRanges(t2t_deletions)\n",
    "len(t2t_deletions)\n",
    "\n",
    "#t2t_mean_individual_deletions.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f897932c-b044-407c-8316-951d97e119bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0870913375262474"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Genes \n",
    "deletions_t2t_coordinates_genes = t2t_deletions.nearest(t2t_genes)\n",
    "deletions_t2t_coordinates_genes = deletions_t2t_coordinates_genes.as_df()\n",
    "deletions_t2t_coordinates_genes = deletions_t2t_coordinates_genes[deletions_t2t_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_mean_individual_genes= deletions_t2t_coordinates_genes.iloc[:,4:270].mean()\n",
    "deletions_t2t_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_t2t_coordinates_telomeres = t2t_deletions.nearest(t2t_telomeres)\n",
    "deletions_t2t_coordinates_telomeres = deletions_t2t_coordinates_telomeres.as_df()\n",
    "deletions_t2t_coordinates_telomeres = deletions_t2t_coordinates_telomeres[deletions_t2t_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_t2t_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_t2t_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_mean_individual_telomeres= deletions_t2t_coordinates_telomeres.iloc[:,4:270].mean()\n",
    "deletions_t2t_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_t2t_coordinates_sds = t2t_deletions.nearest(t2t_sds)\n",
    "deletions_t2t_coordinates_sds = deletions_t2t_coordinates_sds.as_df()\n",
    "deletions_t2t_coordinates_sds = deletions_t2t_coordinates_sds[deletions_t2t_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_mean_individual_sds= deletions_t2t_coordinates_sds.iloc[:,4:270].mean()\n",
    "deletions_t2t_coordinates_sds.replace(np.nan,2, inplace=True)\n",
    "\n",
    "deletions_t2t_mean_individual_sds.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "512d00ac-b7cb-41cd-a739-afc405dc88c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_A01</th>\n",
       "      <th>LP6005441-DNA_A03</th>\n",
       "      <th>LP6005441-DNA_A04</th>\n",
       "      <th>LP6005441-DNA_A05</th>\n",
       "      <th>LP6005441-DNA_A06</th>\n",
       "      <th>LP6005441-DNA_A08</th>\n",
       "      <th>...</th>\n",
       "      <th>LP6005677-DNA_G01</th>\n",
       "      <th>LP6007068-DNA_A01</th>\n",
       "      <th>LP6007069-DNA_A01</th>\n",
       "      <th>SS6004471</th>\n",
       "      <th>SS6004477</th>\n",
       "      <th>SS6004478</th>\n",
       "      <th>SS6004480</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 273 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Start, End, Length_cnvr, LP6005441-DNA_A01, LP6005441-DNA_A03, LP6005441-DNA_A04, LP6005441-DNA_A05, LP6005441-DNA_A06, LP6005441-DNA_A08, LP6005441-DNA_A09, LP6005441-DNA_A10, LP6005441-DNA_A11, LP6005441-DNA_A12, LP6005441-DNA_B01, LP6005441-DNA_B02, LP6005441-DNA_B04, LP6005441-DNA_B05, LP6005441-DNA_B06, LP6005441-DNA_B07, LP6005441-DNA_B08, LP6005441-DNA_B09, LP6005441-DNA_B10, LP6005441-DNA_B11, LP6005441-DNA_B12, LP6005441-DNA_C02, LP6005441-DNA_C03, LP6005441-DNA_C05, LP6005441-DNA_C06, LP6005441-DNA_C07, LP6005441-DNA_C08, LP6005441-DNA_C09, LP6005441-DNA_C10, LP6005441-DNA_C11, LP6005441-DNA_D01, LP6005441-DNA_D02, LP6005441-DNA_D03, LP6005441-DNA_D05, LP6005441-DNA_D06, LP6005441-DNA_D07, LP6005441-DNA_D08, LP6005441-DNA_D09, LP6005441-DNA_D10, LP6005441-DNA_D11, LP6005441-DNA_D12, LP6005441-DNA_E02, LP6005441-DNA_E03, LP6005441-DNA_E05, LP6005441-DNA_E06, LP6005441-DNA_E07, LP6005441-DNA_E08, LP6005441-DNA_E09, LP6005441-DNA_E10, LP6005441-DNA_F01, LP6005441-DNA_F02, LP6005441-DNA_F03, LP6005441-DNA_F04, LP6005441-DNA_F05, LP6005441-DNA_F06, LP6005441-DNA_F07, LP6005441-DNA_F08, LP6005441-DNA_F09, LP6005441-DNA_F10, LP6005441-DNA_F12, LP6005441-DNA_G02, LP6005441-DNA_G03, LP6005441-DNA_G04, LP6005441-DNA_G05, LP6005441-DNA_G06, LP6005441-DNA_G07, LP6005441-DNA_G08, LP6005441-DNA_G10, LP6005441-DNA_G11, LP6005441-DNA_G12, LP6005441-DNA_H02, LP6005441-DNA_H03, LP6005441-DNA_H05, LP6005441-DNA_H06, LP6005441-DNA_H07, LP6005441-DNA_H08, LP6005441-DNA_H09, LP6005441-DNA_H10, LP6005441-DNA_H11, LP6005441-DNA_H12, LP6005442-DNA_A02, LP6005442-DNA_A03, LP6005442-DNA_A04, LP6005442-DNA_A08, LP6005442-DNA_A10, LP6005442-DNA_A11, LP6005442-DNA_A12, LP6005442-DNA_B01, LP6005442-DNA_B02, LP6005442-DNA_B03, LP6005442-DNA_B04, LP6005442-DNA_B08, LP6005442-DNA_B10, LP6005442-DNA_B11, LP6005442-DNA_B12, LP6005442-DNA_C02, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 273 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deletions_t2t_coordinates_telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "bc34c6db-f8f1-4340-8631-21eee0d65167",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/2899501434.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  t2t_duplications.replace(2, np.nan, inplace=True)\n",
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/2899501434.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  t2t_duplications.replace(np.nan,2, inplace=True)\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6.014577417095775"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplications \n",
    "t2t_duplications = final_cnv[(final_cnv[final_cnv.iloc[:,4:280].columns[1:]]>1).all(axis=1)]\n",
    "t2t_duplications.replace(2, np.nan, inplace=True)\n",
    "t2t_mean_individual_duplications = t2t_duplications.iloc[:,4:280].mean()\n",
    "t2t_duplications.replace(np.nan,2, inplace=True)\n",
    "t2t_duplications  = PyRanges(t2t_duplications)\n",
    "len(t2t_duplications)\n",
    "\n",
    "\n",
    "t2t_mean_individual_duplications.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "8cd95add-cd52-4cf4-9bde-c39d3efd24cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.396111753350639"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Genes \n",
    "duplications_t2t_coordinates_genes = t2t_duplications.nearest(t2t_genes)\n",
    "duplications_t2t_coordinates_genes = duplications_t2t_coordinates_genes.as_df()\n",
    "duplications_t2t_coordinates_genes = duplications_t2t_coordinates_genes[duplications_t2t_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_mean_individual_genes= duplications_t2t_coordinates_genes.iloc[:,4:270].mean()\n",
    "duplications_t2t_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_t2t_coordinates_telomeres = t2t_duplications.nearest(t2t_telomeres)\n",
    "duplications_t2t_coordinates_telomeres = duplications_t2t_coordinates_telomeres.as_df()\n",
    "duplications_t2t_coordinates_telomeres = duplications_t2t_coordinates_telomeres[duplications_t2t_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_t2t_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_t2t_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_mean_individual_telomeres= duplications_t2t_coordinates_telomeres.iloc[:,4:270].mean()\n",
    "duplications_t2t_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_t2t_coordinates_sds = t2t_duplications.nearest(t2t_sds)\n",
    "duplications_t2t_coordinates_sds = duplications_t2t_coordinates_sds.as_df()\n",
    "duplications_t2t_coordinates_sds = duplications_t2t_coordinates_sds[duplications_t2t_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_mean_individual_sds= duplications_t2t_coordinates_sds.iloc[:,4:270].mean()\n",
    "duplications_t2t_coordinates_sds.replace(np.nan,2, inplace=True)\n",
    "\n",
    "duplications_t2t_mean_individual_sds.mean()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "14965f34-912f-4ea4-8463-27bde0b6e1bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4038"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7ffae1-94ba-45c0-95d7-e4426e4b28ae",
   "metadata": {},
   "source": [
    "<h1> GRCh37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "21bf4b13-6a3b-4874-bebc-d5d9f6eca33d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Anotations\n",
    "grch37_genes = pd.read_csv('../annotation_files/grch37-genes.csv')\n",
    "grch37_telomeres = pd.read_csv('../annotation_files/grch37-telomers-centromers.csv')\n",
    "grch37_sds = pd.read_csv('../annotation_files/grch37-sds.csv')\n",
    "#grch37_sds = grch37_sds[['#chr1', 'start1', 'end1']]\n",
    "\n",
    "\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "\n",
    "\n",
    "#t2t_sds.columns =['Chromosome','Start','End']\n",
    "#t2t_sds\n",
    "grch37_genes.columns = ['#\"bin\"', 'name', 'Chromosome', 'strand', 'Start', 'End', 'cdsStart',\n",
    "       'cdsEnd', 'exonCount', 'exonStarts', 'exonEnds', 'score', 'name2',\n",
    "       'cdsStartStat', 'cdsEndStat', 'exonFrames'] \n",
    "grch37_telomeres = grch37_telomeres[grch37_telomeres['type'] == 'telomere']\n",
    "grch37_sds.columns = ['#\"bin\"', 'Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'otherChrom', 'otherStart', 'otherEnd', 'otherSize', 'uid',\n",
    "       'posBasesHit', 'testResult', 'verdict', 'chits', 'ccov', 'alignfile',\n",
    "       'alignL', 'indelN', 'indelS', 'alignB', 'matchB', 'mismatchB',\n",
    "       'transitionsB', 'transversionsB', 'fracMatch', 'fracMatchIndel', 'jcK',\n",
    "       'k2K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "92a4acd0-37cc-4be4-a1ce-1ac6fba433d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(\"../call_cnvr_cnvnator/individual_cnv_cnvr.txt\", sep='\\t')\n",
    "data.columns = ['Chromosome', 'Start', 'End','CNVR','Sample_ID','CNV_Start', 'CNV_End', 'RD', 'Lenght']\n",
    "data = data.dropna()#.describe()\n",
    "data['Length_cnvr'] = data['End'] - data['Start'] \n",
    "#data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "236ada20-700c-4d5d-8fc7-4e3450891fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Keeping only samples that are on old cnvs \n",
    "data = data.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "34fafde0-5e38-4355-9137-024a0d940ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#REMOVE SAMPLES WITH ABERRATIONS\n",
    "samples_with_aberrations = ['LP6005441-DNA_B03', ' LP6005443-DNA_G01', 'LP6005441-DNA_G09', 'LP6005441-DNA_D04'] \n",
    "\n",
    "data = data[~data.Sample.isin(samples_with_aberrations)]\n",
    "\n",
    "data.columns = ['Chromosome', 'Start', 'End', 'CNVR_ID', 'Sample_iD', 'CNV_start', 'CNV_end', 'CNV_Value','Length', 'Length_cnvr','Sample','Region']\n",
    "#data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f5ab003b-6d04-431b-ac90-0be08e8f2049",
   "metadata": {},
   "outputs": [],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv_grch37 = data.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='CNV_Value',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "\n",
    "#final_cnv_grch37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "3200a16b-2464-4e47-8063-9e46662ba8f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "final_cnv_ranges_grch37 = PyRanges(final_cnv_grch37)\n",
    "#final_cnv_grch37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f4f2cae9-070e-497d-8d8e-57b59af3beb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/819350350.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  grch37_deletions.replace(2, np.nan, inplace=True)\n",
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/819350350.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  grch37_deletions.replace(np.nan,2, inplace=True)\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6110"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Deletions \n",
    "grch37_deletions = final_cnv_grch37[(final_cnv_grch37[final_cnv_grch37.iloc[:,4:280].columns[1:]]<3).all(axis=1)]\n",
    "grch37_deletions.replace(2, np.nan, inplace=True)\n",
    "grch37_mean_individual_deletions = grch37_deletions.iloc[:,4:280].mean()\n",
    "grch37_deletions.replace(np.nan,2, inplace=True)\n",
    "grch37_deletions = PyRanges(grch37_deletions)\n",
    "len(grch37_deletions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "12785261-b753-4a1c-9662-c7c68a3d2f3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/1735373779.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  grch37_duplications.replace(2, np.nan, inplace=True)\n",
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_8332/1735373779.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  grch37_duplications.replace(np.nan,2, inplace=True)\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4.85052458491179"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplications \n",
    "grch37_duplications = final_cnv_grch37[(final_cnv_grch37[final_cnv_grch37.iloc[:,4:280].columns[1:]]>1).all(axis=1)]\n",
    "grch37_duplications.replace(2, np.nan, inplace=True)\n",
    "grch37_mean_individual_duplications = grch37_duplications.iloc[:,4:280].mean()\n",
    "grch37_duplications.replace(np.nan,2, inplace=True)\n",
    "\n",
    "grch37_duplications  = PyRanges(grch37_duplications)\n",
    "grch37_mean_individual_duplications.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f006e6db-c6d3-4dea-9126-9f0ce4e30ed2",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "8e734c44-9ebc-4144-95a1-58803104e4fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "grch37_genes['Chromosome'] = grch37_genes['Chromosome'].str.replace('chr', '')\n",
    "grch37_genes_coordinates = PyRanges(grch37_genes)\n",
    "#grch37_genes_coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b3c811e9-1372-4e7f-9742-02d296c392b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "grch37_coordinates_genes = final_cnv_ranges_grch37.nearest(grch37_genes_coordinates)\n",
    "grch37_coordinates_genes = grch37_coordinates_genes.as_df()\n",
    "grch37_coordinates_genes = grch37_coordinates_genes[grch37_coordinates_genes['Distance'] <1 ]\n",
    "#grch37_coordinates_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "d95bff56-ba99-456e-8d98-67bbd853687f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5427"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_genes = len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "grch37_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55c3b568-3b07-4850-8ec0-a81222f3d5d3",
   "metadata": {},
   "source": [
    "<h1> Telomeres "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "2922fba9-b7e7-48eb-b4c2-c2ed273c1bb0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "grch37_telomeres['Chromosome'] = grch37_telomeres['Chromosome'].str.replace('chr', '')\n",
    "grch37_telomeres = PyRanges(grch37_telomeres)\n",
    "#grch37_telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "edc9e0f3-a3bd-44c8-bf25-7f00562bb24c",
   "metadata": {},
   "outputs": [],
   "source": [
    "grch37_coordinates_telomeres = final_cnv_ranges_grch37.nearest(grch37_telomeres)\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres.as_df()\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres[grch37_coordinates_telomeres['Distance'] <1 ]\n",
    "#grch37_coordinates_telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "2c9ff212-2464-4d29-aa44-90a2ca72d3b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_telomeres_coordinates = len(grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "grch37_telomeres_coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f02f98a8-f339-48e5-856f-b3e018c85e78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "230539.0"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_coordinates_telomeres['Length_cnvr'].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9613889b-ed82-4667-b50b-1c52765edfdb",
   "metadata": {},
   "source": [
    "<h1> SDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "5c1a61ad-684d-422e-b3ef-497c2156f91c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "grch37_sds['Chromosome'] = grch37_sds['Chromosome'].str.replace('chr', '')\n",
    "grch37_sds = PyRanges(grch37_sds)\n",
    "#grch37_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "06ba10f4-0888-4372-b31f-256c71ff793c",
   "metadata": {},
   "outputs": [],
   "source": [
    "grch37_coordinates_sds = final_cnv_ranges_grch37.nearest(grch37_sds)\n",
    "grch37_coordinates_sds = grch37_coordinates_sds.as_df()\n",
    "grch37_coordinates_sds = grch37_coordinates_sds[grch37_coordinates_sds['Distance'] <1 ]\n",
    "#grch37_coordinates_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "cf3b31d8-07cb-4c2b-bc22-66e26786d3f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1151"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_sds_coordinates = len(grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "grch37_sds_coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb640484-cae4-46aa-bd3c-bd4e249db832",
   "metadata": {},
   "source": [
    "GRCh37 Deletions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "a64eefc5-41c8-45f7-b64b-d46d3aa43daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Genes \n",
    "deletions_grch37_coordinates_genes = grch37_deletions.nearest(grch37_genes_coordinates)\n",
    "deletions_grch37_coordinates_genes = deletions_grch37_coordinates_genes.as_df()\n",
    "deletions_grch37_coordinates_genes = deletions_grch37_coordinates_genes[deletions_grch37_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_mean_individual_genes= deletions_grch37_coordinates_genes.iloc[:,4:270].mean()\n",
    "deletions_grch37_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "### Telomeres\n",
    "deletions_grch37_coordinates_telomeres = grch37_deletions.nearest(grch37_telomeres)\n",
    "deletions_grch37_coordinates_telomeres = deletions_grch37_coordinates_telomeres.as_df()\n",
    "deletions_grch37_coordinates_telomeres = deletions_grch37_coordinates_telomeres[deletions_grch37_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_grch37_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_grch37_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_mean_individual_telomeres= deletions_grch37_coordinates_telomeres.iloc[:,4:270].mean()\n",
    "deletions_grch37_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_grch37_coordinates_sds = grch37_deletions.nearest(grch37_sds)\n",
    "deletions_grch37_coordinates_sds = deletions_grch37_coordinates_sds.as_df()\n",
    "deletions_grch37_coordinates_sds = deletions_grch37_coordinates_sds[deletions_grch37_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_mean_individual_sds= deletions_grch37_coordinates_sds.iloc[:,4:270].mean()\n",
    "deletions_grch37_coordinates_sds.replace(np.nan,2, inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90789782-536a-4f10-b4b5-60e5d62a4cee",
   "metadata": {},
   "source": [
    "GRCh37 Duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "3b1e5f53-2adc-4cd2-b32f-afd8c7321ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Genes \n",
    "duplications_grch37_coordinates_genes = grch37_duplications.nearest(grch37_genes_coordinates)\n",
    "duplications_grch37_coordinates_genes = duplications_grch37_coordinates_genes.as_df()\n",
    "duplications_grch37_coordinates_genes = duplications_grch37_coordinates_genes[duplications_grch37_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_mean_individual_genes= duplications_grch37_coordinates_genes.iloc[:,4:270].mean()\n",
    "duplications_grch37_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_grch37_coordinates_telomeres = grch37_duplications.nearest(grch37_telomeres)\n",
    "duplications_grch37_coordinates_telomeres = duplications_grch37_coordinates_telomeres.as_df()\n",
    "duplications_grch37_coordinates_telomeres = duplications_grch37_coordinates_telomeres[duplications_grch37_coordinates_telomeres['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_mean_individual_telomeres= duplications_grch37_coordinates_telomeres.iloc[:,4:270].mean()\n",
    "duplications_grch37_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_grch37_coordinates_sds = grch37_duplications.nearest(grch37_sds)\n",
    "duplications_grch37_coordinates_sds = duplications_grch37_coordinates_sds.as_df()\n",
    "duplications_grch37_coordinates_sds = duplications_grch37_coordinates_sds[duplications_grch37_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_mean_individual_sds= duplications_grch37_coordinates_sds.iloc[:,4:270].mean()\n",
    "duplications_grch37_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "6365e24d-79fe-4c7d-92b7-9e4ab6dc3ddc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Total</th>\n",
       "      <th>Mean size(Bp)</th>\n",
       "      <th>Deletions</th>\n",
       "      <th>Duplications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7780</td>\n",
       "      <td>66614.0</td>\n",
       "      <td>6110</td>\n",
       "      <td>1482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>5427</td>\n",
       "      <td>85647.0</td>\n",
       "      <td>4174</td>\n",
       "      <td>1123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>10</td>\n",
       "      <td>230539.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>1151</td>\n",
       "      <td>210065.0</td>\n",
       "      <td>730</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>178463</td>\n",
       "      <td>24871.0</td>\n",
       "      <td>37823</td>\n",
       "      <td>81920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>84326</td>\n",
       "      <td>27200.0</td>\n",
       "      <td>23203</td>\n",
       "      <td>55258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>42</td>\n",
       "      <td>5842973.0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>4038</td>\n",
       "      <td>171021.0</td>\n",
       "      <td>144</td>\n",
       "      <td>3497</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data   Total  Mean size(Bp)  Deletions  Duplications\n",
       "0       GRCh37    7780        66614.0       6110          1482\n",
       "1        Genes    5427        85647.0       4174          1123\n",
       "2    Telomeres      10       230539.0          4             0\n",
       "3          SDs    1151       210065.0        730           282\n",
       "4    T2T-CHM13  178463        24871.0      37823         81920\n",
       "5        Genes   84326        27200.0      23203         55258\n",
       "6    Telomeres      42      5842973.0          0            41\n",
       "7          SDs    4038       171021.0        144          3497"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'], \n",
    "     'Total': [len(final_cnv_grch37), \n",
    "               len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),  \n",
    "               len(final_cnv), \n",
    "               len(coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())], \n",
    "     'Mean size(Bp)':[final_cnv_grch37['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_genes['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_sds['Length_cnvr'].mean().round(), \n",
    "                      final_cnv['Length_cnvr'].mean().round(),\n",
    "                      coordinates_genes['Length_cnvr'].mean().round(),\n",
    "                      coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      coordinates_sds['Length_cnvr'].mean().round()],\n",
    "     'Deletions':[len(grch37_deletions), \n",
    "                  len(deletions_grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(deletions_grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(deletions_grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(t2t_deletions), \n",
    "                  len(deletions_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(deletions_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(deletions_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), ], \n",
    "     'Duplications':[len(grch37_duplications), \n",
    "                     len(duplications_grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(duplications_grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                     len(duplications_grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(t2t_duplications), \n",
    "                     len(duplications_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(duplications_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                     len(duplications_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()) ]\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=d)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ecae1534-4c99-471d-b7e3-323d1a8e7561",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Total</th>\n",
       "      <th>Mean size(Bp)</th>\n",
       "      <th>Deletions</th>\n",
       "      <th>Duplications</th>\n",
       "      <th>Mean CN Deletions</th>\n",
       "      <th>Mean CN Duplications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7780</td>\n",
       "      <td>66614.0</td>\n",
       "      <td>6110</td>\n",
       "      <td>1482</td>\n",
       "      <td>0.408507</td>\n",
       "      <td>4.850525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>5427</td>\n",
       "      <td>85647.0</td>\n",
       "      <td>4174</td>\n",
       "      <td>1123</td>\n",
       "      <td>0.426343</td>\n",
       "      <td>4.125167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>10</td>\n",
       "      <td>230539.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>1151</td>\n",
       "      <td>210065.0</td>\n",
       "      <td>730</td>\n",
       "      <td>282</td>\n",
       "      <td>0.469472</td>\n",
       "      <td>5.018248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>178463</td>\n",
       "      <td>24871.0</td>\n",
       "      <td>37823</td>\n",
       "      <td>81920</td>\n",
       "      <td>0.996774</td>\n",
       "      <td>6.014577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>84326</td>\n",
       "      <td>27200.0</td>\n",
       "      <td>23203</td>\n",
       "      <td>55258</td>\n",
       "      <td>0.997449</td>\n",
       "      <td>6.087164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>42</td>\n",
       "      <td>5842973.0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.657647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>4038</td>\n",
       "      <td>171021.0</td>\n",
       "      <td>144</td>\n",
       "      <td>3497</td>\n",
       "      <td>1.087091</td>\n",
       "      <td>7.396112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data   Total  Mean size(Bp)  Deletions  Duplications  \\\n",
       "0       GRCh37    7780        66614.0       6110          1482   \n",
       "1        Genes    5427        85647.0       4174          1123   \n",
       "2    Telomeres      10       230539.0          4             0   \n",
       "3          SDs    1151       210065.0        730           282   \n",
       "4    T2T-CHM13  178463        24871.0      37823         81920   \n",
       "5        Genes   84326        27200.0      23203         55258   \n",
       "6    Telomeres      42      5842973.0          0            41   \n",
       "7          SDs    4038       171021.0        144          3497   \n",
       "\n",
       "   Mean CN Deletions  Mean CN Duplications  \n",
       "0           0.408507              4.850525  \n",
       "1           0.426343              4.125167  \n",
       "2           0.625000                   NaN  \n",
       "3           0.469472              5.018248  \n",
       "4           0.996774              6.014577  \n",
       "5           0.997449              6.087164  \n",
       "6                NaN             39.657647  \n",
       "7           1.087091              7.396112  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'], \n",
    "     'Total': [len(final_cnv_grch37), \n",
    "               len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),  \n",
    "               len(final_cnv), \n",
    "               len(coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())], \n",
    "     'Mean size(Bp)':[final_cnv_grch37['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_genes['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_sds['Length_cnvr'].mean().round(), \n",
    "                      final_cnv['Length_cnvr'].mean().round(),\n",
    "                      coordinates_genes['Length_cnvr'].mean().round(),\n",
    "                      coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      coordinates_sds['Length_cnvr'].mean().round()],\n",
    "     'Deletions':[len(grch37_deletions), \n",
    "                  len(deletions_grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(deletions_grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(deletions_grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(t2t_deletions), \n",
    "                  len(deletions_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(deletions_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(deletions_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), ], \n",
    "     'Duplications':[len(grch37_duplications), \n",
    "                     len(duplications_grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(duplications_grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                     len(duplications_grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(t2t_duplications), \n",
    "                     len(duplications_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(duplications_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                     len(duplications_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()) ],\n",
    "     'Mean CN Deletions' :[grch37_mean_individual_deletions.mean(),\n",
    "                           deletions_grch37_mean_individual_genes.mean(),\n",
    "                           deletions_grch37_mean_individual_telomeres.mean(),\n",
    "                           deletions_grch37_mean_individual_sds.mean(),\n",
    "                           t2t_mean_individual_deletions.mean(),\n",
    "                           deletions_t2t_mean_individual_genes.mean(),\n",
    "                           deletions_t2t_mean_individual_telomeres.mean(),\n",
    "                           deletions_t2t_mean_individual_sds.mean()],\n",
    "     'Mean CN Duplications' :[grch37_mean_individual_duplications.mean(),\n",
    "                           duplications_grch37_mean_individual_genes.mean(),\n",
    "                           duplications_grch37_mean_individual_telomeres.mean(),\n",
    "                           duplications_grch37_mean_individual_sds.mean(),\n",
    "                           t2t_mean_individual_duplications.mean(),\n",
    "                           duplications_t2t_mean_individual_genes.mean(),\n",
    "                           duplications_t2t_mean_individual_telomeres.mean(),\n",
    "                           duplications_t2t_mean_individual_sds.mean()],\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=d)\n",
    "df#.to_csv('../data/table1_cnv_calls_between_assemblies.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "272eef6c-7ec2-444d-9396-1f972933c73a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     4.850525\n",
       "1     4.125167\n",
       "2          NaN\n",
       "3     5.018248\n",
       "4     6.014577\n",
       "5     6.087164\n",
       "6    39.657647\n",
       "7     7.396112\n",
       "Name: Mean CN Duplications, dtype: float64"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Mean CN Duplications']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be934881-8a16-42b7-aee7-9f936e845cf2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2899651b-9ed8-49c8-bf2a-06cac346ce5b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
