{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00f332d4-1c1c-42c7-8382-7cb5fcab05d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyranges import PyRanges\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b272d5b-706a-4c50-9f84-cc3e31f479e3",
   "metadata": {},
   "source": [
    "<h1> T2T CHM13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "6b0442f5-74b7-4d26-9bc8-dab9a2b8ff71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_91372/1889853910.py:2: DtypeWarning: Columns (23,36) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  t2t_genes = pd.read_csv('../annotation_files/T2T-chm13-catLiftOffGenes.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3</td>\n",
       "      <td>140066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16</td>\n",
       "      <td>198196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83511</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62453929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83512</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62454747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83513</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62456950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83514</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62457871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83515</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62458697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83516 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End\n",
       "0           chr1         1     13731\n",
       "1           chr1         3    140066\n",
       "2           chr1        16     13731\n",
       "3           chr1        16     13731\n",
       "4           chr1        16    198196\n",
       "...          ...       ...       ...\n",
       "83511       chrY  62445192  62453929\n",
       "83512       chrY  62445192  62454747\n",
       "83513       chrY  62445192  62456950\n",
       "83514       chrY  62445192  62457871\n",
       "83515       chrY  62445192  62458697\n",
       "\n",
       "[83516 rows x 3 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Anotations\n",
    "t2t_genes = pd.read_csv('../annotation_files/T2T-chm13-catLiftOffGenes.csv')\n",
    "t2t_telomeres = pd.read_csv('../annotation_files/chm13v2.0_telomere.bed', sep='\\t', names=['Chromosome','Start', 'End'])\n",
    "t2t_sds = pd.read_csv('../annotation_files/chm13v2.0_SD.full.bed', sep='\\t')\n",
    "t2t_sds = t2t_sds[['#chr1', 'start1', 'end1']]\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "t2t_sds.columns =['Chromosome','Start','End']\n",
    "t2t_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "d730b68f-174e-41e2-ac7a-b23e1aa2e6eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>ext1</th>\n",
       "      <th>ext2</th>\n",
       "      <th>ext3</th>\n",
       "      <th>ext4</th>\n",
       "      <th>ext5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>9083</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0,255,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9083</td>\n",
       "      <td>10803</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32,178,170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10803</td>\n",
       "      <td>14487</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0,255,255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14487</td>\n",
       "      <td>18731</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>255,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18731</td>\n",
       "      <td>20013</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>255,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179222</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51301662</td>\n",
       "      <td>51302662</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32,178,170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179223</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51302662</td>\n",
       "      <td>51303662</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32,178,170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179224</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51303662</td>\n",
       "      <td>51306044</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32,178,170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179225</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51306044</td>\n",
       "      <td>51312584</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0,255,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179226</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51312584</td>\n",
       "      <td>51324926</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0,255,255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1179227 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Start       End  CNV_Value  ext1 ext2  ext3  ext4  \\\n",
       "0             chr1         0      9083         44     0    +     0     0   \n",
       "1             chr1      9083     10803         34     0    +     0     0   \n",
       "2             chr1     10803     14487         20     0    +     0     0   \n",
       "3             chr1     14487     18731         14     0    +     0     0   \n",
       "4             chr1     18731     20013         14     0    +     0     0   \n",
       "...            ...       ...       ...        ...   ...  ...   ...   ...   \n",
       "1179222      chr22  51301662  51302662         34     0    +     0     0   \n",
       "1179223      chr22  51302662  51303662         34     0    +     0     0   \n",
       "1179224      chr22  51303662  51306044         36     0    +     0     0   \n",
       "1179225      chr22  51306044  51312584         45     0    +     0     0   \n",
       "1179226      chr22  51312584  51324926         25     0    +     0     0   \n",
       "\n",
       "               ext5  \n",
       "0           0,255,0  \n",
       "1        32,178,170  \n",
       "2         0,255,255  \n",
       "3           255,0,0  \n",
       "4           255,0,0  \n",
       "...             ...  \n",
       "1179222  32,178,170  \n",
       "1179223  32,178,170  \n",
       "1179224  32,178,170  \n",
       "1179225     0,255,0  \n",
       "1179226   0,255,255  \n",
       "\n",
       "[1179227 rows x 9 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# is not working \n",
    "#chm13 = pd.read_csv('../call_cnvr_chm13/individual_cnv_cnvr.txt', sep='\\t')\n",
    "chm13 = pd.read_csv('/Users/luciabazan/Downloads/AMR_Mayan_LP6005441-DNA_G07', sep='\\t', header= None)\n",
    "chm13.columns = ['Chromosome', 'Start', 'End','CNV_Value','ext1','ext2','ext3','ext4','ext5']\n",
    "chm13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "55cc3484-958c-4a32-94f5-7b0e88698454",
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13 = chm13.dropna()\n",
    "chm13['Length_cnvr'] = chm13['End'] - chm13['Start']\n",
    "\n",
    "\n",
    "#chm13 = chm13[chm13['Sample_ID'] == 'LP6005441-DNA_G07']\n",
    "chm13['Sample_ID'] = 'LP6005441-DNA_G07'\n",
    "chm13 = chm13[~chm13['CNV_Value'].isin([2])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "55b427d6-ec1f-4c01-8621-426f5f660809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>9083</td>\n",
       "      <td>44</td>\n",
       "      <td>9083</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9083</td>\n",
       "      <td>10803</td>\n",
       "      <td>34</td>\n",
       "      <td>1720</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10803</td>\n",
       "      <td>14487</td>\n",
       "      <td>20</td>\n",
       "      <td>3684</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14487</td>\n",
       "      <td>18731</td>\n",
       "      <td>14</td>\n",
       "      <td>4244</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18731</td>\n",
       "      <td>20013</td>\n",
       "      <td>14</td>\n",
       "      <td>1282</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103808</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51301662</td>\n",
       "      <td>51302662</td>\n",
       "      <td>34</td>\n",
       "      <td>1000</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103809</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51302662</td>\n",
       "      <td>51303662</td>\n",
       "      <td>34</td>\n",
       "      <td>1000</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103810</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51303662</td>\n",
       "      <td>51306044</td>\n",
       "      <td>36</td>\n",
       "      <td>2382</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103811</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51306044</td>\n",
       "      <td>51312584</td>\n",
       "      <td>45</td>\n",
       "      <td>6540</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103812</th>\n",
       "      <td>chr22</td>\n",
       "      <td>51312584</td>\n",
       "      <td>51324926</td>\n",
       "      <td>25</td>\n",
       "      <td>12342</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103813 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Start       End  CNV_Value  Length_cnvr  \\\n",
       "0            chr1         0      9083         44         9083   \n",
       "1            chr1      9083     10803         34         1720   \n",
       "2            chr1     10803     14487         20         3684   \n",
       "3            chr1     14487     18731         14         4244   \n",
       "4            chr1     18731     20013         14         1282   \n",
       "...           ...       ...       ...        ...          ...   \n",
       "103808      chr22  51301662  51302662         34         1000   \n",
       "103809      chr22  51302662  51303662         34         1000   \n",
       "103810      chr22  51303662  51306044         36         2382   \n",
       "103811      chr22  51306044  51312584         45         6540   \n",
       "103812      chr22  51312584  51324926         25        12342   \n",
       "\n",
       "                Sample_ID             Sample   Region  \n",
       "0       LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "1       LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "2       LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "3       LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "4       LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "...                   ...                ...      ...  \n",
       "103808  LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "103809  LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "103810  LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "103811  LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "103812  LP6005441-DNA_G07  LP6005441-DNA_G07  America  \n",
       "\n",
       "[103813 rows x 8 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Keeping only samples that are on old cnvs \n",
    "chm13 = chm13.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])\n",
    "chm13.drop(columns=['ext1','ext2','ext3','ext4','ext5'], inplace=True)\n",
    "chm13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "ac0ab3f0-db9c-4b57-a3aa-06fdecb87f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "#REMOVE SAMPLES WITH ABERRATIONS\n",
    "\n",
    "chm13#.value_counts(by=[\"Sample_ID\"])\n",
    "#chm13.columns = ['Chromosome', 'Start', 'End', 'Sample_ID', 'CNV_start', 'CNV_end', 'Length', 'Length_cnvr','Sample','Region']\n",
    "\n",
    "chm13= chm13[chm13['Chromosome'] != 'chrM']\n",
    "chm13 = chm13[chm13['Chromosome'] != 'chrX']\n",
    "chm13 = chm13[chm13['Chromosome'] != 'chrY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "dca9507e-4cd0-46d6-811b-db80775baa27",
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13['Chromosome'] = chm13['Chromosome'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "a4ff43d7-b75b-406c-8601-e49e532c98b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9083</td>\n",
       "      <td>9083</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9083</td>\n",
       "      <td>10803</td>\n",
       "      <td>1720</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10803</td>\n",
       "      <td>14487</td>\n",
       "      <td>3684</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14487</td>\n",
       "      <td>18731</td>\n",
       "      <td>4244</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>18731</td>\n",
       "      <td>20013</td>\n",
       "      <td>1282</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92829</th>\n",
       "      <td>9</td>\n",
       "      <td>150593359</td>\n",
       "      <td>150597012</td>\n",
       "      <td>3653</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92830</th>\n",
       "      <td>9</td>\n",
       "      <td>150597012</td>\n",
       "      <td>150599679</td>\n",
       "      <td>2667</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92831</th>\n",
       "      <td>9</td>\n",
       "      <td>150599679</td>\n",
       "      <td>150601193</td>\n",
       "      <td>1514</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92832</th>\n",
       "      <td>9</td>\n",
       "      <td>150601193</td>\n",
       "      <td>150609226</td>\n",
       "      <td>8033</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92833</th>\n",
       "      <td>9</td>\n",
       "      <td>150609226</td>\n",
       "      <td>150617247</td>\n",
       "      <td>8021</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92834 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome      Start        End  Length_cnvr  LP6005441-DNA_G07\n",
       "0               1          0       9083         9083               44.0\n",
       "1               1       9083      10803         1720               34.0\n",
       "2               1      10803      14487         3684               20.0\n",
       "3               1      14487      18731         4244               14.0\n",
       "4               1      18731      20013         1282               14.0\n",
       "...           ...        ...        ...          ...                ...\n",
       "92829           9  150593359  150597012         3653               22.0\n",
       "92830           9  150597012  150599679         2667               26.0\n",
       "92831           9  150599679  150601193         1514               45.0\n",
       "92832           9  150601193  150609226         8033               39.0\n",
       "92833           9  150609226  150617247         8021               26.0\n",
       "\n",
       "[92834 rows x 5 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv = chm13.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='CNV_Value',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "final_cnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "8ee60b12-5353-4636-9d8f-bab3747f6a83",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "t2t_deletions_gb = final_cnv[final_cnv['LP6005441-DNA_G07'] < 3]\n",
    "t2t_deletions_gb_mean = t2t_deletions_gb['LP6005441-DNA_G07'].mean()\n",
    "t2t_duplications_gb = final_cnv[final_cnv['LP6005441-DNA_G07'] > 1]\n",
    "t2t_duplications_gb_mean = t2t_duplications_gb['LP6005441-DNA_G07'].mean()\n",
    "t2t_duplications_gb = PyRanges(t2t_duplications_gb)\n",
    "t2t_deletions_gb = PyRanges(t2t_deletions_gb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "c3dd5fce-15e1-4360-ab58-17d628a16655",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "final_cnv_ranges = PyRanges(final_cnv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "1ebf6094-c3e2-453d-b1f9-3b733adfb293",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2t_genes.columns = ['Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'thickStart', 'thickEnd', 'reserved', 'blockCount', 'blockSizes',\n",
    "       'chromStarts', 'name2', 'cdsStartStat', 'cdsEndStat', 'exonFrames',\n",
    "       'txId', 'type', 'geneName', 'geneType', 'sourceGene',\n",
    "       'sourceTranscript', 'alignmentId', 'alternativeSourceTranscripts',\n",
    "       'Paralogy', 'UnfilteredParalogy', 'collapsedGeneIds',\n",
    "       'collapsedGeneNames', 'frameshift', 'exonAnnotationSupport',\n",
    "       'intronAnnotationSupport', 'transcriptClass', 'transcriptModes',\n",
    "       'validStart', 'validStop', 'properOrf', 'extra_paralog']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "7ab381cf-130e-4900-b541-f7756d20c889",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "t2t_genes['Chromosome'] = t2t_genes['Chromosome'].str.replace('chr', '')\n",
    "t2t_genes = PyRanges(t2t_genes)\n",
    "\n",
    "t2t_telomeres['Chromosome'] = t2t_telomeres['Chromosome'].str.replace('chr', '')\n",
    "t2t_telomeres = PyRanges(t2t_telomeres)\n",
    "\n",
    "t2t_sds['Chromosome'] = t2t_sds['Chromosome'].str.replace('chr', '')\n",
    "t2t_sds = PyRanges(t2t_sds)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "270bf858-3acc-485c-90d3-88cb319bd888",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "6679ce82-0b3c-45fd-97d4-3fced891d9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### t2t genes \n",
    "coordinates_genes = final_cnv_ranges.nearest(t2t_genes)\n",
    "coordinates_genes = coordinates_genes.as_df()\n",
    "coordinates_genes = coordinates_genes[coordinates_genes['Distance'] <1 ]\n",
    "\n",
    "## T2T telomeres\n",
    "coordinates_telomeres = final_cnv_ranges.nearest(t2t_telomeres)\n",
    "coordinates_telomeres = coordinates_telomeres.as_df()\n",
    "coordinates_telomeres = coordinates_telomeres[coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs \n",
    "coordinates_sds = final_cnv_ranges.nearest(t2t_sds)\n",
    "coordinates_sds = coordinates_sds.as_df()\n",
    "coordinates_sds = coordinates_sds[coordinates_sds['Distance'] <1 ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "975af7c8-8fdf-4a44-934a-68192a6ad882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>frameshift</th>\n",
       "      <th>exonAnnotationSupport</th>\n",
       "      <th>intronAnnotationSupport</th>\n",
       "      <th>transcriptClass</th>\n",
       "      <th>transcriptModes</th>\n",
       "      <th>validStart</th>\n",
       "      <th>validStop</th>\n",
       "      <th>properOrf</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10803</td>\n",
       "      <td>14487</td>\n",
       "      <td>3684</td>\n",
       "      <td>20.0</td>\n",
       "      <td>11135</td>\n",
       "      <td>12457</td>\n",
       "      <td>AL627309.3-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14487</td>\n",
       "      <td>18731</td>\n",
       "      <td>4244</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14252</td>\n",
       "      <td>21099</td>\n",
       "      <td>AC114498.1-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>18731</td>\n",
       "      <td>20013</td>\n",
       "      <td>1282</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14252</td>\n",
       "      <td>21099</td>\n",
       "      <td>AC114498.1-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>20013</td>\n",
       "      <td>28884</td>\n",
       "      <td>8871</td>\n",
       "      <td>13.0</td>\n",
       "      <td>14252</td>\n",
       "      <td>21099</td>\n",
       "      <td>AC114498.1-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>51814</td>\n",
       "      <td>54076</td>\n",
       "      <td>2262</td>\n",
       "      <td>16.0</td>\n",
       "      <td>52975</td>\n",
       "      <td>54613</td>\n",
       "      <td>AL731661.1-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90914</th>\n",
       "      <td>22</td>\n",
       "      <td>51301662</td>\n",
       "      <td>51302662</td>\n",
       "      <td>1000</td>\n",
       "      <td>34.0</td>\n",
       "      <td>51270445</td>\n",
       "      <td>51302683</td>\n",
       "      <td>RPL23AP82-208</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90915</th>\n",
       "      <td>22</td>\n",
       "      <td>51302662</td>\n",
       "      <td>51303662</td>\n",
       "      <td>1000</td>\n",
       "      <td>34.0</td>\n",
       "      <td>51270445</td>\n",
       "      <td>51302683</td>\n",
       "      <td>RPL23AP82-208</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90916</th>\n",
       "      <td>22</td>\n",
       "      <td>51303662</td>\n",
       "      <td>51306044</td>\n",
       "      <td>2382</td>\n",
       "      <td>36.0</td>\n",
       "      <td>51296750</td>\n",
       "      <td>51312835</td>\n",
       "      <td>RPL23AP82-205</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90917</th>\n",
       "      <td>22</td>\n",
       "      <td>51306044</td>\n",
       "      <td>51312584</td>\n",
       "      <td>6540</td>\n",
       "      <td>45.0</td>\n",
       "      <td>51296750</td>\n",
       "      <td>51312835</td>\n",
       "      <td>RPL23AP82-205</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90918</th>\n",
       "      <td>22</td>\n",
       "      <td>51312584</td>\n",
       "      <td>51324926</td>\n",
       "      <td>12342</td>\n",
       "      <td>25.0</td>\n",
       "      <td>51296750</td>\n",
       "      <td>51312835</td>\n",
       "      <td>RPL23AP82-205</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63950 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End  Length_cnvr  LP6005441-DNA_G07  \\\n",
       "0              1     10803     14487         3684               20.0   \n",
       "1              1     14487     18731         4244               14.0   \n",
       "2              1     18731     20013         1282               14.0   \n",
       "3              1     20013     28884         8871               13.0   \n",
       "4              1     51814     54076         2262               16.0   \n",
       "...          ...       ...       ...          ...                ...   \n",
       "90914         22  51301662  51302662         1000               34.0   \n",
       "90915         22  51302662  51303662         1000               34.0   \n",
       "90916         22  51303662  51306044         2382               36.0   \n",
       "90917         22  51306044  51312584         6540               45.0   \n",
       "90918         22  51312584  51324926        12342               25.0   \n",
       "\n",
       "        Start_b     End_b            name  score strand  ...  frameshift  \\\n",
       "0         11135     12457    AL627309.3-1      0      -  ...         NaN   \n",
       "1         14252     21099  AC114498.1-201      0      +  ...         NaN   \n",
       "2         14252     21099  AC114498.1-201      0      +  ...         NaN   \n",
       "3         14252     21099  AC114498.1-201      0      +  ...         NaN   \n",
       "4         52975     54613    AL731661.1-1      0      -  ...         NaN   \n",
       "...         ...       ...             ...    ...    ...  ...         ...   \n",
       "90914  51270445  51302683   RPL23AP82-208      0      +  ...         NaN   \n",
       "90915  51270445  51302683   RPL23AP82-208      0      +  ...         NaN   \n",
       "90916  51296750  51312835   RPL23AP82-205      0      +  ...         NaN   \n",
       "90917  51296750  51312835   RPL23AP82-205      0      +  ...         NaN   \n",
       "90918  51296750  51312835   RPL23AP82-205      0      +  ...         NaN   \n",
       "\n",
       "       exonAnnotationSupport intronAnnotationSupport  transcriptClass  \\\n",
       "0                        NaN                     NaN          paralog   \n",
       "1                        1,1                       1         ortholog   \n",
       "2                        1,1                       1         ortholog   \n",
       "3                        1,1                       1         ortholog   \n",
       "4                        NaN                     NaN          paralog   \n",
       "...                      ...                     ...              ...   \n",
       "90914                1,1,1,1                   1,1,1         ortholog   \n",
       "90915                1,1,1,1                   1,1,1         ortholog   \n",
       "90916                  1,1,1                     1,1         ortholog   \n",
       "90917                  1,1,1                     1,1         ortholog   \n",
       "90918                  1,1,1                     1,1         ortholog   \n",
       "\n",
       "      transcriptModes validStart validStop properOrf extra_paralog Distance  \n",
       "0             Liftoff        NaN       NaN       NaN          True        0  \n",
       "1            transMap       True      True      True         False        0  \n",
       "2            transMap       True      True      True         False        0  \n",
       "3            transMap       True      True      True         False        0  \n",
       "4             Liftoff        NaN       NaN       NaN          True        0  \n",
       "...               ...        ...       ...       ...           ...      ...  \n",
       "90914        transMap       True      True      True         False        0  \n",
       "90915        transMap       True      True      True         False        0  \n",
       "90916        transMap       True      True      True         False        0  \n",
       "90917        transMap       True      True      True         False        0  \n",
       "90918        transMap       True      True      True         False        0  \n",
       "\n",
       "[63950 rows x 42 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "4428325c-0e80-48a8-a7e4-85bf0538832f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>frameshift</th>\n",
       "      <th>exonAnnotationSupport</th>\n",
       "      <th>intronAnnotationSupport</th>\n",
       "      <th>transcriptClass</th>\n",
       "      <th>transcriptModes</th>\n",
       "      <th>validStart</th>\n",
       "      <th>validStop</th>\n",
       "      <th>properOrf</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10803</td>\n",
       "      <td>14487</td>\n",
       "      <td>3684</td>\n",
       "      <td>20.0</td>\n",
       "      <td>11135</td>\n",
       "      <td>12457</td>\n",
       "      <td>AL627309.3-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14487</td>\n",
       "      <td>18731</td>\n",
       "      <td>4244</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14252</td>\n",
       "      <td>21099</td>\n",
       "      <td>AC114498.1-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>18731</td>\n",
       "      <td>20013</td>\n",
       "      <td>1282</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14252</td>\n",
       "      <td>21099</td>\n",
       "      <td>AC114498.1-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>20013</td>\n",
       "      <td>28884</td>\n",
       "      <td>8871</td>\n",
       "      <td>13.0</td>\n",
       "      <td>14252</td>\n",
       "      <td>21099</td>\n",
       "      <td>AC114498.1-201</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>51814</td>\n",
       "      <td>54076</td>\n",
       "      <td>2262</td>\n",
       "      <td>16.0</td>\n",
       "      <td>52975</td>\n",
       "      <td>54613</td>\n",
       "      <td>AL731661.1-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>paralog</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90914</th>\n",
       "      <td>22</td>\n",
       "      <td>51301662</td>\n",
       "      <td>51302662</td>\n",
       "      <td>1000</td>\n",
       "      <td>34.0</td>\n",
       "      <td>51270445</td>\n",
       "      <td>51302683</td>\n",
       "      <td>RPL23AP82-208</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90915</th>\n",
       "      <td>22</td>\n",
       "      <td>51302662</td>\n",
       "      <td>51303662</td>\n",
       "      <td>1000</td>\n",
       "      <td>34.0</td>\n",
       "      <td>51270445</td>\n",
       "      <td>51302683</td>\n",
       "      <td>RPL23AP82-208</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1,1</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90916</th>\n",
       "      <td>22</td>\n",
       "      <td>51303662</td>\n",
       "      <td>51306044</td>\n",
       "      <td>2382</td>\n",
       "      <td>36.0</td>\n",
       "      <td>51296750</td>\n",
       "      <td>51312835</td>\n",
       "      <td>RPL23AP82-205</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90917</th>\n",
       "      <td>22</td>\n",
       "      <td>51306044</td>\n",
       "      <td>51312584</td>\n",
       "      <td>6540</td>\n",
       "      <td>45.0</td>\n",
       "      <td>51296750</td>\n",
       "      <td>51312835</td>\n",
       "      <td>RPL23AP82-205</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90918</th>\n",
       "      <td>22</td>\n",
       "      <td>51312584</td>\n",
       "      <td>51324926</td>\n",
       "      <td>12342</td>\n",
       "      <td>25.0</td>\n",
       "      <td>51296750</td>\n",
       "      <td>51312835</td>\n",
       "      <td>RPL23AP82-205</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,1,1</td>\n",
       "      <td>1,1</td>\n",
       "      <td>ortholog</td>\n",
       "      <td>transMap</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63950 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End  Length_cnvr  LP6005441-DNA_G07  \\\n",
       "0              1     10803     14487         3684               20.0   \n",
       "1              1     14487     18731         4244               14.0   \n",
       "2              1     18731     20013         1282               14.0   \n",
       "3              1     20013     28884         8871               13.0   \n",
       "4              1     51814     54076         2262               16.0   \n",
       "...          ...       ...       ...          ...                ...   \n",
       "90914         22  51301662  51302662         1000               34.0   \n",
       "90915         22  51302662  51303662         1000               34.0   \n",
       "90916         22  51303662  51306044         2382               36.0   \n",
       "90917         22  51306044  51312584         6540               45.0   \n",
       "90918         22  51312584  51324926        12342               25.0   \n",
       "\n",
       "        Start_b     End_b            name  score strand  ...  frameshift  \\\n",
       "0         11135     12457    AL627309.3-1      0      -  ...         NaN   \n",
       "1         14252     21099  AC114498.1-201      0      +  ...         NaN   \n",
       "2         14252     21099  AC114498.1-201      0      +  ...         NaN   \n",
       "3         14252     21099  AC114498.1-201      0      +  ...         NaN   \n",
       "4         52975     54613    AL731661.1-1      0      -  ...         NaN   \n",
       "...         ...       ...             ...    ...    ...  ...         ...   \n",
       "90914  51270445  51302683   RPL23AP82-208      0      +  ...         NaN   \n",
       "90915  51270445  51302683   RPL23AP82-208      0      +  ...         NaN   \n",
       "90916  51296750  51312835   RPL23AP82-205      0      +  ...         NaN   \n",
       "90917  51296750  51312835   RPL23AP82-205      0      +  ...         NaN   \n",
       "90918  51296750  51312835   RPL23AP82-205      0      +  ...         NaN   \n",
       "\n",
       "       exonAnnotationSupport intronAnnotationSupport  transcriptClass  \\\n",
       "0                        NaN                     NaN          paralog   \n",
       "1                        1,1                       1         ortholog   \n",
       "2                        1,1                       1         ortholog   \n",
       "3                        1,1                       1         ortholog   \n",
       "4                        NaN                     NaN          paralog   \n",
       "...                      ...                     ...              ...   \n",
       "90914                1,1,1,1                   1,1,1         ortholog   \n",
       "90915                1,1,1,1                   1,1,1         ortholog   \n",
       "90916                  1,1,1                     1,1         ortholog   \n",
       "90917                  1,1,1                     1,1         ortholog   \n",
       "90918                  1,1,1                     1,1         ortholog   \n",
       "\n",
       "      transcriptModes validStart validStop properOrf extra_paralog Distance  \n",
       "0             Liftoff        NaN       NaN       NaN          True        0  \n",
       "1            transMap       True      True      True         False        0  \n",
       "2            transMap       True      True      True         False        0  \n",
       "3            transMap       True      True      True         False        0  \n",
       "4             Liftoff        NaN       NaN       NaN          True        0  \n",
       "...               ...        ...       ...       ...           ...      ...  \n",
       "90914        transMap       True      True      True         False        0  \n",
       "90915        transMap       True      True      True         False        0  \n",
       "90916        transMap       True      True      True         False        0  \n",
       "90917        transMap       True      True      True         False        0  \n",
       "90918        transMap       True      True      True         False        0  \n",
       "\n",
       "[63950 rows x 42 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates_genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "19a26eee-0ee7-43ca-a74c-ca187b130d69",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Deletions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "2edbccd4-87d0-49d4-9161-0ee32c091008",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Genes \n",
    "deletions_t2t_coordinates_genes = t2t_deletions_gb.nearest(t2t_genes)\n",
    "deletions_t2t_coordinates_genes = deletions_t2t_coordinates_genes.as_df()\n",
    "deletions_t2t_coordinates_genes = deletions_t2t_coordinates_genes[deletions_t2t_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_mean_individual_genes= deletions_t2t_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_t2t_coordinates_telomeres = t2t_deletions_gb.nearest(t2t_telomeres)\n",
    "deletions_t2t_coordinates_telomeres = deletions_t2t_coordinates_telomeres.as_df()\n",
    "deletions_t2t_coordinates_telomeres = deletions_t2t_coordinates_telomeres[deletions_t2t_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_t2t_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_t2t_mean_individual_telomeres= deletions_t2t_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_t2t_coordinates_sds = t2t_deletions_gb.nearest(t2t_sds)\n",
    "deletions_t2t_coordinates_sds = deletions_t2t_coordinates_sds.as_df()\n",
    "deletions_t2t_coordinates_sds = deletions_t2t_coordinates_sds[deletions_t2t_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_mean_individual_sds= deletions_t2t_coordinates_sds['LP6005441-DNA_G07'].mean()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "49a1d1e1-3ea6-4c9a-a9fb-39eb950a80ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "bb365324-be37-4e25-b96e-cc6640763704",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Genes \n",
    "duplications_t2t_coordinates_genes = t2t_duplications_gb.nearest(t2t_genes)\n",
    "duplications_t2t_coordinates_genes = duplications_t2t_coordinates_genes.as_df()\n",
    "duplications_t2t_coordinates_genes = duplications_t2t_coordinates_genes[duplications_t2t_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_mean_individual_genes= duplications_t2t_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_t2t_coordinates_telomeres = t2t_duplications_gb.nearest(t2t_telomeres)\n",
    "duplications_t2t_coordinates_telomeres = duplications_t2t_coordinates_telomeres.as_df()\n",
    "duplications_t2t_coordinates_telomeres = duplications_t2t_coordinates_telomeres[duplications_t2t_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_t2t_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_t2t_mean_individual_telomeres= duplications_t2t_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_t2t_coordinates_sds = t2t_duplications_gb.nearest(t2t_sds)\n",
    "duplications_t2t_coordinates_sds = duplications_t2t_coordinates_sds.as_df()\n",
    "duplications_t2t_coordinates_sds = duplications_t2t_coordinates_sds[duplications_t2t_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_mean_individual_sds= duplications_t2t_coordinates_sds['LP6005441-DNA_G07'].mean()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccab1fb8-1a97-43df-9041-235962b1cff4",
   "metadata": {},
   "source": [
    "<h2> T2T CNVnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "d68610aa-58f1-4657-9dde-03ef84a89829",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>25001</td>\n",
       "      <td>45000</td>\n",
       "      <td>3.3208</td>\n",
       "      <td>19999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>48001</td>\n",
       "      <td>78000</td>\n",
       "      <td>3.2608</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>88001</td>\n",
       "      <td>96000</td>\n",
       "      <td>0.5038</td>\n",
       "      <td>7999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>2133001</td>\n",
       "      <td>2162000</td>\n",
       "      <td>7.1520</td>\n",
       "      <td>28999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>2193001</td>\n",
       "      <td>2223000</td>\n",
       "      <td>5.9894</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3408</th>\n",
       "      <td>22</td>\n",
       "      <td>24379001</td>\n",
       "      <td>24409000</td>\n",
       "      <td>3.0316</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3409</th>\n",
       "      <td>22</td>\n",
       "      <td>24409001</td>\n",
       "      <td>24427000</td>\n",
       "      <td>1.0380</td>\n",
       "      <td>17999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3410</th>\n",
       "      <td>22</td>\n",
       "      <td>24430001</td>\n",
       "      <td>24463000</td>\n",
       "      <td>1.1298</td>\n",
       "      <td>32999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3412</th>\n",
       "      <td>22</td>\n",
       "      <td>39109001</td>\n",
       "      <td>39114000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413</th>\n",
       "      <td>22</td>\n",
       "      <td>39369001</td>\n",
       "      <td>39373000</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1855 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome     Start       End  CNV_value  Length_cnvr\n",
       "0             1     25001     45000     3.3208        19999\n",
       "1             1     48001     78000     3.2608        29999\n",
       "2             1     88001     96000     0.5038         7999\n",
       "6             1   2133001   2162000     7.1520        28999\n",
       "8             1   2193001   2223000     5.9894        29999\n",
       "...         ...       ...       ...        ...          ...\n",
       "3408         22  24379001  24409000     3.0316        29999\n",
       "3409         22  24409001  24427000     1.0380        17999\n",
       "3410         22  24430001  24463000     1.1298        32999\n",
       "3412         22  39109001  39114000     0.0000         4999\n",
       "3413         22  39369001  39373000     0.0014         3999\n",
       "\n",
       "[1855 rows x 5 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header_list = ['Type', 'Position', 'Lenght', 'RD', 'T-TEST-EVAL', 'GAUSSIAN-EVAL', 'T-TEST-EVAL(MIDDLE)','GAUSSIAN-EVAL(MIDDLE)', 'FRAQ-READ-QUALITY', 't', 't2']\n",
    "\n",
    "t2t_data = pd.read_csv('/Users/luciabazan/t2t_LP6005441-DNA_G07_cnvs.tsv', sep='\\t' , names=header_list)\n",
    "t2t_data\n",
    "\n",
    "\n",
    "t2t_data = t2t_data[t2t_data['T-TEST-EVAL'] <= 0.05]\n",
    "t2t_data\n",
    "\n",
    "cnvnator_data = t2t_data\n",
    "\n",
    "#SEPARATE COORDINATES\n",
    "coordinates = cnvnator_data['Position'].str.split(':', expand=True)\n",
    "CHROMOSOME = coordinates[0]\n",
    "\n",
    "split_position = coordinates[1].str.split('-', expand=True)\n",
    "START = split_position[0]\n",
    "STOP = split_position[1]\n",
    "\n",
    "cnvnator_style= pd.DataFrame()\n",
    "cnvnator_style['Chromosome'] = CHROMOSOME\n",
    "cnvnator_style['Start'] = START\n",
    "cnvnator_style['End'] = STOP\n",
    "cnvnator_style['CNV_value'] = cnvnator_data['RD']\n",
    "\n",
    "\n",
    "cnvnator_style['Start'] = cnvnator_style['Start'].astype(int)\n",
    "cnvnator_style['End'] = cnvnator_style['End'].astype(int)\n",
    "cnvnator_style['CNV_value'] = cnvnator_style['CNV_value'].astype(float)\n",
    "\n",
    "#cnvnator_style['Type'] = cnvnator_data['Type']\n",
    "\n",
    "cnvnator_style['CNV_value'] = cnvnator_style['CNV_value'] * 2\n",
    "\n",
    "cnvnator_style['Chromosome'] = cnvnator_style['Chromosome'].str.replace('chr','')\n",
    "\n",
    "t2t_cnvnator = cnvnator_style\n",
    "t2t_cnvnator['Length_cnvr'] = t2t_cnvnator['End'] - t2t_cnvnator['Start']\n",
    "t2t_cnvnator = t2t_cnvnator[t2t_cnvnator['Chromosome'] != 'M']\n",
    "t2t_cnvnator = t2t_cnvnator[t2t_cnvnator['Chromosome'] != 'X']\n",
    "t2t_cnvnator = t2t_cnvnator[t2t_cnvnator['Chromosome'] != 'Y']\n",
    "t2t_cnvnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "5c1906b5-86f4-481e-89e2-42c528a62792",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>25001</td>\n",
       "      <td>45000</td>\n",
       "      <td>3.3208</td>\n",
       "      <td>19999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>48001</td>\n",
       "      <td>78000</td>\n",
       "      <td>3.2608</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>88001</td>\n",
       "      <td>96000</td>\n",
       "      <td>0.5038</td>\n",
       "      <td>7999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2133001</td>\n",
       "      <td>2162000</td>\n",
       "      <td>7.1520</td>\n",
       "      <td>28999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2193001</td>\n",
       "      <td>2223000</td>\n",
       "      <td>5.9894</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850</th>\n",
       "      <td>22</td>\n",
       "      <td>24379001</td>\n",
       "      <td>24409000</td>\n",
       "      <td>3.0316</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1851</th>\n",
       "      <td>22</td>\n",
       "      <td>24409001</td>\n",
       "      <td>24427000</td>\n",
       "      <td>1.0380</td>\n",
       "      <td>17999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1852</th>\n",
       "      <td>22</td>\n",
       "      <td>24430001</td>\n",
       "      <td>24463000</td>\n",
       "      <td>1.1298</td>\n",
       "      <td>32999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>22</td>\n",
       "      <td>39109001</td>\n",
       "      <td>39114000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>22</td>\n",
       "      <td>39369001</td>\n",
       "      <td>39373000</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1855 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "| Chromosome   | Start     | End       | CNV_value   | Length_cnvr   |\n",
       "| (category)   | (int64)   | (int64)   | (float64)   | (int64)       |\n",
       "|--------------+-----------+-----------+-------------+---------------|\n",
       "| 1            | 25001     | 45000     | 3.3208      | 19999         |\n",
       "| 1            | 48001     | 78000     | 3.2608      | 29999         |\n",
       "| 1            | 88001     | 96000     | 0.5038      | 7999          |\n",
       "| 1            | 2133001   | 2162000   | 7.152       | 28999         |\n",
       "| ...          | ...       | ...       | ...         | ...           |\n",
       "| 22           | 24409001  | 24427000  | 1.038       | 17999         |\n",
       "| 22           | 24430001  | 24463000  | 1.1298      | 32999         |\n",
       "| 22           | 39109001  | 39114000  | 0.0         | 4999          |\n",
       "| 22           | 39369001  | 39373000  | 0.0014      | 3999          |\n",
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "Unstranded PyRanges object has 1,855 rows and 5 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2t_deletions_cnvnator = t2t_cnvnator[t2t_cnvnator['CNV_value'] < 2]\n",
    "t2t_deletions_cnvnator_mean = t2t_deletions_cnvnator['CNV_value'].mean()\n",
    "t2t_duplications_cnvnator = t2t_cnvnator[t2t_cnvnator['CNV_value'] > 2]\n",
    "t2t_duplications_cnvnator_mean = t2t_duplications_cnvnator['CNV_value'].mean()\n",
    "t2t_duplications_cnvnator = PyRanges(t2t_duplications_cnvnator)\n",
    "t2t_deletions_cnvnator = PyRanges(t2t_deletions_cnvnator)\n",
    "\n",
    "t2t_cnvnator= PyRanges(t2t_cnvnator)\n",
    "t2t_cnvnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "083a418f-d419-43af-9452-5a3a7d30d028",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "8627f457-8192-48d5-93cf-0caf1a2186ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### t2t genes \n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator.nearest(t2t_genes)\n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator_coordinates_genes.as_df()\n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator_coordinates_genes[t2t_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "\n",
    "## T2T telomeres\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator.nearest(t2t_telomeres)\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator_coordinates_telomeres.as_df()\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator_coordinates_telomeres[t2t_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs \n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator.nearest(t2t_sds)\n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator_coordinates_sds.as_df()\n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator_coordinates_sds[t2t_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "b938fddc-1393-401e-9c13-b20443820381",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Deletions\n",
    "#### Genes \n",
    "deletions_t2t_cnvnator_coordinates_genes = t2t_deletions_cnvnator.nearest(t2t_genes)\n",
    "deletions_t2t_cnvnator_coordinates_genes = deletions_t2t_cnvnator_coordinates_genes.as_df()\n",
    "deletions_t2t_cnvnator_coordinates_genes = deletions_t2t_cnvnator_coordinates_genes[deletions_t2t_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_cnvnator_mean_individual_genes= deletions_t2t_cnvnator_coordinates_genes['CNV_value'].mean()\n",
    "deletions_t2t_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_t2t_cnvnator_coordinates_telomeres = t2t_deletions_cnvnator.nearest(t2t_telomeres)\n",
    "deletions_t2t_cnvnator_coordinates_telomeres = deletions_t2t_cnvnator_coordinates_telomeres.as_df()\n",
    "deletions_t2t_cnvnator_coordinates_telomeres = deletions_t2t_cnvnator_coordinates_telomeres[deletions_t2t_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_t2t_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_t2t_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_cnvnator_mean_individual_telomeres= deletions_t2t_cnvnator_coordinates_telomeres['CNV_value'].mean()\n",
    "deletions_t2t_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_t2t_cnvnator_coordinates_sds = t2t_deletions_cnvnator.nearest(t2t_sds)\n",
    "deletions_t2t_cnvnator_coordinates_sds = deletions_t2t_cnvnator_coordinates_sds.as_df()\n",
    "deletions_t2t_cnvnator_coordinates_sds = deletions_t2t_cnvnator_coordinates_sds[deletions_t2t_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_t2t_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_t2t_cnvnator_mean_individual_sds= deletions_t2t_cnvnator_coordinates_sds['CNV_value'].mean()\n",
    "deletions_t2t_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "8e147424-6af8-4dec-a7d2-afd746f88002",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.981063976945245"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deletions_t2t_cnvnator_mean_individual_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "59b41445-b950-4946-90e5-ae4dd6f8e960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.5933225225225227"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Duplications\n",
    "#### Genes \n",
    "duplications_t2t_cnvnator_coordinates_genes = t2t_duplications_cnvnator.nearest(t2t_genes)\n",
    "duplications_t2t_cnvnator_coordinates_genes = duplications_t2t_cnvnator_coordinates_genes.as_df()\n",
    "duplications_t2t_cnvnator_coordinates_genes = duplications_t2t_cnvnator_coordinates_genes[duplications_t2t_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_t2t_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_genes= duplications_t2t_cnvnator_coordinates_genes['CNV_value'].mean()\n",
    "duplications_t2t_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_t2t_cnvnator_coordinates_telomeres = t2t_duplications_cnvnator.nearest(t2t_telomeres)\n",
    "duplications_t2t_cnvnator_coordinates_telomeres = duplications_t2t_cnvnator_coordinates_telomeres.as_df()\n",
    "duplications_t2t_cnvnator_coordinates_telomeres = duplications_t2t_cnvnator_coordinates_telomeres[duplications_t2t_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_t2t_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_t2t_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_telomeres= duplications_t2t_cnvnator_coordinates_telomeres['CNV_value'].mean()\n",
    "duplications_t2t_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_t2t_cnvnator_coordinates_sds = t2t_duplications_cnvnator.nearest(t2t_sds)\n",
    "duplications_t2t_cnvnator_coordinates_sds = duplications_t2t_cnvnator_coordinates_sds.as_df()\n",
    "duplications_t2t_cnvnator_coordinates_sds = duplications_t2t_cnvnator_coordinates_sds[duplications_t2t_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "\n",
    "duplications_t2t_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_sds= duplications_t2t_cnvnator_coordinates_sds['CNV_value'].mean()\n",
    "duplications_t2t_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)\n",
    "duplications_t2t_cnvnator_mean_individual_sds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7ffae1-94ba-45c0-95d7-e4426e4b28ae",
   "metadata": {},
   "source": [
    "<h1> GRCh37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "21bf4b13-6a3b-4874-bebc-d5d9f6eca33d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Anotations\n",
    "grch37_genes = pd.read_csv('../annotation_files/grch37-genes.csv')\n",
    "grch37_telomeres = pd.read_csv('../annotation_files/grch37-telomers-centromers.csv')\n",
    "grch37_sds = pd.read_csv('../annotation_files/grch37-sds.csv')\n",
    "#grch37_sds = grch37_sds[['#chr1', 'start1', 'end1']]\n",
    "\n",
    "\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "\n",
    "\n",
    "#t2t_sds.columns =['Chromosome','Start','End']\n",
    "#t2t_sds\n",
    "grch37_genes.columns = ['#\"bin\"', 'name', 'Chromosome', 'strand', 'Start', 'End', 'cdsStart',\n",
    "       'cdsEnd', 'exonCount', 'exonStarts', 'exonEnds', 'score', 'name2',\n",
    "       'cdsStartStat', 'cdsEndStat', 'exonFrames'] \n",
    "grch37_telomeres = grch37_telomeres[grch37_telomeres['type'] == 'telomere']\n",
    "grch37_sds.columns = ['#\"bin\"', 'Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'otherChrom', 'otherStart', 'otherEnd', 'otherSize', 'uid',\n",
    "       'posBasesHit', 'testResult', 'verdict', 'chits', 'ccov', 'alignfile',\n",
    "       'alignL', 'indelN', 'indelS', 'alignB', 'matchB', 'mismatchB',\n",
    "       'transitionsB', 'transversionsB', 'fracMatch', 'fracMatchIndel', 'jcK',\n",
    "       'k2K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "92a4acd0-37cc-4be4-a1ce-1ac6fba433d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_91372/4141503609.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_old['RD'] = sample_old['RD'] * 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Type</th>\n",
       "      <th>RD</th>\n",
       "      <th>Lenght</th>\n",
       "      <th>T-TEST-EVAL</th>\n",
       "      <th>GAUSSIAN-EVAL</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51663</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>234001</td>\n",
       "      <td>238300</td>\n",
       "      <td>3.0</td>\n",
       "      <td>duplication</td>\n",
       "      <td>4.170320</td>\n",
       "      <td>4300.0</td>\n",
       "      <td>2.965070e-10</td>\n",
       "      <td>1.407390e+02</td>\n",
       "      <td>4299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51664</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.010921</td>\n",
       "      <td>50700.0</td>\n",
       "      <td>3.143440e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51665</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>50100.0</td>\n",
       "      <td>3.181090e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51666</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.444038</td>\n",
       "      <td>3900.0</td>\n",
       "      <td>4.086470e-11</td>\n",
       "      <td>2.141070e-31</td>\n",
       "      <td>3899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51667</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.013284</td>\n",
       "      <td>3400.0</td>\n",
       "      <td>4.687430e-11</td>\n",
       "      <td>1.314760e-38</td>\n",
       "      <td>3399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52386</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138194201</td>\n",
       "      <td>138197800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.323550</td>\n",
       "      <td>3600.0</td>\n",
       "      <td>1.411230e-02</td>\n",
       "      <td>2.138410e+09</td>\n",
       "      <td>3599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52387</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138214101</td>\n",
       "      <td>138217600</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.019674</td>\n",
       "      <td>3500.0</td>\n",
       "      <td>4.553500e-11</td>\n",
       "      <td>1.185690e-11</td>\n",
       "      <td>3499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52388</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138479101</td>\n",
       "      <td>138480200</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>1.448840e-10</td>\n",
       "      <td>3.214710e-77</td>\n",
       "      <td>1099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52389</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>139663601</td>\n",
       "      <td>139665300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.604050</td>\n",
       "      <td>1700.0</td>\n",
       "      <td>1.392830e-03</td>\n",
       "      <td>2.771230e-02</td>\n",
       "      <td>1699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52390</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>140253401</td>\n",
       "      <td>140256000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.986446</td>\n",
       "      <td>2600.0</td>\n",
       "      <td>6.129710e-11</td>\n",
       "      <td>5.009630e-27</td>\n",
       "      <td>2599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Sample_ID Chromosome      Start        End  CNV_Value  \\\n",
       "51663  LP6005441-DNA_G07          1     234001     238300        3.0   \n",
       "51664  LP6005441-DNA_G07          1     267101     317800        1.0   \n",
       "51665  LP6005441-DNA_G07          1     471301     521400        1.0   \n",
       "51666  LP6005441-DNA_G07          1    1010801    1014700        1.0   \n",
       "51667  LP6005441-DNA_G07          1    4124401    4127800        1.0   \n",
       "...                  ...        ...        ...        ...        ...   \n",
       "52386  LP6005441-DNA_G07          9  138194201  138197800        1.0   \n",
       "52387  LP6005441-DNA_G07          9  138214101  138217600        1.0   \n",
       "52388  LP6005441-DNA_G07          9  138479101  138480200        1.0   \n",
       "52389  LP6005441-DNA_G07          9  139663601  139665300        1.0   \n",
       "52390  LP6005441-DNA_G07          9  140253401  140256000        1.0   \n",
       "\n",
       "              Type        RD   Lenght   T-TEST-EVAL  GAUSSIAN-EVAL  \\\n",
       "51663  duplication  4.170320   4300.0  2.965070e-10   1.407390e+02   \n",
       "51664     deletion  0.010921  50700.0  3.143440e-12   0.000000e+00   \n",
       "51665     deletion  0.000109  50100.0  3.181090e-12   0.000000e+00   \n",
       "51666     deletion  0.444038   3900.0  4.086470e-11   2.141070e-31   \n",
       "51667     deletion  1.013284   3400.0  4.687430e-11   1.314760e-38   \n",
       "...            ...       ...      ...           ...            ...   \n",
       "52386     deletion  1.323550   3600.0  1.411230e-02   2.138410e+09   \n",
       "52387     deletion  1.019674   3500.0  4.553500e-11   1.185690e-11   \n",
       "52388     deletion  0.000000   1100.0  1.448840e-10   3.214710e-77   \n",
       "52389     deletion  0.604050   1700.0  1.392830e-03   2.771230e-02   \n",
       "52390     deletion  0.986446   2600.0  6.129710e-11   5.009630e-27   \n",
       "\n",
       "       Length_cnvr  \n",
       "51663         4299  \n",
       "51664        50699  \n",
       "51665        50099  \n",
       "51666         3899  \n",
       "51667         3399  \n",
       "...            ...  \n",
       "52386         3599  \n",
       "52387         3499  \n",
       "52388         1099  \n",
       "52389         1699  \n",
       "52390         2599  \n",
       "\n",
       "[728 rows x 11 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"../call_cnvr_cnvnator/individual_cnv_cnvr.txt\", sep='\\t')\n",
    "data = pd.read_csv(\"../data/cnvnator_cnvs.csv\")\n",
    "data.columns = ['Sample_ID','Chromosome', 'Start', 'End','CNV_Value','Type','RD','Lenght', 'T-TEST-EVAL','GAUSSIAN-EVAL',]\n",
    "data = data.dropna()#.describe()\n",
    "data['Length_cnvr'] = data['End'] - data['Start'] \n",
    "sample_old = data[data['Sample_ID'] == 'LP6005441-DNA_G07']\n",
    "#sample_old = old_assembly[old_assembly['Sample_ID'] == 'LP6005441-DNA_B10']\n",
    "#sample_old = sample_old[['Chromosome', 'Start', 'End', 'RD']]\n",
    "sample_old['RD'] = sample_old['RD'] * 2\n",
    "#sample_old['Chromosome'] = str('chr')+sample_old['Chromosome']\n",
    "    #sample_old.to_csv('LP6005441-DNA_B10_grch37.bed', index=False, header=False, sep='\\t')\n",
    "    #list(sample_old.columns)\n",
    "\n",
    "#sample_old\n",
    "\n",
    "sample_old = sample_old[~sample_old['RD'].isin([2])]\n",
    "\n",
    "sample_old = sample_old[sample_old['Chromosome'] != 'M']\n",
    "sample_old = sample_old[sample_old['Chromosome'] != 'X']\n",
    "sample_old = sample_old[sample_old['Chromosome'] != 'Y']\n",
    "\n",
    "sample_old "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "f87a0006-3796-4c37-969d-d3dedb628afc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Type</th>\n",
       "      <th>RD</th>\n",
       "      <th>Lenght</th>\n",
       "      <th>T-TEST-EVAL</th>\n",
       "      <th>GAUSSIAN-EVAL</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>234001</td>\n",
       "      <td>238300</td>\n",
       "      <td>3.0</td>\n",
       "      <td>duplication</td>\n",
       "      <td>4.170320</td>\n",
       "      <td>4300.0</td>\n",
       "      <td>2.965070e-10</td>\n",
       "      <td>1.407390e+02</td>\n",
       "      <td>4299</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.010921</td>\n",
       "      <td>50700.0</td>\n",
       "      <td>3.143440e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50699</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>50100.0</td>\n",
       "      <td>3.181090e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>50099</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.444038</td>\n",
       "      <td>3900.0</td>\n",
       "      <td>4.086470e-11</td>\n",
       "      <td>2.141070e-31</td>\n",
       "      <td>3899</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.013284</td>\n",
       "      <td>3400.0</td>\n",
       "      <td>4.687430e-11</td>\n",
       "      <td>1.314760e-38</td>\n",
       "      <td>3399</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138194201</td>\n",
       "      <td>138197800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.323550</td>\n",
       "      <td>3600.0</td>\n",
       "      <td>1.411230e-02</td>\n",
       "      <td>2.138410e+09</td>\n",
       "      <td>3599</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138214101</td>\n",
       "      <td>138217600</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1.019674</td>\n",
       "      <td>3500.0</td>\n",
       "      <td>4.553500e-11</td>\n",
       "      <td>1.185690e-11</td>\n",
       "      <td>3499</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>138479101</td>\n",
       "      <td>138480200</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>1.448840e-10</td>\n",
       "      <td>3.214710e-77</td>\n",
       "      <td>1099</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>139663601</td>\n",
       "      <td>139665300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.604050</td>\n",
       "      <td>1700.0</td>\n",
       "      <td>1.392830e-03</td>\n",
       "      <td>2.771230e-02</td>\n",
       "      <td>1699</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>9</td>\n",
       "      <td>140253401</td>\n",
       "      <td>140256000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>0.986446</td>\n",
       "      <td>2600.0</td>\n",
       "      <td>6.129710e-11</td>\n",
       "      <td>5.009630e-27</td>\n",
       "      <td>2599</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>America</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample_ID Chromosome      Start        End  CNV_Value  \\\n",
       "0    LP6005441-DNA_G07          1     234001     238300        3.0   \n",
       "1    LP6005441-DNA_G07          1     267101     317800        1.0   \n",
       "2    LP6005441-DNA_G07          1     471301     521400        1.0   \n",
       "3    LP6005441-DNA_G07          1    1010801    1014700        1.0   \n",
       "4    LP6005441-DNA_G07          1    4124401    4127800        1.0   \n",
       "..                 ...        ...        ...        ...        ...   \n",
       "723  LP6005441-DNA_G07          9  138194201  138197800        1.0   \n",
       "724  LP6005441-DNA_G07          9  138214101  138217600        1.0   \n",
       "725  LP6005441-DNA_G07          9  138479101  138480200        1.0   \n",
       "726  LP6005441-DNA_G07          9  139663601  139665300        1.0   \n",
       "727  LP6005441-DNA_G07          9  140253401  140256000        1.0   \n",
       "\n",
       "            Type        RD   Lenght   T-TEST-EVAL  GAUSSIAN-EVAL  Length_cnvr  \\\n",
       "0    duplication  4.170320   4300.0  2.965070e-10   1.407390e+02         4299   \n",
       "1       deletion  0.010921  50700.0  3.143440e-12   0.000000e+00        50699   \n",
       "2       deletion  0.000109  50100.0  3.181090e-12   0.000000e+00        50099   \n",
       "3       deletion  0.444038   3900.0  4.086470e-11   2.141070e-31         3899   \n",
       "4       deletion  1.013284   3400.0  4.687430e-11   1.314760e-38         3399   \n",
       "..           ...       ...      ...           ...            ...          ...   \n",
       "723     deletion  1.323550   3600.0  1.411230e-02   2.138410e+09         3599   \n",
       "724     deletion  1.019674   3500.0  4.553500e-11   1.185690e-11         3499   \n",
       "725     deletion  0.000000   1100.0  1.448840e-10   3.214710e-77         1099   \n",
       "726     deletion  0.604050   1700.0  1.392830e-03   2.771230e-02         1699   \n",
       "727     deletion  0.986446   2600.0  6.129710e-11   5.009630e-27         2599   \n",
       "\n",
       "                Sample   Region  \n",
       "0    LP6005441-DNA_G07  America  \n",
       "1    LP6005441-DNA_G07  America  \n",
       "2    LP6005441-DNA_G07  America  \n",
       "3    LP6005441-DNA_G07  America  \n",
       "4    LP6005441-DNA_G07  America  \n",
       "..                 ...      ...  \n",
       "723  LP6005441-DNA_G07  America  \n",
       "724  LP6005441-DNA_G07  America  \n",
       "725  LP6005441-DNA_G07  America  \n",
       "726  LP6005441-DNA_G07  America  \n",
       "727  LP6005441-DNA_G07  America  \n",
       "\n",
       "[728 rows x 13 columns]"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###### Keeping only samples that are on old cnvs \n",
    "sample_old = sample_old.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])\n",
    "\n",
    "#REMOVE SAMPLES WITH ABERRATIONS\n",
    "#sample_old.columns = ['Chromosome', 'Start', 'End', 'CNVR_ID', 'Sample_iD', 'CNV_start', 'CNV_end', 'CNV_Value','Length', 'Length_cnvr','Sample','Region']\n",
    "sample_old "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "f5ab003b-6d04-431b-ac90-0be08e8f2049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>234001</td>\n",
       "      <td>238300</td>\n",
       "      <td>4299</td>\n",
       "      <td>4.170320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>50699</td>\n",
       "      <td>0.010921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>50099</td>\n",
       "      <td>0.000109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>3899</td>\n",
       "      <td>0.444038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>3399</td>\n",
       "      <td>1.013284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>9</td>\n",
       "      <td>138194201</td>\n",
       "      <td>138197800</td>\n",
       "      <td>3599</td>\n",
       "      <td>1.323550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>9</td>\n",
       "      <td>138214101</td>\n",
       "      <td>138217600</td>\n",
       "      <td>3499</td>\n",
       "      <td>1.019674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>9</td>\n",
       "      <td>138479101</td>\n",
       "      <td>138480200</td>\n",
       "      <td>1099</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>9</td>\n",
       "      <td>139663601</td>\n",
       "      <td>139665300</td>\n",
       "      <td>1699</td>\n",
       "      <td>0.604050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>9</td>\n",
       "      <td>140253401</td>\n",
       "      <td>140256000</td>\n",
       "      <td>2599</td>\n",
       "      <td>0.986446</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome      Start        End  Length_cnvr  LP6005441-DNA_G07\n",
       "0               1     234001     238300         4299           4.170320\n",
       "1               1     267101     317800        50699           0.010921\n",
       "2               1     471301     521400        50099           0.000109\n",
       "3               1    1010801    1014700         3899           0.444038\n",
       "4               1    4124401    4127800         3399           1.013284\n",
       "..            ...        ...        ...          ...                ...\n",
       "723             9  138194201  138197800         3599           1.323550\n",
       "724             9  138214101  138217600         3499           1.019674\n",
       "725             9  138479101  138480200         1099           0.000000\n",
       "726             9  139663601  139665300         1699           0.604050\n",
       "727             9  140253401  140256000         2599           0.986446\n",
       "\n",
       "[728 rows x 5 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv_grch37 = sample_old.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='RD',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "\n",
    "final_cnv_grch37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "3200a16b-2464-4e47-8063-9e46662ba8f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>50699</td>\n",
       "      <td>0.010921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>50099</td>\n",
       "      <td>0.000109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>3899</td>\n",
       "      <td>0.444038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>3399</td>\n",
       "      <td>1.013284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>8360201</td>\n",
       "      <td>8364000</td>\n",
       "      <td>3799</td>\n",
       "      <td>0.017621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>22</td>\n",
       "      <td>21880701</td>\n",
       "      <td>21887200</td>\n",
       "      <td>6499</td>\n",
       "      <td>0.925102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>22</td>\n",
       "      <td>21899601</td>\n",
       "      <td>21905500</td>\n",
       "      <td>5899</td>\n",
       "      <td>1.209440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>22</td>\n",
       "      <td>23478501</td>\n",
       "      <td>23479700</td>\n",
       "      <td>1199</td>\n",
       "      <td>0.186410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>22</td>\n",
       "      <td>37143101</td>\n",
       "      <td>37147900</td>\n",
       "      <td>4799</td>\n",
       "      <td>0.057464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>22</td>\n",
       "      <td>39294101</td>\n",
       "      <td>39298700</td>\n",
       "      <td>4599</td>\n",
       "      <td>0.003570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+---------------+---------------------+\n",
       "| Chromosome   | Start     | End       | Length_cnvr   | LP6005441-DNA_G07   |\n",
       "| (category)   | (int64)   | (int64)   | (int64)       | (float64)           |\n",
       "|--------------+-----------+-----------+---------------+---------------------|\n",
       "| 1            | 267101    | 317800    | 50699         | 0.01092072          |\n",
       "| 1            | 471301    | 521400    | 50099         | 0.0001087124        |\n",
       "| 1            | 1010801   | 1014700   | 3899          | 0.444038            |\n",
       "| 1            | 4124401   | 4127800   | 3399          | 1.013284            |\n",
       "| ...          | ...       | ...       | ...           | ...                 |\n",
       "| 22           | 21899601  | 21905500  | 5899          | 1.20944             |\n",
       "| 22           | 23478501  | 23479700  | 1199          | 0.1864096           |\n",
       "| 22           | 37143101  | 37147900  | 4799          | 0.057464            |\n",
       "| 22           | 39294101  | 39298700  | 4599          | 0.00356992          |\n",
       "+--------------+-----------+-----------+---------------+---------------------+\n",
       "Unstranded PyRanges object has 508 rows and 5 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_deletions_cnvnator = final_cnv_grch37[final_cnv_grch37['LP6005441-DNA_G07'] < 2]\n",
    "grch37_deletions_cnvnator_mean = grch37_deletions_cnvnator['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "grch37_duplications_cnvnator = final_cnv_grch37[final_cnv_grch37['LP6005441-DNA_G07'] > 2]\n",
    "grch37_duplications_cnvnator_mean = grch37_duplications_cnvnator['LP6005441-DNA_G07'].mean()\n",
    "\n",
    "grch37_duplications_cnvnator = PyRanges(grch37_duplications_cnvnator)\n",
    "grch37_deletions_cnvnator = PyRanges(grch37_deletions_cnvnator)\n",
    "\n",
    "\n",
    "final_cnv_ranges_grch37 = PyRanges(final_cnv_grch37)\n",
    "grch37_deletions_cnvnator"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f006e6db-c6d3-4dea-9126-9f0ce4e30ed2",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "8e734c44-9ebc-4144-95a1-58803104e4fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "grch37_genes['Chromosome'] = grch37_genes['Chromosome'].str.replace('chr', '')\n",
    "grch37_genes_coordinates = PyRanges(grch37_genes)\n",
    "\n",
    "grch37_telomeres['Chromosome'] = grch37_telomeres['Chromosome'].str.replace('chr', '')\n",
    "grch37_telomeres = PyRanges(grch37_telomeres)\n",
    "\n",
    "grch37_sds['Chromosome'] = grch37_sds['Chromosome'].str.replace('chr', '')\n",
    "grch37_sds = PyRanges(grch37_sds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "b3c811e9-1372-4e7f-9742-02d296c392b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genes\n",
    "grch37_coordinates_genes = final_cnv_ranges_grch37.nearest(grch37_genes_coordinates)\n",
    "grch37_coordinates_genes = grch37_coordinates_genes.as_df()\n",
    "grch37_coordinates_genes = grch37_coordinates_genes[grch37_coordinates_genes['Distance'] <1 ]\n",
    "#grch37_coordinates_genes\n",
    "\n",
    "#Telomeres\n",
    "grch37_coordinates_telomeres = final_cnv_ranges_grch37.nearest(grch37_telomeres)\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres.as_df()\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres[grch37_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs\n",
    "grch37_coordinates_sds = final_cnv_ranges_grch37.nearest(grch37_sds)\n",
    "grch37_coordinates_sds = grch37_coordinates_sds.as_df()\n",
    "grch37_coordinates_sds = grch37_coordinates_sds[grch37_coordinates_sds['Distance'] <1 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "d95bff56-ba99-456e-8d98-67bbd853687f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#grch37_genes = len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "#grch37_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eb30162-a0ee-47c5-9dce-657e51b770f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "3324939e-9862-4985-b021-884aaa8c9347",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Deletions\n",
    "#### Genes \n",
    "deletions_grch37_cnvnator_coordinates_genes = grch37_deletions_cnvnator.nearest(grch37_genes_coordinates)\n",
    "deletions_grch37_cnvnator_coordinates_genes = deletions_grch37_cnvnator_coordinates_genes.as_df()\n",
    "deletions_grch37_cnvnator_coordinates_genes = deletions_grch37_cnvnator_coordinates_genes[deletions_grch37_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_cnvnator_mean_individual_genes= deletions_grch37_cnvnator_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "deletions_grch37_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_grch37_cnvnator_coordinates_telomeres = grch37_deletions_cnvnator.nearest(grch37_telomeres)\n",
    "deletions_grch37_cnvnator_coordinates_telomeres = deletions_grch37_cnvnator_coordinates_telomeres.as_df()\n",
    "deletions_grch37_cnvnator_coordinates_telomeres = deletions_grch37_cnvnator_coordinates_telomeres[deletions_grch37_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_grch37_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_grch37_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_cnvnator_mean_individual_telomeres= deletions_grch37_cnvnator_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "deletions_grch37_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_grch37_cnvnator_coordinates_sds = grch37_deletions_cnvnator.nearest(grch37_sds)\n",
    "deletions_grch37_cnvnator_coordinates_sds = deletions_grch37_cnvnator_coordinates_sds.as_df()\n",
    "deletions_grch37_cnvnator_coordinates_sds = deletions_grch37_cnvnator_coordinates_sds[deletions_grch37_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_cnvnator_mean_individual_sds= deletions_grch37_cnvnator_coordinates_sds['LP6005441-DNA_G07'].mean()\n",
    "deletions_grch37_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "b1318180-3c75-4c62-a2d3-2cfc72cc820f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Duplications\n",
    "#### Genes \n",
    "duplications_grch37_cnvnator_coordinates_genes = grch37_duplications_cnvnator.nearest(grch37_genes_coordinates)\n",
    "duplications_grch37_cnvnator_coordinates_genes = duplications_grch37_cnvnator_coordinates_genes.as_df()\n",
    "duplications_grch37_cnvnator_coordinates_genes = duplications_grch37_cnvnator_coordinates_genes[duplications_grch37_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_cnvnator_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_cnvnator_mean_individual_genes= duplications_grch37_cnvnator_coordinates_genes['LP6005441-DNA_G07'].mean()\n",
    "duplications_grch37_cnvnator_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_grch37_cnvnator_coordinates_telomeres = grch37_duplications_cnvnator.nearest(grch37_telomeres)\n",
    "duplications_grch37_cnvnator_coordinates_telomeres = duplications_grch37_cnvnator_coordinates_telomeres.as_df()\n",
    "duplications_grch37_cnvnator_coordinates_telomeres = duplications_grch37_cnvnator_coordinates_telomeres[duplications_grch37_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_grch37_cnvnator_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_grch37_cnvnator_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_cnvnator_mean_individual_telomeres= duplications_grch37_cnvnator_coordinates_telomeres['LP6005441-DNA_G07'].mean()\n",
    "duplications_grch37_cnvnator_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_grch37_cnvnator_coordinates_sds = grch37_duplications_cnvnator.nearest(grch37_sds)\n",
    "duplications_grch37_cnvnator_coordinates_sds = duplications_grch37_cnvnator_coordinates_sds.as_df()\n",
    "duplications_grch37_cnvnator_coordinates_sds = duplications_grch37_cnvnator_coordinates_sds[duplications_grch37_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_cnvnator_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_cnvnator_mean_individual_sds= duplications_grch37_cnvnator_coordinates_sds['LP6005441-DNA_G07'].mean()\n",
    "duplications_grch37_cnvnator_coordinates_sds.replace(np.nan,2, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02a2ca0-2689-4d75-a250-04f12f78179b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "258edfda-e062-4a97-b450-3c3ff2798083",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "55c3b568-3b07-4850-8ec0-a81222f3d5d3",
   "metadata": {},
   "source": [
    "<h2> Peter Sudmant data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "cf3b31d8-07cb-4c2b-bc22-66e26786d3f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>177033</td>\n",
       "      <td>22</td>\n",
       "      <td>164985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>89237</td>\n",
       "      <td>10</td>\n",
       "      <td>77189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>228416</td>\n",
       "      <td>267160</td>\n",
       "      <td>24</td>\n",
       "      <td>38744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>471226</td>\n",
       "      <td>471326</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>318047</td>\n",
       "      <td>471226</td>\n",
       "      <td>21</td>\n",
       "      <td>153179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4311</th>\n",
       "      <td>22</td>\n",
       "      <td>42973857</td>\n",
       "      <td>42980083</td>\n",
       "      <td>4</td>\n",
       "      <td>6226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4312</th>\n",
       "      <td>22</td>\n",
       "      <td>42973707</td>\n",
       "      <td>42973857</td>\n",
       "      <td>4</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4313</th>\n",
       "      <td>22</td>\n",
       "      <td>49013936</td>\n",
       "      <td>49016261</td>\n",
       "      <td>5</td>\n",
       "      <td>2325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4314</th>\n",
       "      <td>22</td>\n",
       "      <td>51223959</td>\n",
       "      <td>51224959</td>\n",
       "      <td>4</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4315</th>\n",
       "      <td>22</td>\n",
       "      <td>51201920</td>\n",
       "      <td>51223959</td>\n",
       "      <td>4</td>\n",
       "      <td>22039</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4316 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "| Chromosome   | Start     | End       | CNV_value   | Length_cnvr   |\n",
       "| (category)   | (int64)   | (int64)   | (int64)     | (int64)       |\n",
       "|--------------+-----------+-----------+-------------+---------------|\n",
       "| 1            | 12048     | 177033    | 22          | 164985        |\n",
       "| 1            | 12048     | 89237     | 10          | 77189         |\n",
       "| 1            | 228416    | 267160    | 24          | 38744         |\n",
       "| 1            | 471226    | 471326    | 10          | 100           |\n",
       "| ...          | ...       | ...       | ...         | ...           |\n",
       "| 22           | 42973707  | 42973857  | 4           | 150           |\n",
       "| 22           | 49013936  | 49016261  | 5           | 2325          |\n",
       "| 22           | 51223959  | 51224959  | 4           | 1000          |\n",
       "| 22           | 51201920  | 51223959  | 4           | 22039         |\n",
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "Unstranded PyRanges object has 4,316 rows and 5 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sudmant = pd.read_csv('/Users/luciabazan/Downloads/sudmant_data.csv')\n",
    "sudmant = sudmant[['contig','start','end','AMR_Maya_HGDP00855_F']]\n",
    "sudmant.columns =['Chromosome', 'Start', 'End', 'CNV_value']\n",
    "\n",
    "\n",
    "#sudmant['Chromosome'] = sudmant['Chromosome'].str.replace('chr','hs')\n",
    "n_uncalled = np.sum(sudmant==-1,1)\n",
    "sudmant = sudmant[n_uncalled==0]\n",
    "\n",
    "sudmant['Length_cnvr'] = sudmant['End'] - sudmant['Start'] \n",
    "sudmant = sudmant[~sudmant['CNV_value'].isin([2])]\n",
    "\n",
    "\n",
    "sudmant['Chromosome'] = sudmant['Chromosome'].str.replace('chr', '')\n",
    "sudmant = sudmant[sudmant['Chromosome'] != 'M']\n",
    "sudmant = sudmant[sudmant['Chromosome'] != 'X']\n",
    "sudmant = sudmant[sudmant['Chromosome'] != 'Y']\n",
    "\n",
    "\n",
    "grch37_deletions_sudmant = sudmant[sudmant['CNV_value'] < 3]\n",
    "grch37_deletions_sudmant_mean = grch37_deletions_sudmant['CNV_value'].mean()\n",
    "\n",
    "grch37_duplications_sudmant = sudmant[sudmant['CNV_value'] > 1]\n",
    "grch37_duplications_sudmant_mean = grch37_duplications_sudmant['CNV_value'].mean()\n",
    "\n",
    "grch37_duplications_sudmant = PyRanges(grch37_duplications_sudmant)\n",
    "grch37_deletions_sudmant = PyRanges(grch37_deletions_sudmant)\n",
    "\n",
    "\n",
    "sudmant = PyRanges(sudmant)\n",
    "sudmant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "a64eefc5-41c8-45f7-b64b-d46d3aa43daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genes\n",
    "peter_coordinates_genes = sudmant.nearest(grch37_genes_coordinates)\n",
    "peter_coordinates_genes = peter_coordinates_genes.as_df()\n",
    "peter_coordinates_genes = peter_coordinates_genes[peter_coordinates_genes['Distance'] <1 ]\n",
    "#grch37_coordinates_genes\n",
    "\n",
    "#Telomeres\n",
    "peter_coordinates_telomeres = sudmant.nearest(grch37_telomeres)\n",
    "peter_coordinates_telomeres = peter_coordinates_telomeres.as_df()\n",
    "peter_coordinates_telomeres = peter_coordinates_telomeres[peter_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs\n",
    "peter_coordinates_sds = sudmant.nearest(grch37_sds)\n",
    "peter_coordinates_sds = peter_coordinates_sds.as_df()\n",
    "peter_coordinates_sds = peter_coordinates_sds[peter_coordinates_sds['Distance'] <1 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "382fef50-c3b3-4737-8d72-1dc5078ba5c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>#\"bin\"</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>...</th>\n",
       "      <th>alignB</th>\n",
       "      <th>matchB</th>\n",
       "      <th>mismatchB</th>\n",
       "      <th>transitionsB</th>\n",
       "      <th>transversionsB</th>\n",
       "      <th>fracMatch</th>\n",
       "      <th>fracMatchIndel</th>\n",
       "      <th>jcK</th>\n",
       "      <th>k2K</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>177033</td>\n",
       "      <td>22</td>\n",
       "      <td>164985</td>\n",
       "      <td>585</td>\n",
       "      <td>10000</td>\n",
       "      <td>87112</td>\n",
       "      <td>chr15:102446355</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>74269</td>\n",
       "      <td>73742</td>\n",
       "      <td>527</td>\n",
       "      <td>332</td>\n",
       "      <td>195</td>\n",
       "      <td>0.992904</td>\n",
       "      <td>0.991956</td>\n",
       "      <td>0.007130</td>\n",
       "      <td>0.007133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>89237</td>\n",
       "      <td>10</td>\n",
       "      <td>77189</td>\n",
       "      <td>585</td>\n",
       "      <td>10000</td>\n",
       "      <td>87112</td>\n",
       "      <td>chr15:102446355</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>74269</td>\n",
       "      <td>73742</td>\n",
       "      <td>527</td>\n",
       "      <td>332</td>\n",
       "      <td>195</td>\n",
       "      <td>0.992904</td>\n",
       "      <td>0.991956</td>\n",
       "      <td>0.007130</td>\n",
       "      <td>0.007133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>228416</td>\n",
       "      <td>267160</td>\n",
       "      <td>24</td>\n",
       "      <td>38744</td>\n",
       "      <td>586</td>\n",
       "      <td>227417</td>\n",
       "      <td>238719</td>\n",
       "      <td>chr17:81176011</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>11249</td>\n",
       "      <td>11065</td>\n",
       "      <td>184</td>\n",
       "      <td>123</td>\n",
       "      <td>61</td>\n",
       "      <td>0.983643</td>\n",
       "      <td>0.981984</td>\n",
       "      <td>0.016538</td>\n",
       "      <td>0.016561</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>471226</td>\n",
       "      <td>471326</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "      <td>73</td>\n",
       "      <td>317719</td>\n",
       "      <td>471368</td>\n",
       "      <td>chr5:180744334</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>153555</td>\n",
       "      <td>153444</td>\n",
       "      <td>111</td>\n",
       "      <td>75</td>\n",
       "      <td>36</td>\n",
       "      <td>0.999277</td>\n",
       "      <td>0.999121</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>318047</td>\n",
       "      <td>471226</td>\n",
       "      <td>21</td>\n",
       "      <td>153179</td>\n",
       "      <td>73</td>\n",
       "      <td>317719</td>\n",
       "      <td>471368</td>\n",
       "      <td>chr5:180744334</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>153555</td>\n",
       "      <td>153444</td>\n",
       "      <td>111</td>\n",
       "      <td>75</td>\n",
       "      <td>36</td>\n",
       "      <td>0.999277</td>\n",
       "      <td>0.999121</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0.000723</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4290</th>\n",
       "      <td>22</td>\n",
       "      <td>42973857</td>\n",
       "      <td>42980083</td>\n",
       "      <td>4</td>\n",
       "      <td>6226</td>\n",
       "      <td>912</td>\n",
       "      <td>42962426</td>\n",
       "      <td>42978456</td>\n",
       "      <td>chr22:42902112</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>13422</td>\n",
       "      <td>13031</td>\n",
       "      <td>391</td>\n",
       "      <td>245</td>\n",
       "      <td>146</td>\n",
       "      <td>0.970869</td>\n",
       "      <td>0.967122</td>\n",
       "      <td>0.029712</td>\n",
       "      <td>0.029771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4291</th>\n",
       "      <td>22</td>\n",
       "      <td>42973707</td>\n",
       "      <td>42973857</td>\n",
       "      <td>4</td>\n",
       "      <td>150</td>\n",
       "      <td>912</td>\n",
       "      <td>42962426</td>\n",
       "      <td>42978456</td>\n",
       "      <td>chr22:42902112</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>13422</td>\n",
       "      <td>13031</td>\n",
       "      <td>391</td>\n",
       "      <td>245</td>\n",
       "      <td>146</td>\n",
       "      <td>0.970869</td>\n",
       "      <td>0.967122</td>\n",
       "      <td>0.029712</td>\n",
       "      <td>0.029771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4292</th>\n",
       "      <td>22</td>\n",
       "      <td>49013936</td>\n",
       "      <td>49016261</td>\n",
       "      <td>5</td>\n",
       "      <td>2325</td>\n",
       "      <td>958</td>\n",
       "      <td>49014584</td>\n",
       "      <td>49015948</td>\n",
       "      <td>chr15:22514585</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1331</td>\n",
       "      <td>1270</td>\n",
       "      <td>61</td>\n",
       "      <td>44</td>\n",
       "      <td>17</td>\n",
       "      <td>0.954170</td>\n",
       "      <td>0.949888</td>\n",
       "      <td>0.047290</td>\n",
       "      <td>0.047556</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4293</th>\n",
       "      <td>22</td>\n",
       "      <td>51223959</td>\n",
       "      <td>51224959</td>\n",
       "      <td>4</td>\n",
       "      <td>1000</td>\n",
       "      <td>975</td>\n",
       "      <td>51198214</td>\n",
       "      <td>51236901</td>\n",
       "      <td>chr2:114370005</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>38593</td>\n",
       "      <td>38225</td>\n",
       "      <td>368</td>\n",
       "      <td>242</td>\n",
       "      <td>126</td>\n",
       "      <td>0.990465</td>\n",
       "      <td>0.989260</td>\n",
       "      <td>0.009597</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4294</th>\n",
       "      <td>22</td>\n",
       "      <td>51201920</td>\n",
       "      <td>51223959</td>\n",
       "      <td>4</td>\n",
       "      <td>22039</td>\n",
       "      <td>975</td>\n",
       "      <td>51198214</td>\n",
       "      <td>51236901</td>\n",
       "      <td>chr2:114370005</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>38593</td>\n",
       "      <td>38225</td>\n",
       "      <td>368</td>\n",
       "      <td>242</td>\n",
       "      <td>126</td>\n",
       "      <td>0.990465</td>\n",
       "      <td>0.989260</td>\n",
       "      <td>0.009597</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3704 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome     Start       End  CNV_value  Length_cnvr  #\"bin\"   Start_b  \\\n",
       "0             1     12048    177033         22       164985     585     10000   \n",
       "1             1     12048     89237         10        77189     585     10000   \n",
       "2             1    228416    267160         24        38744     586    227417   \n",
       "3             1    471226    471326         10          100      73    317719   \n",
       "4             1    318047    471226         21       153179      73    317719   \n",
       "...         ...       ...       ...        ...          ...     ...       ...   \n",
       "4290         22  42973857  42980083          4         6226     912  42962426   \n",
       "4291         22  42973707  42973857          4          150     912  42962426   \n",
       "4292         22  49013936  49016261          5         2325     958  49014584   \n",
       "4293         22  51223959  51224959          4         1000     975  51198214   \n",
       "4294         22  51201920  51223959          4        22039     975  51198214   \n",
       "\n",
       "         End_b             name  score  ...  alignB  matchB  mismatchB  \\\n",
       "0        87112  chr15:102446355      0  ...   74269   73742        527   \n",
       "1        87112  chr15:102446355      0  ...   74269   73742        527   \n",
       "2       238719   chr17:81176011      0  ...   11249   11065        184   \n",
       "3       471368   chr5:180744334      0  ...  153555  153444        111   \n",
       "4       471368   chr5:180744334      0  ...  153555  153444        111   \n",
       "...        ...              ...    ...  ...     ...     ...        ...   \n",
       "4290  42978456   chr22:42902112      0  ...   13422   13031        391   \n",
       "4291  42978456   chr22:42902112      0  ...   13422   13031        391   \n",
       "4292  49015948   chr15:22514585      0  ...    1331    1270         61   \n",
       "4293  51236901   chr2:114370005      0  ...   38593   38225        368   \n",
       "4294  51236901   chr2:114370005      0  ...   38593   38225        368   \n",
       "\n",
       "      transitionsB  transversionsB  fracMatch  fracMatchIndel       jcK  \\\n",
       "0              332             195   0.992904        0.991956  0.007130   \n",
       "1              332             195   0.992904        0.991956  0.007130   \n",
       "2              123              61   0.983643        0.981984  0.016538   \n",
       "3               75              36   0.999277        0.999121  0.000723   \n",
       "4               75              36   0.999277        0.999121  0.000723   \n",
       "...            ...             ...        ...             ...       ...   \n",
       "4290           245             146   0.970869        0.967122  0.029712   \n",
       "4291           245             146   0.970869        0.967122  0.029712   \n",
       "4292            44              17   0.954170        0.949888  0.047290   \n",
       "4293           242             126   0.990465        0.989260  0.009597   \n",
       "4294           242             126   0.990465        0.989260  0.009597   \n",
       "\n",
       "           k2K  Distance  \n",
       "0     0.007133         0  \n",
       "1     0.007133         0  \n",
       "2     0.016561         0  \n",
       "3     0.000723         0  \n",
       "4     0.000723         0  \n",
       "...        ...       ...  \n",
       "4290  0.029771         0  \n",
       "4291  0.029771         0  \n",
       "4292  0.047556         0  \n",
       "4293  0.009604         0  \n",
       "4294  0.009604         0  \n",
       "\n",
       "[3704 rows x 35 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peter_coordinates_sds#[['Chromosome','Start', 'End']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d720a419-476a-4e26-8d5c-9225f2ce38d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "3e2e6080-b29f-4c35-b894-00ac5173dcef",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Deletions\n",
    "#### Genes \n",
    "deletions_grch37_sudmant_coordinates_genes = grch37_deletions_sudmant.nearest(grch37_genes_coordinates)\n",
    "deletions_grch37_sudmant_coordinates_genes = deletions_grch37_sudmant_coordinates_genes.as_df()\n",
    "deletions_grch37_sudmant_coordinates_genes = deletions_grch37_sudmant_coordinates_genes[deletions_grch37_sudmant_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_sudmant_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_sudmant_mean_individual_genes= deletions_grch37_sudmant_coordinates_genes['CNV_value'].mean()\n",
    "deletions_grch37_sudmant_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "deletions_grch37_sudmant_coordinates_telomeres = grch37_deletions_sudmant.nearest(grch37_telomeres)\n",
    "deletions_grch37_sudmant_coordinates_telomeres = deletions_grch37_sudmant_coordinates_telomeres.as_df()\n",
    "deletions_grch37_sudmant_coordinates_telomeres = deletions_grch37_sudmant_coordinates_telomeres[deletions_grch37_sudmant_coordinates_telomeres['Distance'] <1 ]\n",
    "deletions_grch37_sudmant_coordinates_telomeres\n",
    "# Mean CN \n",
    "deletions_grch37_sudmant_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_sudmant_mean_individual_telomeres= deletions_grch37_sudmant_coordinates_telomeres['CNV_value'].mean()\n",
    "deletions_grch37_sudmant_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "deletions_grch37_sudmant_coordinates_sds = grch37_deletions_sudmant.nearest(grch37_sds)\n",
    "deletions_grch37_sudmant_coordinates_sds = deletions_grch37_sudmant_coordinates_sds.as_df()\n",
    "deletions_grch37_sudmant_coordinates_sds = deletions_grch37_sudmant_coordinates_sds[deletions_grch37_sudmant_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "deletions_grch37_sudmant_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "deletions_grch37_sudmant_mean_individual_sds= deletions_grch37_sudmant_coordinates_sds['CNV_value'].mean()\n",
    "deletions_grch37_sudmant_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "1f5d27a5-ae43-417b-960e-42e27ce8f2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Duplications\n",
    "#### Genes \n",
    "duplications_grch37_sudmant_coordinates_genes = grch37_duplications_sudmant.nearest(grch37_genes_coordinates)\n",
    "duplications_grch37_sudmant_coordinates_genes = duplications_grch37_sudmant_coordinates_genes.as_df()\n",
    "duplications_grch37_sudmant_coordinates_genes = duplications_grch37_sudmant_coordinates_genes[duplications_grch37_sudmant_coordinates_genes['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_sudmant_coordinates_genes.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_sudmant_mean_individual_genes= duplications_grch37_sudmant_coordinates_genes['CNV_value'].mean()\n",
    "duplications_grch37_sudmant_coordinates_genes.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### Telomeres\n",
    "duplications_grch37_sudmant_coordinates_telomeres = grch37_duplications_sudmant.nearest(grch37_telomeres)\n",
    "duplications_grch37_sudmant_coordinates_telomeres = duplications_grch37_sudmant_coordinates_telomeres.as_df()\n",
    "duplications_grch37_sudmant_coordinates_telomeres = duplications_grch37_sudmant_coordinates_telomeres[duplications_grch37_sudmant_coordinates_telomeres['Distance'] <1 ]\n",
    "duplications_grch37_sudmant_coordinates_telomeres\n",
    "# Mean CN \n",
    "duplications_grch37_sudmant_coordinates_telomeres.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_sudmant_mean_individual_telomeres= duplications_grch37_sudmant_coordinates_telomeres['CNV_value'].mean()\n",
    "duplications_grch37_sudmant_coordinates_telomeres.replace(np.nan,2, inplace=True)\n",
    "\n",
    "\n",
    "#### SDs \n",
    "duplications_grch37_sudmant_coordinates_sds = grch37_duplications_sudmant.nearest(grch37_sds)\n",
    "duplications_grch37_sudmant_coordinates_sds = duplications_grch37_sudmant_coordinates_sds.as_df()\n",
    "duplications_grch37_sudmant_coordinates_sds = duplications_grch37_sudmant_coordinates_sds[duplications_grch37_sudmant_coordinates_sds['Distance'] <1 ]\n",
    "# Mean CN \n",
    "duplications_grch37_sudmant_coordinates_sds.replace(2, np.nan, inplace=True)\n",
    "duplications_grch37_sudmant_mean_individual_sds= duplications_grch37_sudmant_coordinates_sds['CNV_value'].mean()\n",
    "duplications_grch37_sudmant_coordinates_sds.replace(np.nan,2, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d598726-8adc-4696-b5e0-0c4267040c6e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "90789782-536a-4f10-b4b5-60e5d62a4cee",
   "metadata": {},
   "source": [
    "GRCh37 Duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "b2db73c9-02d8-47e0-a506-82ac31c4ba7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Literature</th>\n",
       "      <th>Mean size(Bp) Literature</th>\n",
       "      <th>Literature  Deletions</th>\n",
       "      <th>Literature Mean CN Deletions</th>\n",
       "      <th>Literature  Duplications</th>\n",
       "      <th>Literature Mean CN Duplications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>4316</td>\n",
       "      <td>32098.0</td>\n",
       "      <td>333</td>\n",
       "      <td>0.726727</td>\n",
       "      <td>3983</td>\n",
       "      <td>10.283204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>3087</td>\n",
       "      <td>39603.0</td>\n",
       "      <td>180</td>\n",
       "      <td>0.738889</td>\n",
       "      <td>2907</td>\n",
       "      <td>10.553492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>3704</td>\n",
       "      <td>36046.0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.850000</td>\n",
       "      <td>3684</td>\n",
       "      <td>10.779045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>92834</td>\n",
       "      <td>4636.0</td>\n",
       "      <td>7174</td>\n",
       "      <td>0.980485</td>\n",
       "      <td>85660</td>\n",
       "      <td>31.152393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>63950</td>\n",
       "      <td>3086.0</td>\n",
       "      <td>4179</td>\n",
       "      <td>0.986360</td>\n",
       "      <td>59771</td>\n",
       "      <td>20.399742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>43</td>\n",
       "      <td>126802.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43</td>\n",
       "      <td>37.186047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>66176</td>\n",
       "      <td>5227.0</td>\n",
       "      <td>51</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66125</td>\n",
       "      <td>38.810541</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data  Literature  Mean size(Bp) Literature  Literature  Deletions  \\\n",
       "0       GRCh37        4316                   32098.0                    333   \n",
       "1        Genes        3087                   39603.0                    180   \n",
       "2    Telomeres           0                       NaN                      0   \n",
       "3          SDs        3704                   36046.0                     20   \n",
       "4    T2T-CHM13       92834                    4636.0                   7174   \n",
       "5        Genes       63950                    3086.0                   4179   \n",
       "6    Telomeres          43                  126802.0                      0   \n",
       "7          SDs       66176                    5227.0                     51   \n",
       "\n",
       "   Literature Mean CN Deletions  Literature  Duplications  \\\n",
       "0                      0.726727                      3983   \n",
       "1                      0.738889                      2907   \n",
       "2                           NaN                         0   \n",
       "3                      0.850000                      3684   \n",
       "4                      0.980485                     85660   \n",
       "5                      0.986360                     59771   \n",
       "6                           NaN                        43   \n",
       "7                      1.000000                     66125   \n",
       "\n",
       "   Literature Mean CN Duplications  \n",
       "0                        10.283204  \n",
       "1                        10.553492  \n",
       "2                              NaN  \n",
       "3                        10.779045  \n",
       "4                        31.152393  \n",
       "5                        20.399742  \n",
       "6                        37.186047  \n",
       "7                        38.810541  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "literature = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'\n",
    "             ], \n",
    "     'Literature':[len(sudmant), \n",
    "                  len(peter_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(peter_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(peter_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(final_cnv_ranges), \n",
    "                  len(coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), ], \n",
    "     'Mean size(Bp) Literature':[round(sudmant.as_df()['Length_cnvr'].mean()), \n",
    "                      round(peter_coordinates_genes['Length_cnvr'].mean()), \n",
    "                      peter_coordinates_telomeres['Length_cnvr'].mean(), \n",
    "                      round(peter_coordinates_sds['Length_cnvr'].mean()), \n",
    "                      round(final_cnv['Length_cnvr'].mean()),\n",
    "                      round(coordinates_genes['Length_cnvr'].mean()),\n",
    "                      round(coordinates_telomeres['Length_cnvr'].mean()), \n",
    "                      round(coordinates_sds['Length_cnvr'].mean())],\n",
    "    'Literature  Deletions':[len(grch37_deletions_sudmant), \n",
    "                              len(deletions_grch37_sudmant_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(deletions_grch37_sudmant_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_grch37_sudmant_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_deletions_gb),\n",
    "                                     len(deletions_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "     \n",
    "     'Literature Mean CN Deletions':[grch37_deletions_sudmant_mean, \n",
    "                                     deletions_grch37_sudmant_mean_individual_genes, \n",
    "                                     deletions_grch37_sudmant_mean_individual_telomeres,\n",
    "                                     deletions_grch37_sudmant_mean_individual_sds,\n",
    "                                     t2t_deletions_gb_mean,\n",
    "                                     deletions_t2t_mean_individual_genes,\n",
    "                                     deletions_t2t_mean_individual_telomeres,\n",
    "                                     deletions_t2t_mean_individual_sds\n",
    "                                    ],\n",
    "\n",
    "     \n",
    "     'Literature  Duplications':[len(grch37_duplications_sudmant), \n",
    "                              len(duplications_grch37_sudmant_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(duplications_grch37_sudmant_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_grch37_sudmant_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_duplications_gb),\n",
    "                                     len(duplications_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "     'Literature Mean CN Duplications':[grch37_duplications_sudmant_mean,\n",
    "                                       duplications_grch37_sudmant_mean_individual_genes,\n",
    "                                       duplications_grch37_sudmant_mean_individual_telomeres,\n",
    "                                       duplications_grch37_sudmant_mean_individual_sds,\n",
    "                                       t2t_duplications_gb_mean,\n",
    "                                        duplications_t2t_mean_individual_genes,\n",
    "                                        duplications_t2t_mean_individual_telomeres,\n",
    "                                        duplications_t2t_mean_individual_sds],\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=literature)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "ed1f8e95-2418-464b-9402-cfa0f13215f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>CNVNator</th>\n",
       "      <th>Mean size(Bp)</th>\n",
       "      <th>Deletions</th>\n",
       "      <th>CNVnator Mean CN Deletions</th>\n",
       "      <th>Duplications</th>\n",
       "      <th>CNVnator Mean CN Duplications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>728</td>\n",
       "      <td>18512.0</td>\n",
       "      <td>508</td>\n",
       "      <td>0.773704</td>\n",
       "      <td>220</td>\n",
       "      <td>4.375335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>383</td>\n",
       "      <td>23790.0</td>\n",
       "      <td>272</td>\n",
       "      <td>0.786549</td>\n",
       "      <td>111</td>\n",
       "      <td>4.271579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>3</td>\n",
       "      <td>47899.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.099335</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>430</td>\n",
       "      <td>27020.0</td>\n",
       "      <td>234</td>\n",
       "      <td>0.852147</td>\n",
       "      <td>196</td>\n",
       "      <td>4.463832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>1855</td>\n",
       "      <td>66703.0</td>\n",
       "      <td>1420</td>\n",
       "      <td>0.894372</td>\n",
       "      <td>435</td>\n",
       "      <td>3.969565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>386</td>\n",
       "      <td>64483.0</td>\n",
       "      <td>224</td>\n",
       "      <td>0.997807</td>\n",
       "      <td>162</td>\n",
       "      <td>3.344081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>5</td>\n",
       "      <td>101599.0</td>\n",
       "      <td>3</td>\n",
       "      <td>1.049467</td>\n",
       "      <td>2</td>\n",
       "      <td>4.295300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>569</td>\n",
       "      <td>59707.0</td>\n",
       "      <td>347</td>\n",
       "      <td>0.981064</td>\n",
       "      <td>222</td>\n",
       "      <td>3.593323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data  CNVNator  Mean size(Bp)  Deletions  \\\n",
       "0       GRCh37       728        18512.0        508   \n",
       "1        Genes       383        23790.0        272   \n",
       "2    Telomeres         3        47899.0          3   \n",
       "3          SDs       430        27020.0        234   \n",
       "4    T2T-CHM13      1855        66703.0       1420   \n",
       "5        Genes       386        64483.0        224   \n",
       "6    Telomeres         5       101599.0          3   \n",
       "7          SDs       569        59707.0        347   \n",
       "\n",
       "   CNVnator Mean CN Deletions  Duplications  CNVnator Mean CN Duplications  \n",
       "0                    0.773704           220                       4.375335  \n",
       "1                    0.786549           111                       4.271579  \n",
       "2                    0.099335             0                            NaN  \n",
       "3                    0.852147           196                       4.463832  \n",
       "4                    0.894372           435                       3.969565  \n",
       "5                    0.997807           162                       3.344081  \n",
       "6                    1.049467             2                       4.295300  \n",
       "7                    0.981064           222                       3.593323  "
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'\n",
    "             ], \n",
    "     'CNVNator': [len(final_cnv_grch37), \n",
    "               len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),  \n",
    "               len(t2t_cnvnator), \n",
    "               len(t2t_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(t2t_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(t2t_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "              ], \n",
    "     'Mean size(Bp)':[final_cnv_grch37['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_genes['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_sds['Length_cnvr'].mean().round(), \n",
    "                      t2t_cnvnator.as_df()['Length_cnvr'].mean().round(),\n",
    "                      t2t_cnvnator_coordinates_genes['Length_cnvr'].mean().round(),\n",
    "                      t2t_cnvnator_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      t2t_cnvnator_coordinates_sds['Length_cnvr'].mean().round()],\n",
    "     \n",
    "     'Deletions':[len(grch37_deletions_cnvnator), \n",
    "                              len(deletions_grch37_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(deletions_grch37_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_grch37_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_deletions_cnvnator),\n",
    "                                     len(deletions_t2t_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(deletions_t2t_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "\n",
    "     'CNVnator Mean CN Deletions':[grch37_deletions_cnvnator_mean,\n",
    "                                   deletions_grch37_cnvnator_mean_individual_genes,\n",
    "                                   deletions_grch37_cnvnator_mean_individual_telomeres,\n",
    "                                   deletions_grch37_cnvnator_mean_individual_sds,\n",
    "                                   t2t_deletions_cnvnator_mean, \n",
    "                                   deletions_t2t_cnvnator_mean_individual_genes,\n",
    "                                   deletions_t2t_cnvnator_mean_individual_telomeres,\n",
    "                                   deletions_t2t_cnvnator_mean_individual_sds],\n",
    "\n",
    "    'Duplications':[len(grch37_duplications_cnvnator), \n",
    "                              len(duplications_grch37_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                                     len(duplications_grch37_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_grch37_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(t2t_duplications_cnvnator),\n",
    "                                     len(duplications_t2t_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                                     len(duplications_t2t_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())],\n",
    "\n",
    "     'CNVnator Mean CN Duplications':[grch37_duplications_cnvnator_mean,\n",
    "                                      duplications_grch37_cnvnator_mean_individual_genes,\n",
    "                                      duplications_grch37_cnvnator_mean_individual_telomeres,\n",
    "                                      duplications_grch37_cnvnator_mean_individual_sds,\n",
    "                                      t2t_duplications_cnvnator_mean,\n",
    "                                      duplications_t2t_cnvnator_mean_individual_genes,\n",
    "                                      duplications_t2t_cnvnator_mean_individual_telomeres,\n",
    "                                      duplications_t2t_cnvnator_mean_individual_sds\n",
    "                                      \n",
    "                                     ],\n",
    "\n",
    "\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=d)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "95d6753a-e4b7-4fb0-b072-1f7647bae1a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3704"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(peter_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "800cc5c7-e145-4663-b41e-bf515f7f3615",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
