{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "00f332d4-1c1c-42c7-8382-7cb5fcab05d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyranges import PyRanges\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b272d5b-706a-4c50-9f84-cc3e31f479e3",
   "metadata": {},
   "source": [
    "<h1> T2T CHM13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "6b0442f5-74b7-4d26-9bc8-dab9a2b8ff71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_22719/1889853910.py:2: DtypeWarning: Columns (23,36) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  t2t_genes = pd.read_csv('../annotation_files/T2T-chm13-catLiftOffGenes.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3</td>\n",
       "      <td>140066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16</td>\n",
       "      <td>13731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16</td>\n",
       "      <td>198196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83511</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62453929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83512</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62454747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83513</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62456950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83514</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62457871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83515</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62445192</td>\n",
       "      <td>62458697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83516 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End\n",
       "0           chr1         1     13731\n",
       "1           chr1         3    140066\n",
       "2           chr1        16     13731\n",
       "3           chr1        16     13731\n",
       "4           chr1        16    198196\n",
       "...          ...       ...       ...\n",
       "83511       chrY  62445192  62453929\n",
       "83512       chrY  62445192  62454747\n",
       "83513       chrY  62445192  62456950\n",
       "83514       chrY  62445192  62457871\n",
       "83515       chrY  62445192  62458697\n",
       "\n",
       "[83516 rows x 3 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Anotations\n",
    "t2t_genes = pd.read_csv('../annotation_files/T2T-chm13-catLiftOffGenes.csv')\n",
    "t2t_telomeres = pd.read_csv('../annotation_files/chm13v2.0_telomere.bed', sep='\\t', names=['Chromosome','Start', 'End'])\n",
    "t2t_sds = pd.read_csv('../annotation_files/chm13v2.0_SD.full.bed', sep='\\t')\n",
    "t2t_sds = t2t_sds[['#chr1', 'start1', 'end1']]\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "t2t_sds.columns =['Chromosome','Start','End']\n",
    "t2t_sds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "d730b68f-174e-41e2-ac7a-b23e1aa2e6eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/05/4pfxtbrs5459ytz3j1jhcgt00000gn/T/ipykernel_22719/2170477776.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  chm13 = pd.read_csv('../call_cnvr_chm13/individual_cnv_cnvr.txt', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "# is not working \n",
    "chm13 = pd.read_csv('../call_cnvr_chm13/individual_cnv_cnvr.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "55cc3484-958c-4a32-94f5-7b0e88698454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNVR_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>CNV_Start</th>\n",
       "      <th>CNV_End</th>\n",
       "      <th>CNV_Value</th>\n",
       "      <th>Length</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1508570</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8285141</td>\n",
       "      <td>8289096</td>\n",
       "      <td>1</td>\n",
       "      <td>3956</td>\n",
       "      <td>12764469.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508571</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8292271</td>\n",
       "      <td>8294316</td>\n",
       "      <td>1</td>\n",
       "      <td>2046</td>\n",
       "      <td>12764469.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508572</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8294316</td>\n",
       "      <td>8297751</td>\n",
       "      <td>1</td>\n",
       "      <td>3436</td>\n",
       "      <td>12764469.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508573</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8297751</td>\n",
       "      <td>8303368</td>\n",
       "      <td>1</td>\n",
       "      <td>5618</td>\n",
       "      <td>12764469.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508574</th>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>CNVR_109274</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>8304488</td>\n",
       "      <td>8306331</td>\n",
       "      <td>1</td>\n",
       "      <td>1844</td>\n",
       "      <td>12764469.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14499930</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248364795</td>\n",
       "      <td>248366351</td>\n",
       "      <td>26</td>\n",
       "      <td>1557</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14499931</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248366351</td>\n",
       "      <td>248367622</td>\n",
       "      <td>27</td>\n",
       "      <td>1272</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14499932</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248367622</td>\n",
       "      <td>248368745</td>\n",
       "      <td>27</td>\n",
       "      <td>1124</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14499933</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248368745</td>\n",
       "      <td>248370339</td>\n",
       "      <td>42</td>\n",
       "      <td>1595</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14499934</th>\n",
       "      <td>1</td>\n",
       "      <td>248304532.0</td>\n",
       "      <td>248387328.0</td>\n",
       "      <td>CNVR_11357</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>248370339</td>\n",
       "      <td>248387328</td>\n",
       "      <td>37</td>\n",
       "      <td>16990</td>\n",
       "      <td>82796.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92810 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chr        Start          End      CNVR_ID          Sample_ID  \\\n",
       "1508570   22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "1508571   22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "1508572   22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "1508573   22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "1508574   22          0.0   12764469.0  CNVR_109274  LP6005441-DNA_G07   \n",
       "...       ..          ...          ...          ...                ...   \n",
       "14499930   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "14499931   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "14499932   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "14499933   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "14499934   1  248304532.0  248387328.0   CNVR_11357  LP6005441-DNA_G07   \n",
       "\n",
       "          CNV_Start    CNV_End  CNV_Value  Length  Length_cnvr  \n",
       "1508570     8285141    8289096          1    3956   12764469.0  \n",
       "1508571     8292271    8294316          1    2046   12764469.0  \n",
       "1508572     8294316    8297751          1    3436   12764469.0  \n",
       "1508573     8297751    8303368          1    5618   12764469.0  \n",
       "1508574     8304488    8306331          1    1844   12764469.0  \n",
       "...             ...        ...        ...     ...          ...  \n",
       "14499930  248364795  248366351         26    1557      82796.0  \n",
       "14499931  248366351  248367622         27    1272      82796.0  \n",
       "14499932  248367622  248368745         27    1124      82796.0  \n",
       "14499933  248368745  248370339         42    1595      82796.0  \n",
       "14499934  248370339  248387328         37   16990      82796.0  \n",
       "\n",
       "[92810 rows x 10 columns]"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13 = chm13.dropna()\n",
    "chm13['Length_cnvr'] = chm13['End'] - chm13['Start']\n",
    "\n",
    "\n",
    "chm13 = chm13[chm13['Sample_ID'] == 'LP6005441-DNA_G07']\n",
    "\n",
    "chm13 = chm13[~chm13['CNV_Value'].isin([2])]\n",
    "chm13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "55b427d6-ec1f-4c01-8621-426f5f660809",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Keeping only samples that are on old cnvs \n",
    "chm13 = chm13.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "ac0ab3f0-db9c-4b57-a3aa-06fdecb87f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "#REMOVE SAMPLES WITH ABERRATIONS\n",
    "\n",
    "chm13#.value_counts(by=[\"Sample_ID\"])\n",
    "chm13.columns = ['Chromosome', 'Start', 'End', 'CNVR_ID', 'Sample_ID', 'CNV_start', 'CNV_end', 'CNV_Value','Length', 'Length_cnvr','Sample','Region']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "a4ff43d7-b75b-406c-8601-e49e532c98b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>891494.0</td>\n",
       "      <td>891494.0</td>\n",
       "      <td>5.377358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>900018.0</td>\n",
       "      <td>1000503.0</td>\n",
       "      <td>100485.0</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1575421.0</td>\n",
       "      <td>1594102.0</td>\n",
       "      <td>18681.0</td>\n",
       "      <td>19.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1758355.0</td>\n",
       "      <td>1945034.0</td>\n",
       "      <td>186679.0</td>\n",
       "      <td>31.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>2268719.0</td>\n",
       "      <td>2287580.0</td>\n",
       "      <td>18861.0</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14867</th>\n",
       "      <td>9</td>\n",
       "      <td>148771046.0</td>\n",
       "      <td>149276995.0</td>\n",
       "      <td>505949.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14868</th>\n",
       "      <td>9</td>\n",
       "      <td>149278663.0</td>\n",
       "      <td>149484014.0</td>\n",
       "      <td>205351.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14869</th>\n",
       "      <td>9</td>\n",
       "      <td>149620306.0</td>\n",
       "      <td>149794372.0</td>\n",
       "      <td>174066.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14870</th>\n",
       "      <td>9</td>\n",
       "      <td>149803678.0</td>\n",
       "      <td>149871435.0</td>\n",
       "      <td>67757.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14871</th>\n",
       "      <td>9</td>\n",
       "      <td>150171200.0</td>\n",
       "      <td>150617247.0</td>\n",
       "      <td>446047.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14872 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome        Start          End  Length_cnvr  LP6005441-DNA_G07\n",
       "0               3          0.0     891494.0     891494.0           5.377358\n",
       "1               3     900018.0    1000503.0     100485.0           6.000000\n",
       "2               3    1575421.0    1594102.0      18681.0          19.000000\n",
       "3               3    1758355.0    1945034.0     186679.0          31.000000\n",
       "4               3    2268719.0    2287580.0      18861.0           3.000000\n",
       "...           ...          ...          ...          ...                ...\n",
       "14867           9  148771046.0  149276995.0     505949.0           1.000000\n",
       "14868           9  149278663.0  149484014.0     205351.0           1.000000\n",
       "14869           9  149620306.0  149794372.0     174066.0           1.000000\n",
       "14870           9  149803678.0  149871435.0      67757.0           1.000000\n",
       "14871           9  150171200.0  150617247.0     446047.0           1.000000\n",
       "\n",
       "[14872 rows x 5 columns]"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv = chm13.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='CNV_Value',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "final_cnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "c3dd5fce-15e1-4360-ab58-17d628a16655",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "final_cnv_ranges = PyRanges(final_cnv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "1ebf6094-c3e2-453d-b1f9-3b733adfb293",
   "metadata": {},
   "outputs": [],
   "source": [
    "t2t_genes.columns = ['Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'thickStart', 'thickEnd', 'reserved', 'blockCount', 'blockSizes',\n",
    "       'chromStarts', 'name2', 'cdsStartStat', 'cdsEndStat', 'exonFrames',\n",
    "       'txId', 'type', 'geneName', 'geneType', 'sourceGene',\n",
    "       'sourceTranscript', 'alignmentId', 'alternativeSourceTranscripts',\n",
    "       'Paralogy', 'UnfilteredParalogy', 'collapsedGeneIds',\n",
    "       'collapsedGeneNames', 'frameshift', 'exonAnnotationSupport',\n",
    "       'intronAnnotationSupport', 'transcriptClass', 'transcriptModes',\n",
    "       'validStart', 'validStop', 'properOrf', 'extra_paralog']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "270bf858-3acc-485c-90d3-88cb319bd888",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "7ab381cf-130e-4900-b541-f7756d20c889",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "t2t_genes['Chromosome'] = t2t_genes['Chromosome'].str.replace('chr', '')\n",
    "t2t_genes = PyRanges(t2t_genes)\n",
    "\n",
    "t2t_telomeres['Chromosome'] = t2t_telomeres['Chromosome'].str.replace('chr', '')\n",
    "t2t_telomeres = PyRanges(t2t_telomeres)\n",
    "\n",
    "t2t_sds['Chromosome'] = t2t_sds['Chromosome'].str.replace('chr', '')\n",
    "t2t_sds = PyRanges(t2t_sds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "6679ce82-0b3c-45fd-97d4-3fced891d9c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>403567</td>\n",
       "      <td>403567.0</td>\n",
       "      <td>17.124031</td>\n",
       "      <td>1</td>\n",
       "      <td>13731</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>453948</td>\n",
       "      <td>466894</td>\n",
       "      <td>12946.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>459773</td>\n",
       "      <td>460982</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>518975</td>\n",
       "      <td>605035</td>\n",
       "      <td>86060.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>567686</td>\n",
       "      <td>568750</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>700181</td>\n",
       "      <td>782637</td>\n",
       "      <td>82456.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>717376</td>\n",
       "      <td>721103</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>878563</td>\n",
       "      <td>973010</td>\n",
       "      <td>94447.0</td>\n",
       "      <td>4.025000</td>\n",
       "      <td>900919</td>\n",
       "      <td>924969</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14867</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>139.091432</td>\n",
       "      <td>4</td>\n",
       "      <td>293068</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14868</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>45090682</td>\n",
       "      <td>45090682.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>293068</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14869</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>4.877301</td>\n",
       "      <td>15712005</td>\n",
       "      <td>15938486</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14870</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12764469</td>\n",
       "      <td>12764469.0</td>\n",
       "      <td>137.948830</td>\n",
       "      <td>1</td>\n",
       "      <td>66809</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14871</th>\n",
       "      <td>22</td>\n",
       "      <td>15787747</td>\n",
       "      <td>51324926</td>\n",
       "      <td>35537179.0</td>\n",
       "      <td>9.741935</td>\n",
       "      <td>15712005</td>\n",
       "      <td>15938486</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3703 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End  Length_cnvr  LP6005441-DNA_G07  \\\n",
       "0              1         0    403567     403567.0          17.124031   \n",
       "1              1    453948    466894      12946.0           1.000000   \n",
       "2              1    518975    605035      86060.0           3.000000   \n",
       "3              1    700181    782637      82456.0           1.000000   \n",
       "4              1    878563    973010      94447.0           4.025000   \n",
       "...          ...       ...       ...          ...                ...   \n",
       "14867         21         0  45090682   45090682.0         139.091432   \n",
       "14868         21         0  45090682   45090682.0           1.000000   \n",
       "14869         22  15787747  51324926   35537179.0           4.877301   \n",
       "14870         22         0  12764469   12764469.0         137.948830   \n",
       "14871         22  15787747  51324926   35537179.0           9.741935   \n",
       "\n",
       "        Start_b     End_b  Distance  \n",
       "0             1     13731         0  \n",
       "1        459773    460982         0  \n",
       "2        567686    568750         0  \n",
       "3        717376    721103         0  \n",
       "4        900919    924969         0  \n",
       "...         ...       ...       ...  \n",
       "14867         4    293068         0  \n",
       "14868         4    293068         0  \n",
       "14869  15712005  15938486         0  \n",
       "14870         1     66809         0  \n",
       "14871  15712005  15938486         0  \n",
       "\n",
       "[3703 rows x 8 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### t2t genes \n",
    "coordinates_genes = final_cnv_ranges.nearest(t2t_genes)\n",
    "coordinates_genes = coordinates_genes.as_df()\n",
    "coordinates_genes = coordinates_genes[coordinates_genes['Distance'] <1 ]\n",
    "\n",
    "## T2T telomeres\n",
    "coordinates_telomeres = final_cnv_ranges.nearest(t2t_telomeres)\n",
    "coordinates_telomeres = coordinates_telomeres.as_df()\n",
    "coordinates_telomeres = coordinates_telomeres[coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs \n",
    "coordinates_sds = final_cnv_ranges.nearest(t2t_sds)\n",
    "coordinates_sds = coordinates_sds.as_df()\n",
    "coordinates_sds = coordinates_sds[coordinates_sds['Distance'] <1 ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccab1fb8-1a97-43df-9041-235962b1cff4",
   "metadata": {},
   "source": [
    "<h2> T2T CNVnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "d68610aa-58f1-4657-9dde-03ef84a89829",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>25001</td>\n",
       "      <td>45000</td>\n",
       "      <td>3.3208</td>\n",
       "      <td>19999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>48001</td>\n",
       "      <td>78000</td>\n",
       "      <td>3.2608</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>88001</td>\n",
       "      <td>96000</td>\n",
       "      <td>0.5038</td>\n",
       "      <td>7999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2133001</td>\n",
       "      <td>2162000</td>\n",
       "      <td>7.1520</td>\n",
       "      <td>28999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2193001</td>\n",
       "      <td>2223000</td>\n",
       "      <td>5.9894</td>\n",
       "      <td>29999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>Y</td>\n",
       "      <td>2253001</td>\n",
       "      <td>5216000</td>\n",
       "      <td>0.1908</td>\n",
       "      <td>2962999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>Y</td>\n",
       "      <td>5218001</td>\n",
       "      <td>11518000</td>\n",
       "      <td>0.0382</td>\n",
       "      <td>6299999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>Y</td>\n",
       "      <td>11522001</td>\n",
       "      <td>62337000</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>50814999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>Y</td>\n",
       "      <td>62339001</td>\n",
       "      <td>62399000</td>\n",
       "      <td>1.2280</td>\n",
       "      <td>59999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>Y</td>\n",
       "      <td>62401001</td>\n",
       "      <td>62456000</td>\n",
       "      <td>1.1834</td>\n",
       "      <td>54999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1943 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "| Chromosome   | Start     | End       | CNV_value   | Length_cnvr   |\n",
       "| (category)   | (int64)   | (int64)   | (float64)   | (int64)       |\n",
       "|--------------+-----------+-----------+-------------+---------------|\n",
       "| 1            | 25001     | 45000     | 3.3208      | 19999         |\n",
       "| 1            | 48001     | 78000     | 3.2608      | 29999         |\n",
       "| 1            | 88001     | 96000     | 0.5038      | 7999          |\n",
       "| 1            | 2133001   | 2162000   | 7.152       | 28999         |\n",
       "| ...          | ...       | ...       | ...         | ...           |\n",
       "| Y            | 5218001   | 11518000  | 0.0382      | 6299999       |\n",
       "| Y            | 11522001  | 62337000  | 0.0086      | 50814999      |\n",
       "| Y            | 62339001  | 62399000  | 1.228       | 59999         |\n",
       "| Y            | 62401001  | 62456000  | 1.1834      | 54999         |\n",
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "Unstranded PyRanges object has 1,943 rows and 5 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header_list = ['Type', 'Position', 'Lenght', 'RD', 'T-TEST-EVAL', 'GAUSSIAN-EVAL', 'T-TEST-EVAL(MIDDLE)','GAUSSIAN-EVAL(MIDDLE)', 'FRAQ-READ-QUALITY', 't', 't2']\n",
    "\n",
    "t2t_data = pd.read_csv('/Users/luciabazan/t2t_LP6005441-DNA_G07_cnvs.tsv', sep='\\t' , names=header_list)\n",
    "t2t_data\n",
    "\n",
    "\n",
    "t2t_data = t2t_data[t2t_data['T-TEST-EVAL'] <= 0.05]\n",
    "t2t_data\n",
    "\n",
    "cnvnator_data = t2t_data\n",
    "\n",
    "#SEPARATE COORDINATES\n",
    "coordinates = cnvnator_data['Position'].str.split(':', expand=True)\n",
    "CHROMOSOME = coordinates[0]\n",
    "\n",
    "split_position = coordinates[1].str.split('-', expand=True)\n",
    "START = split_position[0]\n",
    "STOP = split_position[1]\n",
    "\n",
    "cnvnator_style= pd.DataFrame()\n",
    "cnvnator_style['Chromosome'] = CHROMOSOME\n",
    "cnvnator_style['Start'] = START\n",
    "cnvnator_style['End'] = STOP\n",
    "cnvnator_style['CNV_value'] = cnvnator_data['RD']\n",
    "\n",
    "\n",
    "cnvnator_style['Start'] = cnvnator_style['Start'].astype(int)\n",
    "cnvnator_style['End'] = cnvnator_style['End'].astype(int)\n",
    "cnvnator_style['CNV_value'] = cnvnator_style['CNV_value'].astype(float)\n",
    "\n",
    "#cnvnator_style['Type'] = cnvnator_data['Type']\n",
    "\n",
    "cnvnator_style['CNV_value'] = cnvnator_style['CNV_value'] * 2\n",
    "\n",
    "cnvnator_style['Chromosome'] = cnvnator_style['Chromosome'].str.replace('chr','')\n",
    "\n",
    "t2t_cnvnator = cnvnator_style\n",
    "t2t_cnvnator['Length_cnvr'] = t2t_cnvnator['End'] - t2t_cnvnator['Start']\n",
    "t2t_cnvnator= PyRanges(t2t_cnvnator)\n",
    "t2t_cnvnator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "8627f457-8192-48d5-93cf-0caf1a2186ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### t2t genes \n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator.nearest(t2t_genes)\n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator_coordinates_genes.as_df()\n",
    "t2t_cnvnator_coordinates_genes = t2t_cnvnator_coordinates_genes[t2t_cnvnator_coordinates_genes['Distance'] <1 ]\n",
    "\n",
    "## T2T telomeres\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator.nearest(t2t_telomeres)\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator_coordinates_telomeres.as_df()\n",
    "t2t_cnvnator_coordinates_telomeres = t2t_cnvnator_coordinates_telomeres[t2t_cnvnator_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs \n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator.nearest(t2t_sds)\n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator_coordinates_sds.as_df()\n",
    "t2t_cnvnator_coordinates_sds = t2t_cnvnator_coordinates_sds[t2t_cnvnator_coordinates_sds['Distance'] <1 ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7ffae1-94ba-45c0-95d7-e4426e4b28ae",
   "metadata": {},
   "source": [
    "<h1> GRCh37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "21bf4b13-6a3b-4874-bebc-d5d9f6eca33d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Anotations\n",
    "grch37_genes = pd.read_csv('../annotation_files/grch37-genes.csv')\n",
    "grch37_telomeres = pd.read_csv('../annotation_files/grch37-telomers-centromers.csv')\n",
    "grch37_sds = pd.read_csv('../annotation_files/grch37-sds.csv')\n",
    "#grch37_sds = grch37_sds[['#chr1', 'start1', 'end1']]\n",
    "\n",
    "\n",
    "## Read and merge files \n",
    "anotation = pd.read_csv('../data/SGDP_anotation.csv', sep=',', encoding='latin-1')\n",
    "\n",
    "ids_hg19 = pd.DataFrame()\n",
    "ids_hg19['Sample'] = anotation['3-Illumina_ID']\n",
    "ids_hg19['Region'] = anotation['10-Region']\n",
    "\n",
    "\n",
    "#t2t_sds.columns =['Chromosome','Start','End']\n",
    "#t2t_sds\n",
    "grch37_genes.columns = ['#\"bin\"', 'name', 'Chromosome', 'strand', 'Start', 'End', 'cdsStart',\n",
    "       'cdsEnd', 'exonCount', 'exonStarts', 'exonEnds', 'score', 'name2',\n",
    "       'cdsStartStat', 'cdsEndStat', 'exonFrames'] \n",
    "grch37_telomeres = grch37_telomeres[grch37_telomeres['type'] == 'telomere']\n",
    "grch37_sds.columns = ['#\"bin\"', 'Chromosome', 'Start', 'End', 'name', 'score', 'strand',\n",
    "       'otherChrom', 'otherStart', 'otherEnd', 'otherSize', 'uid',\n",
    "       'posBasesHit', 'testResult', 'verdict', 'chits', 'ccov', 'alignfile',\n",
    "       'alignL', 'indelN', 'indelS', 'alignB', 'matchB', 'mismatchB',\n",
    "       'transitionsB', 'transversionsB', 'fracMatch', 'fracMatchIndel', 'jcK',\n",
    "       'k2K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "92a4acd0-37cc-4be4-a1ce-1ac6fba433d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNVR</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>CNV_Start</th>\n",
       "      <th>CNV_End</th>\n",
       "      <th>RD</th>\n",
       "      <th>Lenght</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51663</th>\n",
       "      <td>1</td>\n",
       "      <td>176101.0</td>\n",
       "      <td>355800.0</td>\n",
       "      <td>CNVR_1</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>234001</td>\n",
       "      <td>238300</td>\n",
       "      <td>4</td>\n",
       "      <td>4300.0</td>\n",
       "      <td>179699.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51664</th>\n",
       "      <td>1</td>\n",
       "      <td>176101.0</td>\n",
       "      <td>355800.0</td>\n",
       "      <td>CNVR_1</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>267101</td>\n",
       "      <td>317800</td>\n",
       "      <td>0</td>\n",
       "      <td>50700.0</td>\n",
       "      <td>179699.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51665</th>\n",
       "      <td>1</td>\n",
       "      <td>449901.0</td>\n",
       "      <td>564900.0</td>\n",
       "      <td>CNVR_2</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>471301</td>\n",
       "      <td>521400</td>\n",
       "      <td>0</td>\n",
       "      <td>50100.0</td>\n",
       "      <td>114999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51666</th>\n",
       "      <td>1</td>\n",
       "      <td>1007801.0</td>\n",
       "      <td>1015200.0</td>\n",
       "      <td>CNVR_5</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>1010801</td>\n",
       "      <td>1014700</td>\n",
       "      <td>0</td>\n",
       "      <td>3900.0</td>\n",
       "      <td>7399.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51667</th>\n",
       "      <td>1</td>\n",
       "      <td>3995301.0</td>\n",
       "      <td>4237400.0</td>\n",
       "      <td>CNVR_15</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>4124401</td>\n",
       "      <td>4127800</td>\n",
       "      <td>1</td>\n",
       "      <td>3400.0</td>\n",
       "      <td>242099.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52386</th>\n",
       "      <td>9</td>\n",
       "      <td>137097701.0</td>\n",
       "      <td>138417800.0</td>\n",
       "      <td>CNVR_4724</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>138194201</td>\n",
       "      <td>138197800</td>\n",
       "      <td>1</td>\n",
       "      <td>3600.0</td>\n",
       "      <td>1320099.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52387</th>\n",
       "      <td>9</td>\n",
       "      <td>137097701.0</td>\n",
       "      <td>138417800.0</td>\n",
       "      <td>CNVR_4724</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>138214101</td>\n",
       "      <td>138217600</td>\n",
       "      <td>1</td>\n",
       "      <td>3500.0</td>\n",
       "      <td>1320099.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52388</th>\n",
       "      <td>9</td>\n",
       "      <td>138418601.0</td>\n",
       "      <td>138480200.0</td>\n",
       "      <td>CNVR_4725</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>138479101</td>\n",
       "      <td>138480200</td>\n",
       "      <td>0</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>61599.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52389</th>\n",
       "      <td>9</td>\n",
       "      <td>139663001.0</td>\n",
       "      <td>139715700.0</td>\n",
       "      <td>CNVR_4729</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>139663601</td>\n",
       "      <td>139665300</td>\n",
       "      <td>0</td>\n",
       "      <td>1700.0</td>\n",
       "      <td>52699.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52390</th>\n",
       "      <td>9</td>\n",
       "      <td>140245301.0</td>\n",
       "      <td>140317100.0</td>\n",
       "      <td>CNVR_4735</td>\n",
       "      <td>LP6005441-DNA_G07</td>\n",
       "      <td>140253401</td>\n",
       "      <td>140256000</td>\n",
       "      <td>0</td>\n",
       "      <td>2600.0</td>\n",
       "      <td>71799.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>690 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome        Start          End       CNVR          Sample_ID  \\\n",
       "51663          1     176101.0     355800.0     CNVR_1  LP6005441-DNA_G07   \n",
       "51664          1     176101.0     355800.0     CNVR_1  LP6005441-DNA_G07   \n",
       "51665          1     449901.0     564900.0     CNVR_2  LP6005441-DNA_G07   \n",
       "51666          1    1007801.0    1015200.0     CNVR_5  LP6005441-DNA_G07   \n",
       "51667          1    3995301.0    4237400.0    CNVR_15  LP6005441-DNA_G07   \n",
       "...          ...          ...          ...        ...                ...   \n",
       "52386          9  137097701.0  138417800.0  CNVR_4724  LP6005441-DNA_G07   \n",
       "52387          9  137097701.0  138417800.0  CNVR_4724  LP6005441-DNA_G07   \n",
       "52388          9  138418601.0  138480200.0  CNVR_4725  LP6005441-DNA_G07   \n",
       "52389          9  139663001.0  139715700.0  CNVR_4729  LP6005441-DNA_G07   \n",
       "52390          9  140245301.0  140317100.0  CNVR_4735  LP6005441-DNA_G07   \n",
       "\n",
       "       CNV_Start    CNV_End  RD   Lenght  Length_cnvr  \n",
       "51663     234001     238300   4   4300.0     179699.0  \n",
       "51664     267101     317800   0  50700.0     179699.0  \n",
       "51665     471301     521400   0  50100.0     114999.0  \n",
       "51666    1010801    1014700   0   3900.0       7399.0  \n",
       "51667    4124401    4127800   1   3400.0     242099.0  \n",
       "...          ...        ...  ..      ...          ...  \n",
       "52386  138194201  138197800   1   3600.0    1320099.0  \n",
       "52387  138214101  138217600   1   3500.0    1320099.0  \n",
       "52388  138479101  138480200   0   1100.0      61599.0  \n",
       "52389  139663601  139665300   0   1700.0      52699.0  \n",
       "52390  140253401  140256000   0   2600.0      71799.0  \n",
       "\n",
       "[690 rows x 10 columns]"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"../call_cnvr_cnvnator/individual_cnv_cnvr.txt\", sep='\\t')\n",
    "data.columns = ['Chromosome', 'Start', 'End','CNVR','Sample_ID','CNV_Start', 'CNV_End', 'RD', 'Lenght']\n",
    "data = data.dropna()#.describe()\n",
    "data['Length_cnvr'] = data['End'] - data['Start'] \n",
    "sample_old = data[data['Sample_ID'] == 'LP6005441-DNA_G07']\n",
    "#sample_old = old_assembly[old_assembly['Sample_ID'] == 'LP6005441-DNA_B10']\n",
    "#sample_old = sample_old[['Chromosome', 'Start', 'End', 'RD']]\n",
    "#sample_old['RD'] = sample_old['RD'] * 2\n",
    "#sample_old['Chromosome'] = str('chr')+sample_old['Chromosome']\n",
    "    #sample_old.to_csv('LP6005441-DNA_B10_grch37.bed', index=False, header=False, sep='\\t')\n",
    "    #list(sample_old.columns)\n",
    "\n",
    "#sample_old\n",
    "\n",
    "sample_old = sample_old[~sample_old['RD'].isin([2])]\n",
    "#sample_old.columns = ['Chromosome', 'Start', 'End', 'RD']\n",
    "#sample_old['assembly'] = 'GRCh37'\n",
    "sample_old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "f87a0006-3796-4c37-969d-d3dedb628afc",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Keeping only samples that are on old cnvs \n",
    "sample_old = sample_old.merge(ids_hg19, right_on=['Sample'], left_on=['Sample_ID'])\n",
    "\n",
    "#REMOVE SAMPLES WITH ABERRATIONS\n",
    "sample_old.columns = ['Chromosome', 'Start', 'End', 'CNVR_ID', 'Sample_iD', 'CNV_start', 'CNV_end', 'CNV_Value','Length', 'Length_cnvr','Sample','Region']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "f5ab003b-6d04-431b-ac90-0be08e8f2049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>LP6005441-DNA_G07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>176101.0</td>\n",
       "      <td>355800.0</td>\n",
       "      <td>179699.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>449901.0</td>\n",
       "      <td>564900.0</td>\n",
       "      <td>114999.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1007801.0</td>\n",
       "      <td>1015200.0</td>\n",
       "      <td>7399.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3995301.0</td>\n",
       "      <td>4237400.0</td>\n",
       "      <td>242099.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>8360201.0</td>\n",
       "      <td>8364200.0</td>\n",
       "      <td>3999.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>9</td>\n",
       "      <td>136453701.0</td>\n",
       "      <td>136980500.0</td>\n",
       "      <td>526799.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>9</td>\n",
       "      <td>137097701.0</td>\n",
       "      <td>138417800.0</td>\n",
       "      <td>1320099.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>9</td>\n",
       "      <td>138418601.0</td>\n",
       "      <td>138480200.0</td>\n",
       "      <td>61599.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>9</td>\n",
       "      <td>139663001.0</td>\n",
       "      <td>139715700.0</td>\n",
       "      <td>52699.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>9</td>\n",
       "      <td>140245301.0</td>\n",
       "      <td>140317100.0</td>\n",
       "      <td>71799.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>417 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Chromosome        Start          End  Length_cnvr  LP6005441-DNA_G07\n",
       "0               1     176101.0     355800.0     179699.0                2.0\n",
       "1               1     449901.0     564900.0     114999.0                0.0\n",
       "2               1    1007801.0    1015200.0       7399.0                0.0\n",
       "3               1    3995301.0    4237400.0     242099.0                1.0\n",
       "4               1    8360201.0    8364200.0       3999.0                0.0\n",
       "..            ...          ...          ...          ...                ...\n",
       "412             9  136453701.0  136980500.0     526799.0                0.0\n",
       "413             9  137097701.0  138417800.0    1320099.0                1.0\n",
       "414             9  138418601.0  138480200.0      61599.0                0.0\n",
       "415             9  139663001.0  139715700.0      52699.0                0.0\n",
       "416             9  140245301.0  140317100.0      71799.0                0.0\n",
       "\n",
       "[417 rows x 5 columns]"
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#change to df when using telomeres and centromeres filtration\n",
    "final_cnv_grch37 = sample_old.pivot_table(index=[\"Chromosome\", \"Start\", \"End\", 'Length_cnvr'], \n",
    "                    columns='Sample', \n",
    "                    values='CNV_Value',\n",
    "                        fill_value=2).reset_index()\n",
    "\n",
    "\n",
    "final_cnv_grch37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "3200a16b-2464-4e47-8063-9e46662ba8f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "final_cnv_ranges_grch37 = PyRanges(final_cnv_grch37)\n",
    "#final_cnv_grch37"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f006e6db-c6d3-4dea-9126-9f0ce4e30ed2",
   "metadata": {},
   "source": [
    "<h2> Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "8e734c44-9ebc-4144-95a1-58803104e4fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "grch37_genes['Chromosome'] = grch37_genes['Chromosome'].str.replace('chr', '')\n",
    "grch37_genes_coordinates = PyRanges(grch37_genes)\n",
    "\n",
    "grch37_telomeres['Chromosome'] = grch37_telomeres['Chromosome'].str.replace('chr', '')\n",
    "grch37_telomeres = PyRanges(grch37_telomeres)\n",
    "\n",
    "grch37_sds['Chromosome'] = grch37_sds['Chromosome'].str.replace('chr', '')\n",
    "grch37_sds = PyRanges(grch37_sds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "b3c811e9-1372-4e7f-9742-02d296c392b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genes\n",
    "grch37_coordinates_genes = final_cnv_ranges_grch37.nearest(grch37_genes_coordinates)\n",
    "grch37_coordinates_genes = grch37_coordinates_genes.as_df()\n",
    "grch37_coordinates_genes = grch37_coordinates_genes[grch37_coordinates_genes['Distance'] <1 ]\n",
    "#grch37_coordinates_genes\n",
    "\n",
    "#Telomeres\n",
    "grch37_coordinates_telomeres = final_cnv_ranges_grch37.nearest(grch37_telomeres)\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres.as_df()\n",
    "grch37_coordinates_telomeres = grch37_coordinates_telomeres[grch37_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs\n",
    "grch37_coordinates_sds = final_cnv_ranges_grch37.nearest(grch37_sds)\n",
    "grch37_coordinates_sds = grch37_coordinates_sds.as_df()\n",
    "grch37_coordinates_sds = grch37_coordinates_sds[grch37_coordinates_sds['Distance'] <1 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "d95bff56-ba99-456e-8d98-67bbd853687f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "301"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_genes = len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "grch37_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55c3b568-3b07-4850-8ec0-a81222f3d5d3",
   "metadata": {},
   "source": [
    "<h2> Peter Sudmant data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "cf3b31d8-07cb-4c2b-bc22-66e26786d3f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>177033</td>\n",
       "      <td>22</td>\n",
       "      <td>164985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12048</td>\n",
       "      <td>89237</td>\n",
       "      <td>10</td>\n",
       "      <td>77189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>228416</td>\n",
       "      <td>267160</td>\n",
       "      <td>24</td>\n",
       "      <td>38744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>471226</td>\n",
       "      <td>471326</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>318047</td>\n",
       "      <td>471226</td>\n",
       "      <td>21</td>\n",
       "      <td>153179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4476</th>\n",
       "      <td>X</td>\n",
       "      <td>75525579</td>\n",
       "      <td>75617789</td>\n",
       "      <td>3</td>\n",
       "      <td>92210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4477</th>\n",
       "      <td>X</td>\n",
       "      <td>84105273</td>\n",
       "      <td>84119490</td>\n",
       "      <td>3</td>\n",
       "      <td>14217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4478</th>\n",
       "      <td>X</td>\n",
       "      <td>134749333</td>\n",
       "      <td>134807363</td>\n",
       "      <td>3</td>\n",
       "      <td>58030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4479</th>\n",
       "      <td>X</td>\n",
       "      <td>29268810</td>\n",
       "      <td>29271507</td>\n",
       "      <td>3</td>\n",
       "      <td>2697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4480</th>\n",
       "      <td>X</td>\n",
       "      <td>91366346</td>\n",
       "      <td>91368944</td>\n",
       "      <td>6</td>\n",
       "      <td>2598</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4481 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "| Chromosome   | Start     | End       | CNV_value   | Length_cnvr   |\n",
       "| (category)   | (int64)   | (int64)   | (int64)     | (int64)       |\n",
       "|--------------+-----------+-----------+-------------+---------------|\n",
       "| 1            | 12048     | 177033    | 22          | 164985        |\n",
       "| 1            | 12048     | 89237     | 10          | 77189         |\n",
       "| 1            | 228416    | 267160    | 24          | 38744         |\n",
       "| 1            | 471226    | 471326    | 10          | 100           |\n",
       "| ...          | ...       | ...       | ...         | ...           |\n",
       "| X            | 84105273  | 84119490  | 3           | 14217         |\n",
       "| X            | 134749333 | 134807363 | 3           | 58030         |\n",
       "| X            | 29268810  | 29271507  | 3           | 2697          |\n",
       "| X            | 91366346  | 91368944  | 6           | 2598          |\n",
       "+--------------+-----------+-----------+-------------+---------------+\n",
       "Unstranded PyRanges object has 4,481 rows and 5 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sudmant = pd.read_csv('/Users/luciabazan/Downloads/sudmant_data.csv')\n",
    "sudmant = sudmant[['contig','start','end','AMR_Maya_HGDP00855_F']]\n",
    "sudmant.columns =['Chromosome', 'Start', 'End', 'CNV_value']\n",
    "\n",
    "\n",
    "#sudmant['Chromosome'] = sudmant['Chromosome'].str.replace('chr','hs')\n",
    "n_uncalled = np.sum(sudmant==-1,1)\n",
    "sudmant = sudmant[n_uncalled==0]\n",
    "\n",
    "sudmant['Length_cnvr'] = sudmant['End'] - sudmant['Start'] \n",
    "sudmant = sudmant[~sudmant['CNV_value'].isin([2])]\n",
    "\n",
    "\n",
    "sudmant['Chromosome'] = sudmant['Chromosome'].str.replace('chr', '')\n",
    "sudmant = PyRanges(sudmant)\n",
    "sudmant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "a64eefc5-41c8-45f7-b64b-d46d3aa43daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genes\n",
    "peter_coordinates_genes = sudmant.nearest(grch37_genes_coordinates)\n",
    "peter_coordinates_genes = peter_coordinates_genes.as_df()\n",
    "peter_coordinates_genes = peter_coordinates_genes[peter_coordinates_genes['Distance'] <1 ]\n",
    "#grch37_coordinates_genes\n",
    "\n",
    "#Telomeres\n",
    "peter_coordinates_telomeres = sudmant.nearest(grch37_telomeres)\n",
    "peter_coordinates_telomeres = peter_coordinates_telomeres.as_df()\n",
    "peter_coordinates_telomeres = peter_coordinates_telomeres[peter_coordinates_telomeres['Distance'] <1 ]\n",
    "\n",
    "#SDs\n",
    "peter_coordinates_sds = sudmant.nearest(grch37_sds)\n",
    "peter_coordinates_sds = peter_coordinates_sds.as_df()\n",
    "peter_coordinates_sds = peter_coordinates_sds[peter_coordinates_sds['Distance'] <1 ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90789782-536a-4f10-b4b5-60e5d62a4cee",
   "metadata": {},
   "source": [
    "GRCh37 Duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "6365e24d-79fe-4c7d-92b7-9e4ab6dc3ddc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Calls</th>\n",
       "      <th>Mean size(Bp)</th>\n",
       "      <th>Literature</th>\n",
       "      <th>Mean size(Bp) Literature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>417</td>\n",
       "      <td>160454.0</td>\n",
       "      <td>4481</td>\n",
       "      <td>32000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>301</td>\n",
       "      <td>211600.0</td>\n",
       "      <td>3179</td>\n",
       "      <td>39641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>6</td>\n",
       "      <td>164382.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>163</td>\n",
       "      <td>310071.0</td>\n",
       "      <td>3849</td>\n",
       "      <td>35872.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>1943</td>\n",
       "      <td>98033.0</td>\n",
       "      <td>14872</td>\n",
       "      <td>99326.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>433</td>\n",
       "      <td>207664.0</td>\n",
       "      <td>12264</td>\n",
       "      <td>108634.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>7</td>\n",
       "      <td>127570.0</td>\n",
       "      <td>42</td>\n",
       "      <td>8181399.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>630</td>\n",
       "      <td>158253.0</td>\n",
       "      <td>3374</td>\n",
       "      <td>268004.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data  Calls  Mean size(Bp)  Literature  Mean size(Bp) Literature\n",
       "0       GRCh37    417       160454.0        4481                   32000.0\n",
       "1        Genes    301       211600.0        3179                   39641.0\n",
       "2    Telomeres      6       164382.0           0                       NaN\n",
       "3          SDs    163       310071.0        3849                   35872.0\n",
       "4    T2T-CHM13   1943        98033.0       14872                   99326.0\n",
       "5        Genes    433       207664.0       12264                  108634.0\n",
       "6    Telomeres      7       127570.0          42                 8181399.0\n",
       "7          SDs    630       158253.0        3374                  268004.0"
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'\n",
    "             ], \n",
    "     'Calls': [len(final_cnv_grch37), \n",
    "               len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),  \n",
    "               len(t2t_cnvnator), \n",
    "               len(t2t_cnvnator_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(t2t_cnvnator_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(t2t_cnvnator_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())\n",
    "              ], \n",
    "     'Mean size(Bp)':[final_cnv_grch37['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_genes['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_sds['Length_cnvr'].mean().round(), \n",
    "                      t2t_cnvnator.as_df()['Length_cnvr'].mean().round(),\n",
    "                      t2t_cnvnator_coordinates_genes['Length_cnvr'].mean().round(),\n",
    "                      t2t_cnvnator_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      t2t_cnvnator_coordinates_sds['Length_cnvr'].mean().round()],\n",
    "     'Literature':[len(sudmant), \n",
    "                  len(peter_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(peter_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(peter_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(final_cnv_ranges), \n",
    "                  len(coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), ], \n",
    "     'Mean size(Bp) Literature':[sudmant.as_df()['Length_cnvr'].mean().round(), \n",
    "                      peter_coordinates_genes['Length_cnvr'].mean().round(), \n",
    "                      peter_coordinates_telomeres['Length_cnvr'].mean(), \n",
    "                      peter_coordinates_sds['Length_cnvr'].mean().round(), \n",
    "                      final_cnv['Length_cnvr'].mean().round(),\n",
    "                      coordinates_genes['Length_cnvr'].mean().round(),\n",
    "                      coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      coordinates_sds['Length_cnvr'].mean().round()]\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=d)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "b2db73c9-02d8-47e0-a506-82ac31c4ba7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Total</th>\n",
       "      <th>Mean size(Bp)</th>\n",
       "      <th>Deletions</th>\n",
       "      <th>Duplications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37</td>\n",
       "      <td>417</td>\n",
       "      <td>160454.0</td>\n",
       "      <td>417</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genes</td>\n",
       "      <td>301</td>\n",
       "      <td>211600.0</td>\n",
       "      <td>301</td>\n",
       "      <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>6</td>\n",
       "      <td>164382.0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SDs</td>\n",
       "      <td>163</td>\n",
       "      <td>310071.0</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2T-CHM13</td>\n",
       "      <td>14872</td>\n",
       "      <td>108634.0</td>\n",
       "      <td>14872</td>\n",
       "      <td>14872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genes</td>\n",
       "      <td>12264</td>\n",
       "      <td>99326.0</td>\n",
       "      <td>12264</td>\n",
       "      <td>12264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Telomeres</td>\n",
       "      <td>42</td>\n",
       "      <td>8181399.0</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SDs</td>\n",
       "      <td>3374</td>\n",
       "      <td>268004.0</td>\n",
       "      <td>3374</td>\n",
       "      <td>3374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Data  Total  Mean size(Bp)  Deletions  Duplications\n",
       "0       GRCh37    417       160454.0        417           417\n",
       "1        Genes    301       211600.0        301           301\n",
       "2    Telomeres      6       164382.0          6             6\n",
       "3          SDs    163       310071.0        163           163\n",
       "4    T2T-CHM13  14872       108634.0      14872         14872\n",
       "5        Genes  12264        99326.0      12264         12264\n",
       "6    Telomeres     42      8181399.0         42            42\n",
       "7          SDs   3374       268004.0       3374          3374"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'Data': ['GRCh37', '  Genes','  Telomeres','  SDs','T2T-CHM13', '  Genes','  Telomeres','  SDs'], \n",
    "     'Total': [len(final_cnv_grch37), \n",
    "               len(grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),  \n",
    "               len(final_cnv), \n",
    "               len(coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "               len(coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates())], \n",
    "     'Mean size(Bp)':[final_cnv_grch37['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_genes['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      grch37_coordinates_sds['Length_cnvr'].mean().round(), \n",
    "                      coordinates_genes['Length_cnvr'].mean().round(),\n",
    "                      final_cnv['Length_cnvr'].mean().round(),\n",
    "                      coordinates_telomeres['Length_cnvr'].mean().round(), \n",
    "                      coordinates_sds['Length_cnvr'].mean().round()],\n",
    "     'Deletions':[len(grch37_deletions), \n",
    "                  len(deletions_grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(deletions_grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(deletions_grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(t2t_deletions), \n",
    "                  len(deletions_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                  len(deletions_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                  len(deletions_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()), ], \n",
    "     'Duplications':[len(grch37_duplications), \n",
    "                     len(duplications_grch37_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(duplications_grch37_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                     len(duplications_grch37_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(t2t_duplications), \n",
    "                     len(duplications_t2t_coordinates_genes[['Chromosome','Start', 'End']].drop_duplicates()),\n",
    "                     len(duplications_t2t_coordinates_telomeres[['Chromosome','Start', 'End']].drop_duplicates()), \n",
    "                     len(duplications_t2t_coordinates_sds[['Chromosome','Start', 'End']].drop_duplicates()) ]\n",
    "    \n",
    "    }\n",
    "df = pd.DataFrame(data=d)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "272eef6c-7ec2-444d-9396-1f972933c73a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "160454.0"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_cnv_grch37['Length_cnvr'].mean().round()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "be934881-8a16-42b7-aee7-9f936e845cf2",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "('PyRanges object has no attribute', 'mean')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[279], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m t2t_cnvnator[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLength_cnvr\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mmean()\u001b[38;5;241m.\u001b[39mround()\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/pyranges/pyranges_main.py:265\u001b[0m, in \u001b[0;36mPyRanges.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    240\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Return column.\u001b[39;00m\n\u001b[1;32m    241\u001b[0m \n\u001b[1;32m    242\u001b[0m \u001b[38;5;124;03mParameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    260\u001b[0m \u001b[38;5;124;03mName: Start, dtype: int64\u001b[39;00m\n\u001b[1;32m    261\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    263\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpyranges\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmethods\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mattr\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m _getattr\n\u001b[0;32m--> 265\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _getattr(\u001b[38;5;28mself\u001b[39m, name)\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.11/site-packages/pyranges/methods/attr.py:65\u001b[0m, in \u001b[0;36m_getattr\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m     63\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m pd\u001b[38;5;241m.\u001b[39mconcat([df[name] \u001b[38;5;28;01mfor\u001b[39;00m df \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalues()])\n\u001b[1;32m     64\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 65\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPyRanges object has no attribute\u001b[39m\u001b[38;5;124m\"\u001b[39m, name)\n",
      "\u001b[0;31mAttributeError\u001b[0m: ('PyRanges object has no attribute', 'mean')"
     ]
    }
   ],
   "source": [
    "t2t_cnvnator['Length_cnvr'].mean().round()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "2899651b-9ed8-49c8-bf2a-06cac346ce5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          19999\n",
       "1          29999\n",
       "2           7999\n",
       "3          28999\n",
       "4          29999\n",
       "          ...   \n",
       "1938     2962999\n",
       "1939     6299999\n",
       "1940    50814999\n",
       "1941       59999\n",
       "1942       54999\n",
       "Name: Length_cnvr, Length: 1943, dtype: int64"
      ]
     },
     "execution_count": 300,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2t_cnvnator.as_df()['Length_cnvr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "871e7b13-7aab-4546-a23f-3724cb1ddb00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>CNV_value</th>\n",
       "      <th>Length_cnvr</th>\n",
       "      <th>#\"bin\"</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>ix</th>\n",
       "      <th>n</th>\n",
       "      <th>size</th>\n",
       "      <th>type</th>\n",
       "      <th>bridge</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Start, End, CNV_value, Length_cnvr, #\"bin\", Start_b, End_b, ix, n, size, type, bridge, Distance]\n",
       "Index: []"
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peter_coordinates_telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed1f8e95-2418-464b-9402-cfa0f13215f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
